// @openuidev/react-ui/dist/components/Charts/BarChart/BarChart.js
import { jsx as _jsx, jsxs as _jsxs } from "react/jsx-runtime";
import clsx from "clsx";
import React, { useCallback, useEffect, useId, useMemo, useRef, useState } from "react";
import { Bar, BarChart as RechartsBarChart, XAxis, YAxis } from "recharts";
import { usePrintContext } from "../../../context/PrintContext";
import { useTheme } from "../../ThemeProvider";
import { ChartContainer, ChartTooltip } from "../Charts";
import { SideBarTooltipProvider } from "../context/SideBarTooltipContext";
import { useExportChartData, useMaxLabelHeight, useTransformedKeys, useYAxisLabelWidth, } from "../hooks";
import { cartesianGrid, CustomTooltipContent, DefaultLegend, LineInBarShape, SideBarTooltip, XAxisTick, YAxisTick, } from "../shared";
import { ScrollButtonsHorizontal } from "../shared/ScrollButtonsHorizontal/ScrollButtonsHorizontal";
import { useChartPalette } from "../utils/PalletUtils";
import { LabelTooltipProvider } from "../shared/LabelTooltip/LabelTooltip";
import { findNearestSnapPosition, getBarStackInfo, getRadiusArray, } from "../utils/BarCharts/BarChartsUtils";
import { get2dChartConfig, getColorForDataKey, getDataKeys, getLegendItems, } from "../utils/dataUtils";
import { BAR_WIDTH, getPadding, getSnapPositions, getWidthOfData, getWidthOfGroup, } from "./utils/BarChartUtils";
const BAR_GAP = 10; // Gap between bars
const BAR_CATEGORY_GAP = "20%"; // Gap between categories
const BAR_INTERNAL_LINE_WIDTH = 1;
const BAR_RADIUS = 4;
const CHART_CONTAINER_BOTTOM_MARGIN = 10;
const BarChartComponent = ({ data, categoryKey, theme = "ocean", customPalette, variant = "grouped", tickVariant = "multiLine", grid = true, icons = {}, radius, isAnimationActive = false, showYAxis = true, xAxisLabel, yAxisLabel, legend = true, className, height, width, }) => {
    const printContext = usePrintContext();
    isAnimationActive = printContext ? false : isAnimationActive;
    const widthOfGroup = getWidthOfGroup(data, categoryKey, variant);
    const maxLabelHeight = useMaxLabelHeight(data, categoryKey, tickVariant, widthOfGroup);
    const dataKeys = useMemo(() => {
        return getDataKeys(data, categoryKey);
    }, [data, categoryKey]);
    const { yAxisWidth, setLabelWidth } = useYAxisLabelWidth(data, dataKeys);
    const transformedKeys = useTransformedKeys(dataKeys);
    const colors = useChartPalette({
        chartThemeName: theme,
        customPalette,
        themePaletteName: "barChartPalette",
        dataLength: dataKeys.length,
    });
    const chartConfig = useMemo(() => {
        return get2dChartConfig(dataKeys, colors, transformedKeys, undefined, icons);
    }, [dataKeys, icons, colors, transformedKeys]);
    const chartContainerRef = useRef(null);
    const mainContainerRef = useRef(null);
    const [containerWidth, setContainerWidth] = useState(0);
    const [canScrollLeft, setCanScrollLeft] = useState(false);
    const [canScrollRight, setCanScrollRight] = useState(false);
    const [hoveredCategory, setHoveredCategory] = useState(null);
    const [isLegendExpanded, setIsLegendExpanded] = useState(false);
    const [isSideBarTooltipOpen, setIsSideBarTooltipOpen] = useState(false);
    const [sideBarTooltipData, setSideBarTooltipData] = useState({
        title: "",
        values: [],
    });
    // Use provided width or observed width
    const effectiveWidth = useMemo(() => {
        return width ?? containerWidth;
    }, [width, containerWidth]);
    // need this to calculate the padding for the chart container, because the y-axis is rendered in a separate chart
    const effectiveContainerWidth = useMemo(() => {
        const dynamicYAxisWidth = showYAxis ? yAxisWidth : 0;
        return Math.max(0, effectiveWidth - dynamicYAxisWidth);
    }, [effectiveWidth, showYAxis, yAxisWidth]);
    const padding = useMemo(() => {
        return getPadding(data, categoryKey, effectiveContainerWidth, variant);
    }, [data, categoryKey, effectiveContainerWidth, variant]);
    const dataWidth = useMemo(() => {
        return getWidthOfData(data, categoryKey, variant);
    }, [data, categoryKey, variant]);
    // Calculate snap positions for proper group alignment
    const snapPositions = useMemo(() => {
        return getSnapPositions(data, categoryKey, variant);
    }, [data, categoryKey, variant]);
    // self note:
    // Use provided height or calculated height based on container width
    // if height is provided, it will be used to set the height of the chart
    // if height is not provided, it will be calculated based on the container width (effectiveWidth)
    // getChartHeight(effectiveWidth) this function is not used here, request of the designer, we will use fix height
    // 296 is the height of the chart by default, given by designer
    // we want to chart to scale with width but height will be fixed
    const chartHeight = useMemo(() => {
        return height ?? 296 + maxLabelHeight;
    }, [height, maxLabelHeight]);
    // Check scroll boundaries
    const updateScrollState = useCallback(() => {
        if (mainContainerRef.current) {
            const { scrollLeft, scrollWidth, clientWidth } = mainContainerRef.current;
            setCanScrollLeft(scrollLeft > 0);
            setCanScrollRight(scrollLeft < scrollWidth - clientWidth - 1); // -1 for floating point precision
        }
    }, []);
    const scrollLeft = useCallback(() => {
        if (mainContainerRef.current) {
            const currentScroll = mainContainerRef.current.scrollLeft;
            const targetIndex = findNearestSnapPosition(snapPositions, currentScroll, "left");
            const targetPosition = snapPositions[targetIndex] ?? 0;
            mainContainerRef.current.scrollTo({
                left: targetPosition,
                behavior: "smooth",
            });
        }
    }, [snapPositions]);
    const scrollRight = useCallback(() => {
        if (mainContainerRef.current) {
            const currentScroll = mainContainerRef.current.scrollLeft;
            const targetIndex = findNearestSnapPosition(snapPositions, currentScroll, "right");
            const targetPosition = snapPositions[targetIndex] ?? 0;
            mainContainerRef.current.scrollTo({
                left: targetPosition,
                behavior: "smooth",
            });
        }
    }, [snapPositions]);
    useEffect(() => {
        // Only set up ResizeObserver if width is not provided
        if (width || !chartContainerRef.current) {
            return () => { };
        }
        const resizeObserver = new ResizeObserver((entries) => {
            // there is only one entry in the entries array because we are observing the chart container
            for (const entry of entries) {
                setContainerWidth(entry.contentRect.width);
            }
        });
        resizeObserver.observe(chartContainerRef.current);
        return () => {
            resizeObserver.disconnect();
        };
    }, [width]);
    // Update scroll state when container width or data width changes
    useEffect(() => {
        updateScrollState();
    }, [effectiveWidth, dataWidth, updateScrollState]);
    useEffect(() => {
        setIsSideBarTooltipOpen(false);
        setIsLegendExpanded(false);
    }, [dataKeys]);
    // Add scroll event listener to update button states
    useEffect(() => {
        const mainContainer = mainContainerRef.current;
        if (!mainContainer)
            return;
        const handleScroll = () => {
            updateScrollState();
        };
        mainContainer.addEventListener("scroll", handleScroll);
        return () => {
            mainContainer.removeEventListener("scroll", handleScroll);
        };
    }, [updateScrollState]);
    // Memoize legend items creation
    const legendItems = useMemo(() => {
        return getLegendItems(dataKeys, colors, icons);
    }, [dataKeys, colors, icons]);
    const exportData = useExportChartData({
        type: "bar",
        data,
        categoryKey: categoryKey,
        dataKeys,
        colors,
        legend,
        xAxisLabel,
        yAxisLabel,
    });
    const id = useId();
    const yAxis = useMemo(() => {
        if (!showYAxis) {
            return null;
        }
        return (_jsx("div", { className: "openui-bar-chart-y-axis-container", children: _jsxs(RechartsBarChart, { width: yAxisWidth, height: chartHeight, data: data, stackOffset: "sign", margin: {
                    top: 20,
                    bottom: maxLabelHeight + CHART_CONTAINER_BOTTOM_MARGIN, // this is required for to give space for x-axis
                    left: 0,
                    right: 0,
                }, children: [_jsx(YAxis, { width: yAxisWidth, tickLine: false, axisLine: false, tick: _jsx(YAxisTick, { setLabelWidth: setLabelWidth }) }), dataKeys.map((key) => {
                        return (_jsx(Bar, { dataKey: key, fill: "transparent", stackId: variant === "stacked" ? "a" : undefined, isAnimationActive: false, maxBarSize: 0 }, `yaxis-bar-chart-${key}`));
                    })] }, `y-axis-bar-chart-${id}`) }));
    }, [showYAxis, chartHeight, data, dataKeys, variant, id, maxLabelHeight, yAxisWidth]);
    // Handle mouse events for group hovering
    const handleChartMouseMove = useCallback((state) => {
        if (state && state.activeLabel !== undefined) {
            setHoveredCategory(state.activeLabel);
        }
    }, []);
    const handleChartMouseLeave = useCallback(() => {
        setHoveredCategory(null);
    }, []);
    const handleXAxisTickMouseEnter = useCallback((tickProps) => {
        if (typeof tickProps.payload?.value === "string") {
            setHoveredCategory(tickProps.payload.value);
        }
    }, []);
    const handleXAxisTickMouseLeave = useCallback(() => {
        setHoveredCategory(null);
    }, []);
    const { mode, theme: userTheme } = useTheme();
    const barInternalLineColor = useMemo(() => {
        if (mode === "light") {
            return "rgba(255, 255, 255, 0.3)";
        }
        return "rgba(0, 0, 0, 0.3)";
    }, [mode]);
    const calculatedRadius = useMemo(() => {
        let radiusValue = BAR_RADIUS;
        if (typeof radius === "number") {
            radiusValue = radius;
        }
        else {
            const radiusTheme = userTheme.radius2xs;
            if (radiusTheme) {
                radiusValue = typeof radiusTheme === "string" ? parseInt(radiusTheme) : radiusTheme;
            }
        }
        return radiusValue;
    }, [userTheme.radius2xs, radius]);
    const onBarsClick = useCallback((data) => {
        if (data?.activePayload?.length && data.activePayload.length > 10) {
            setIsSideBarTooltipOpen(true);
            setSideBarTooltipData({
                title: data.activeLabel,
                values: data.activePayload.map((payload) => ({
                    value: payload.value,
                    label: payload.name || payload.dataKey,
                    color: getColorForDataKey(payload.dataKey, dataKeys, colors),
                })),
            });
        }
    }, [dataKeys, colors]);
    const barElements = useMemo(() => {
        return dataKeys.map((key) => {
            const transformedKey = transformedKeys[key];
            const color = `var(--color-${transformedKey})`;
            return (_jsx(Bar, { dataKey: key, fill: color, stackId: variant === "stacked" ? "a" : undefined, isAnimationActive: isAnimationActive, maxBarSize: BAR_WIDTH, barSize: BAR_WIDTH, shape: (props) => {
                    const { payload, value, dataKey } = props;
                    const { isNegative, isFirstInStack, isLastInStack, hasNegativeValueInStack } = getBarStackInfo(variant, value, dataKey, payload, dataKeys);
                    const customRadius = getRadiusArray(variant, calculatedRadius, "vertical", isFirstInStack, isLastInStack, isNegative);
                    return (_jsx(LineInBarShape, { ...props, radius: customRadius, internalLineColor: barInternalLineColor, internalLineWidth: BAR_INTERNAL_LINE_WIDTH, isHovered: hoveredCategory !== null, hoveredCategory: hoveredCategory, categoryKey: categoryKey, variant: variant, hasNegativeValueInStack: hasNegativeValueInStack }));
                } }, `main-${key}`));
        });
    }, [
        dataKeys,
        transformedKeys,
        variant,
        calculatedRadius,
        isAnimationActive,
        barInternalLineColor,
        hoveredCategory,
        categoryKey,
    ]);
    return (_jsx(LabelTooltipProvider, { children: _jsx(SideBarTooltipProvider, { isSideBarTooltipOpen: isSideBarTooltipOpen, setIsSideBarTooltipOpen: setIsSideBarTooltipOpen, data: sideBarTooltipData, setData: setSideBarTooltipData, children: _jsxs("div", { className: clsx("openui-bar-chart-container", className), "data-openui-chart": exportData, style: {
                    width: width ? `${width}px` : undefined,
                }, children: [_jsxs("div", { className: "openui-bar-chart-container-inner", ref: chartContainerRef, children: [yAxis, _jsx("div", { className: "openui-bar-chart-main-container", ref: mainContainerRef, children: _jsx(ChartContainer, { config: chartConfig, style: { width: dataWidth, minWidth: "100%", height: chartHeight }, rechartsProps: {
                                        width: "100%",
                                        height: "100%",
                                        minHeight: 1,
                                        minWidth: 1,
                                        initialDimension: { width: 1, height: 1 },
                                    }, children: _jsxs(RechartsBarChart, { stackOffset: "sign", accessibilityLayer: true, data: data, margin: {
                                            top: 20,
                                            bottom: CHART_CONTAINER_BOTTOM_MARGIN,
                                        }, onClick: onBarsClick, onMouseMove: handleChartMouseMove, onMouseLeave: handleChartMouseLeave, barGap: BAR_GAP, barCategoryGap: BAR_CATEGORY_GAP, children: [grid && cartesianGrid(), _jsx(XAxis, { dataKey: categoryKey, tickLine: false, axisLine: false, textAnchor: "middle", interval: 0, height: maxLabelHeight, tick: _jsx(XAxisTick, { variant: tickVariant, widthOfGroup: widthOfGroup, labelHeight: maxLabelHeight, onMouseEnter: handleXAxisTickMouseEnter, onMouseLeave: handleXAxisTickMouseLeave }), orientation: "bottom", 
                                                // gives the padding on the 2 sides see the function for reference
                                                padding: padding }), _jsx(ChartTooltip
                                            // cursor={<SimpleCursor />}
                                            , { 
                                                // cursor={<SimpleCursor />}
                                                cursor: {
                                                    fill: "var(--openui-highlight)",
                                                    stroke: "var(--openui-stroke-default)",
                                                    opacity: 1,
                                                    strokeWidth: 1,
                                                }, content: _jsx(CustomTooltipContent, { parentRef: mainContainerRef }), offset: 15 }), barElements] }, `bar-chart-${id}`) }) }), isSideBarTooltipOpen && _jsx(SideBarTooltip, { height: chartHeight })] }), _jsx(ScrollButtonsHorizontal, { dataWidth: dataWidth, effectiveWidth: effectiveWidth, canScrollLeft: canScrollLeft, canScrollRight: canScrollRight, isSideBarTooltipOpen: isSideBarTooltipOpen, onScrollLeft: scrollLeft, onScrollRight: scrollRight }), legend && (_jsx(DefaultLegend, { items: legendItems, yAxisLabel: yAxisLabel, xAxisLabel: xAxisLabel, containerWidth: effectiveWidth, isExpanded: isLegendExpanded, setIsExpanded: setIsLegendExpanded }))] }) }) }));
};
// Added React.memo for performance optimization to avoid unnecessary re-renders
export const BarChart = React.memo(BarChartComponent);
