// @openuidev/react-ui/dist/components/ImageBlock/stories/ImageBlock.stories.js
import { jsx as _jsx } from "react/jsx-runtime";
import { ImageBlock } from "../ImageBlock";
const meta = {
    title: "Components/ImageBlock",
    component: ImageBlock,
    tags: ["autodocs"],
    parameters: {
        docs: {
            description: {
                component: "```tsx\nimport { ImageBlock } from '@openuidev/react-ui';\n```\n\nA hero/showcase image with a blurred backdrop effect, responsive sizing, and loading/error states.",
            },
        },
    },
    argTypes: {
        src: {
            control: "text",
            description: "Image source URL",
            table: { category: "Content" },
        },
        alt: {
            control: "text",
            description: "Alternative text for the image",
            table: { category: "Content" },
        },
        className: {
            control: false,
            description: "Additional CSS classes",
            table: { category: "Styling" },
        },
    },
};
export default meta;
const sampleImage = "https://images.unsplash.com/photo-1738168246881-40f35f8aba0a?w=800&auto=format&fit=crop&q=60";
export const Default = {
    args: {
        src: sampleImage,
        alt: "Sample landscape",
    },
    decorators: [
        (Story) => (_jsx("div", { style: { maxWidth: 600, margin: "auto" }, children: _jsx(Story, {}) })),
    ],
};
export const Mobile = {
    args: {
        src: sampleImage,
        alt: "Mobile view",
    },
    decorators: [
        (Story) => (_jsx("div", { style: { maxWidth: 320, margin: "auto" }, children: _jsx(Story, {}) })),
    ],
};
export const Loading = {
    args: {
        src: sampleImage,
        alt: "Loading state",
        imageLoading: true,
    },
    decorators: [
        (Story) => (_jsx("div", { style: { maxWidth: 600, margin: "auto" }, children: _jsx(Story, {}) })),
    ],
};
export const Error = {
    args: {
        src: "https://invalid-url-that-will-fail.example/image.jpg",
        alt: "Broken image",
    },
    decorators: [
        (Story) => (_jsx("div", { style: { maxWidth: 600, margin: "auto" }, children: _jsx(Story, {}) })),
    ],
};
