// @openuidev/react-ui/dist/components/CheckBoxItem/index.cjs
Object.defineProperty(exports, Symbol.toStringTag, { value: "Module" });
const require_chunk = require("../../chunk-CKQMccvm.cjs");
let clsx = require("clsx");
clsx = require_chunk.__toESM(clsx);
let react = require("react");
let react_jsx_runtime = require("react/jsx-runtime");
let _radix_ui_react_checkbox = require("@radix-ui/react-checkbox");
_radix_ui_react_checkbox = require_chunk.__toESM(_radix_ui_react_checkbox);
//#region src/components/CheckBoxItem/CheckBoxItem.tsx
const CheckBoxItem = (0, react.forwardRef)((props, ref) => {
	const { label, description, onChange, className, disabled, required, ...rest } = props;
	const id = (0, react.useId)();
	return /* @__PURE__ */ (0, react_jsx_runtime.jsxs)("label", {
		htmlFor: id,
		className: "openui-checkbox-item-container",
		children: [/* @__PURE__ */ (0, react_jsx_runtime.jsx)(_radix_ui_react_checkbox.Root, {
			...rest,
			ref,
			onCheckedChange: onChange,
			id,
			className: (0, clsx.default)("openui-checkbox-item-root", className),
			disabled,
			required,
			children: /* @__PURE__ */ (0, react_jsx_runtime.jsx)(_radix_ui_react_checkbox.Indicator, {
				className: "openui-checkbox-item-indicator",
				children: /* @__PURE__ */ (0, react_jsx_runtime.jsx)("svg", {
					width: "10",
					height: "8",
					viewBox: "0 0 10 8",
					fill: "none",
					xmlns: "http://www.w3.org/2000/svg",
					children: /* @__PURE__ */ (0, react_jsx_runtime.jsx)("path", {
						d: "M9 1L3.5 6.5L1 4",
						stroke: "currentColor",
						strokeWidth: "1.5",
						strokeLinecap: "round",
						strokeLinejoin: "round"
					})
				})
			})
		}), /* @__PURE__ */ (0, react_jsx_runtime.jsxs)("div", {
			className: "openui-checkbox-item-content",
			children: [label && /* @__PURE__ */ (0, react_jsx_runtime.jsx)("label", {
				htmlFor: id,
				className: "openui-checkbox-item-label",
				children: label
			}), description && /* @__PURE__ */ (0, react_jsx_runtime.jsx)("p", {
				className: "openui-checkbox-item-description",
				children: description
			})]
		})]
	});
});
CheckBoxItem.displayName = "CheckBoxItem";
//#endregion
exports.CheckBoxItem = CheckBoxItem;
