// @openuidev/react-ui/dist/components/Charts/shared/PortalTooltip/FloatingUIPortal.js
import { jsx as _jsx, Fragment as _Fragment, jsxs as _jsxs } from "react/jsx-runtime";
import { autoUpdate, flip, hide, offset, useFloating } from "@floating-ui/react-dom";
import clsx from "clsx";
import { useEffect } from "react";
import { createPortal } from "react-dom";
import { useTheme } from "../../../ThemeProvider";
export const FloatingUIPortal = ({ children, className = "", chartId, portalContainer, position, placement = "right-start", offsetDistance = 20, }) => {
    const { refs, floatingStyles, update } = useFloating({
        placement,
        middleware: [offset(offsetDistance), flip(), hide()],
        whileElementsMounted: autoUpdate,
    });
    const { portalThemeClassName } = useTheme();
    useEffect(() => {
        if (position) {
            update();
        }
    }, [position, update]);
    return (_jsxs(_Fragment, { children: [_jsx("div", { ref: refs.setReference, style: {
                    position: "absolute",
                    top: position?.y,
                    left: position?.x,
                } }), createPortal(_jsx("div", { ref: refs.setFloating, className: clsx("openui-portal-tooltip", portalThemeClassName, className), "data-chart": chartId, style: floatingStyles, children: children }), portalContainer?.current || document.body)] }));
};
