// @openuidev/react-ui/dist/genui-lib/Accordion/index.js
"use client";
import { jsx as _jsx, jsxs as _jsxs } from "react/jsx-runtime";
import { defineComponent } from "@openuidev/react-lang";
import React from "react";
import { z } from "zod/v4";
import { Accordion as OpenUIAccordion, AccordionContent as OpenUIAccordionContent, AccordionItem as OpenUIAccordionItem, AccordionTrigger as OpenUIAccordionTrigger, } from "../../components/Accordion";
import { AccordionItemSchema } from "./schema";
export { AccordionItemSchema } from "./schema";
export const AccordionItem = defineComponent({
    name: "AccordionItem",
    props: AccordionItemSchema,
    description: "value is unique id, trigger is section title",
    component: () => null,
});
export const Accordion = defineComponent({
    name: "Accordion",
    props: z.object({
        items: z.array(AccordionItem.ref),
    }),
    description: "Collapsible sections",
    component: ({ props, renderNode }) => {
        const items = props.items ?? [];
        const [openItem, setOpenItem] = React.useState("");
        const userHasInteracted = React.useRef(false);
        const prevContentSizes = React.useRef({});
        React.useEffect(() => {
            const first = items[0];
            if (items.length && !openItem && first) {
                setOpenItem(first.props.value);
            }
        }, [items.length, openItem]);
        React.useEffect(() => {
            if (userHasInteracted.current)
                return;
            let candidate = null;
            const nextSizes = {};
            for (const item of items) {
                const size = JSON.stringify(item.props.content).length;
                const prevSize = prevContentSizes.current[item.props.value] ?? 0;
                nextSizes[item.props.value] = size;
                if (size > prevSize) {
                    candidate = item.props.value;
                }
            }
            prevContentSizes.current = nextSizes;
            if (candidate && candidate !== openItem) {
                setOpenItem(candidate);
            }
        });
        const handleValueChange = (value) => {
            userHasInteracted.current = true;
            setOpenItem(value);
        };
        if (!items.length)
            return null;
        return (_jsx(OpenUIAccordion, { type: "single", collapsible: true, value: openItem, onValueChange: handleValueChange, children: items.map((item) => (_jsxs(OpenUIAccordionItem, { value: item.props.value, children: [_jsx(OpenUIAccordionTrigger, { text: item.props.trigger }), _jsx(OpenUIAccordionContent, { children: renderNode(item.props.content) })] }, item.props.value))) }));
    },
});
