// @openuidev/react-ui/dist/components/Shell/index.cjs
Object.defineProperty(exports, Symbol.toStringTag, { value: "Module" });
const require_ArtifactOverlay = require("../../ArtifactOverlay-7Tt6OY58.cjs");
const require_ArtifactPanel = require("../../ArtifactPanel-c1WD8X43.cjs");
const require_Shell = require("../../Shell-CXCVPlSX.cjs");
exports.ArtifactOverlay = require_ArtifactOverlay.ArtifactOverlay;
exports.ArtifactPanel = require_ArtifactPanel.ArtifactPanel;
exports.ArtifactPortalTarget = require_ArtifactOverlay.ArtifactPortalTarget;
exports.AssistantMessageContainer = require_Shell.AssistantMessageContainer;
exports.Composer = require_Shell.Composer;
exports.Container = require_Shell.Container;
exports.ConversationStarter = require_Shell.ConversationStarter;
exports.DesktopWelcomeComposer = require_Shell.DesktopWelcomeComposer;
exports.MessageLoading = require_Shell.MessageLoading;
exports.Messages = require_Shell.Messages;
exports.MobileHeader = require_Shell.MobileHeader;
exports.NewChatButton = require_Shell.NewChatButton;
exports.RenderMessage = require_Shell.RenderMessage;
exports.ScrollArea = require_Shell.ScrollArea;
exports.ShellStoreContext = require_ArtifactOverlay.ShellStoreContext;
exports.ShellStoreProvider = require_ArtifactOverlay.ShellStoreProvider;
exports.SidebarContainer = require_Shell.SidebarContainer;
exports.SidebarContent = require_Shell.SidebarContent;
exports.SidebarHeader = require_Shell.SidebarHeader;
exports.SidebarSeparator = require_Shell.SidebarSeparator;
exports.ThreadButton = require_Shell.ThreadButton;
exports.ThreadContainer = require_Shell.ThreadContainer;
exports.ThreadError = require_Shell.ThreadError;
exports.ThreadHeader = require_Shell.ThreadHeader;
exports.ThreadList = require_Shell.ThreadList;
exports.UserMessageContainer = require_Shell.UserMessageContainer;
exports.WelcomeScreen = require_Shell.WelcomeScreen;
exports.createShellStore = require_ArtifactOverlay.createShellStore;
exports.useShellStore = require_ArtifactOverlay.useShellStore;
