// @openuidev/react-ui/dist/components/Shell/Sidebar.js
import { jsx as _jsx, Fragment as _Fragment, jsxs as _jsxs } from "react/jsx-runtime";
import { useActiveArtifact } from "@openuidev/react-headless";
import clsx from "clsx";
import { ArrowLeftFromLine, ArrowRightFromLine } from "lucide-react";
import { useEffect } from "react";
import { useLayoutContext } from "../../context/LayoutContext";
import { IconButton } from "../IconButton";
import { useShellStore } from "../_shared/store";
export const SidebarContainer = ({ children, className, }) => {
    const { isSidebarOpen, setIsSidebarOpen } = useShellStore((state) => ({
        isSidebarOpen: state.isSidebarOpen,
        setIsSidebarOpen: state.setIsSidebarOpen,
    }));
    const { isArtifactActive } = useActiveArtifact();
    const { layout } = useLayoutContext() || {};
    const isMobile = layout === "mobile";
    useEffect(() => {
        if (isMobile) {
            setIsSidebarOpen(false);
        }
        else {
            setIsSidebarOpen(true);
        }
    }, [isMobile]);
    return (_jsxs(_Fragment, { children: [isMobile && (_jsx("div", { className: clsx("openui-shell-sidebar-container__overlay", {
                    "openui-shell-sidebar-container__overlay--collapsed": !isSidebarOpen,
                }), onClick: () => {
                    setIsSidebarOpen(false);
                } })), _jsx("div", { className: clsx("openui-shell-sidebar-container", {
                    "openui-shell-sidebar-container--collapsed": !isSidebarOpen,
                    "openui-shell-sidebar-container--hidden": isArtifactActive && !isMobile,
                }, className), children: children })] }));
};
export const SidebarHeader = ({ children, className, }) => {
    const { agentName, logoUrl, setIsSidebarOpen, isSidebarOpen } = useShellStore((state) => ({
        agentName: state.agentName,
        logoUrl: state.logoUrl,
        setIsSidebarOpen: state.setIsSidebarOpen,
        isSidebarOpen: state.isSidebarOpen,
    }));
    return (_jsxs("div", { className: clsx("openui-shell-sidebar-header", { "openui-shell-sidebar-header--collapsed": !isSidebarOpen }, className), children: [_jsxs("div", { className: "openui-shell-sidebar-header__top-row", children: [_jsx("img", { src: logoUrl, alt: agentName, className: "openui-shell-sidebar-header__logo" }), _jsx("div", { className: "openui-shell-sidebar-header__agent-name", children: agentName }), _jsx(IconButton, { icon: isSidebarOpen ? _jsx(ArrowLeftFromLine, { size: "1em" }) : _jsx(ArrowRightFromLine, { size: "1em" }), onClick: () => {
                            setIsSidebarOpen(!isSidebarOpen);
                        }, size: "small", variant: "secondary", className: "openui-shell-sidebar-header__toggle-button" })] }), children] }));
};
export const SidebarContent = ({ children, className, }) => {
    const { isSidebarOpen } = useShellStore((state) => ({
        isSidebarOpen: state.isSidebarOpen,
    }));
    return (_jsx("div", { className: clsx("openui-shell-sidebar-content", className, {
            "openui-shell-sidebar-content--collapsed": !isSidebarOpen,
        }), children: children }));
};
export const SidebarSeparator = () => {
    return _jsx("div", { className: "openui-shell-sidebar-separator" });
};
