// @openuidev/react-ui/dist/components/Shell/Thread.js
import { jsx as _jsx, jsxs as _jsxs, Fragment as _Fragment } from "react/jsx-runtime";
import { MessageProvider, useActiveArtifact, useThread } from "@openuidev/react-headless";
import clsx from "clsx";
import { memo, useRef } from "react";
import { useLayoutContext } from "../../context/LayoutContext";
import { useScrollToBottom } from "../../hooks/useScrollToBottom";
import { separateContentAndContext } from "../../utils/contentParser";
import { ArtifactOverlay, ArtifactPortalTarget } from "../_shared/artifact";
import { useShellStore } from "../_shared/store";
import { Callout } from "../Callout";
import { MarkDownRenderer } from "../MarkDownRenderer";
import { MessageLoading as MessageLoadingComponent } from "../MessageLoading";
import { ToolCallComponent } from "../ToolCall";
import { ToolResult } from "../ToolResult";
import { ResizableSeparator } from "./ResizableSeparator";
import { useArtifactResize } from "./useArtifactResize";
export const ThreadContainer = ({ children, className, }) => {
    const { layout } = useLayoutContext();
    const isMobile = layout === "mobile";
    const { isArtifactActive } = useActiveArtifact();
    const { setIsSidebarOpen } = useShellStore((state) => ({
        setIsSidebarOpen: state.setIsSidebarOpen,
    }));
    const isLoadingMessages = useThread((s) => s.isLoadingMessages);
    const { containerRef, chatPanelRef, artifactPanelRef, isDragging, handleResize, handleDragStart, handleDragEnd, } = useArtifactResize({
        isArtifactActive,
        isMobile,
        setIsSidebarOpen,
    });
    return (_jsx("div", { className: clsx("openui-shell-thread-container", className, {
            "openui-shell-thread-container--artifact-active": isArtifactActive,
        }), style: {
            visibility: isLoadingMessages ? "hidden" : undefined,
        }, children: _jsxs("div", { className: "openui-shell-thread-wrapper", ref: containerRef, children: [_jsxs("div", { ref: chatPanelRef, className: clsx("openui-shell-thread-chat-panel", {
                        "openui-shell-thread-chat-panel--animating": !isDragging,
                    }), children: [children, isMobile && _jsx(ArtifactOverlay, {})] }), !isMobile && isArtifactActive && (_jsxs(_Fragment, { children: [_jsx(ResizableSeparator, { onResize: handleResize, onDragStart: handleDragStart, onDragEnd: handleDragEnd }), _jsx("div", { ref: artifactPanelRef, className: clsx("openui-shell-thread-artifact-panel", {
                                "openui-shell-thread-artifact-panel--animating": !isDragging,
                            }), children: _jsx(ArtifactPortalTarget, {}) })] }))] }) }));
};
export const ScrollArea = ({ children, className, scrollVariant = "user-message-anchor", userMessageSelector, }) => {
    const ref = useRef(null);
    const messages = useThread((s) => s.messages);
    const isRunning = useThread((s) => s.isRunning);
    const isLoadingMessages = useThread((s) => s.isLoadingMessages);
    useScrollToBottom({
        ref,
        lastMessage: messages[messages.length - 1] || { id: "" },
        scrollVariant,
        userMessageSelector,
        isRunning,
        isLoadingMessages,
    });
    return (_jsx("div", { className: "openui-shell-thread-scroll-container", children: _jsx("div", { ref: ref, className: clsx("openui-shell-thread-scroll-area", {
                "openui-shell-thread-scroll-area--user-message-anchor": scrollVariant === "user-message-anchor",
            }, className), children: children }) }));
};
export const AssistantMessageContainer = ({ children, className, }) => {
    const { logoUrl } = useShellStore((store) => ({
        logoUrl: store.logoUrl,
    }));
    return (_jsxs("div", { className: clsx("openui-shell-thread-message-assistant", className), children: [_jsx("img", { src: logoUrl, alt: "Assistant", className: "openui-shell-thread-message-assistant__logo" }), _jsx("div", { className: "openui-shell-thread-message-assistant__content", children: children })] }));
};
export const UserMessageContainer = ({ children, className, }) => {
    return (_jsx("div", { className: clsx("openui-shell-thread-message-user", className), children: _jsx("div", { className: "openui-shell-thread-message-user__content", children: children }) }));
};
const AssistantMessageContent = ({ message, allMessages, }) => {
    // Find tool result messages that correspond to this message's tool calls
    const getToolName = (toolCallId) => {
        const toolCall = message.toolCalls?.find((tc) => tc.id === toolCallId);
        return toolCall?.function.name;
    };
    // Collect tool messages that follow this assistant message
    const toolMessages = [];
    const msgIndex = allMessages.findIndex((m) => m.id === message.id);
    if (msgIndex !== -1) {
        for (let i = msgIndex + 1; i < allMessages.length; i++) {
            const m = allMessages[i];
            if (m && m.role === "tool") {
                toolMessages.push(m);
            }
            else {
                break;
            }
        }
    }
    return (_jsxs(_Fragment, { children: [message.content && (_jsx(MarkDownRenderer, { textMarkdown: message.content, className: "openui-shell-thread-message-assistant__text" })), message.toolCalls?.map((toolCall) => (_jsx(ToolCallComponent, { toolCall: toolCall }, toolCall.id))), toolMessages.map((tm) => (_jsx(ToolResult, { message: tm, toolName: getToolName(tm.toolCallId) }, tm.id)))] }));
};
const UserMessageContent = ({ message }) => {
    if (message.role !== "user")
        return null;
    const content = message.content;
    if (typeof content === "string") {
        // Strip XML wrapper tags (<content>, <context>) so the bubble shows clean text
        const { content: humanText } = separateContentAndContext(content);
        return _jsx(_Fragment, { children: humanText });
    }
    // InputContent[] — render text parts
    return (_jsx(_Fragment, { children: content?.map((part, i) => {
            if (part.type === "text") {
                return _jsx("span", { children: part.text }, i);
            }
            // Binary content — could be image, file, etc.
            if (part.type === "binary" && part.url) {
                return (_jsx("img", { src: part.url, alt: "", className: "openui-shell-thread-message-user__image" }, i));
            }
            return null;
        }) }));
};
export const RenderMessage = memo(({ message, className, allMessages, assistantMessage: CustomAssistantMessage, userMessage: CustomUserMessage, isStreaming, }) => {
    if (message.role === "tool") {
        // Tool messages are rendered inline with their parent assistant message
        return null;
    }
    if (message.role === "assistant") {
        if (CustomAssistantMessage) {
            return _jsx(CustomAssistantMessage, { message: message, isStreaming: isStreaming });
        }
        return (_jsx(AssistantMessageContainer, { className: className, children: _jsx(AssistantMessageContent, { message: message, allMessages: allMessages }) }));
    }
    if (message.role === "user") {
        if (CustomUserMessage) {
            return _jsx(CustomUserMessage, { message: message });
        }
        return (_jsx(UserMessageContainer, { className: className, children: _jsx(UserMessageContent, { message: message }) }));
    }
    // Other roles (system, developer, reasoning, activity) — skip by default
    return null;
});
export const MessageLoading = () => {
    return (_jsx("div", { className: "openui-shell-thread-message-loading", children: _jsx(MessageLoadingComponent, {}) }));
};
export const ThreadError = () => {
    const threadError = useThread((s) => s.threadError);
    if (!threadError)
        return null;
    return (_jsx("div", { className: "openui-shell-thread-error", children: _jsx(Callout, { variant: "danger", title: "Something went wrong", description: threadError.message || "An unexpected error occurred. Please try again." }) }));
};
export const Messages = ({ className, loader, assistantMessage, userMessage, }) => {
    const messages = useThread((s) => s.messages);
    const isRunning = useThread((s) => s.isRunning);
    const threadError = useThread((s) => s.threadError);
    return (_jsxs("div", { className: clsx("openui-shell-thread-messages", className), children: [messages.map((message, i) => {
                return (_jsx(MessageProvider, { message: message, children: _jsx(RenderMessage, { message: message, allMessages: messages, assistantMessage: assistantMessage, userMessage: userMessage, isStreaming: isRunning && i === messages.length - 1 }) }, message.id));
            }), isRunning && _jsx("div", { children: loader }), !isRunning && threadError && _jsx(ThreadError, {})] }));
};
export const ThreadHeader = ({ children, className, }) => {
    return _jsx("div", { className: clsx("openui-shell-thread-header", className), children: children });
};
// Re-export Composer from components
export { Composer } from "./components";
