// @openuidev/react-ui/dist/components/Shell/ThreadList.js
import { jsx as _jsx, jsxs as _jsxs } from "react/jsx-runtime";
import { useThreadList } from "@openuidev/react-headless";
import * as DropdownMenu from "@radix-ui/react-dropdown-menu";
import clsx from "clsx";
import { EllipsisVerticalIcon, Trash2Icon } from "lucide-react";
import { Fragment, useEffect } from "react";
import { useLayoutContext } from "../../context/LayoutContext";
import { useShellStore } from "../_shared/store";
export const ThreadButton = ({ id, title, className, }) => {
    const selectThread = useThreadList((s) => s.selectThread);
    const deleteThread = useThreadList((s) => s.deleteThread);
    const selectedThreadId = useThreadList((s) => s.selectedThreadId);
    const { isSidebarOpen, setIsSidebarOpen } = useShellStore((state) => ({
        isSidebarOpen: state.isSidebarOpen,
        setIsSidebarOpen: state.setIsSidebarOpen,
    }));
    const { layout } = useLayoutContext();
    return (_jsxs("div", { className: clsx("openui-shell-thread-button", {
            "openui-shell-thread-button--selected": selectedThreadId === id,
        }, className), children: [_jsx("button", { className: "openui-shell-thread-button-title", onClick: () => {
                    if (layout === "mobile") {
                        setIsSidebarOpen(!isSidebarOpen);
                    }
                    selectThread(id);
                }, children: title }), _jsxs(DropdownMenu.Root, { children: [_jsx(DropdownMenu.Trigger, { asChild: true, children: _jsx("button", { className: "openui-shell-thread-button-dropdown-trigger", children: _jsx(EllipsisVerticalIcon, { size: 14 }) }) }), _jsx(DropdownMenu.Portal, { children: _jsx(DropdownMenu.Content, { className: "openui-shell-thread-button-dropdown-menu", side: "bottom", align: "start", sideOffset: 2, children: _jsxs(DropdownMenu.Item, { className: "openui-shell-thread-button-dropdown-menu-item", onSelect: () => {
                                    deleteThread(id);
                                }, children: [_jsx(Trash2Icon, { size: 14, className: "openui-shell-thread-button-dropdown-menu-item-icon" }), "Delete"] }) }) })] })] }));
};
export const ThreadList = ({ className }) => {
    const threads = useThreadList((s) => s.threads);
    const loadThreads = useThreadList((s) => s.loadThreads);
    useEffect(() => {
        loadThreads();
    }, []);
    const groupThreads = () => {
        const now = new Date();
        const today = new Date(now.getFullYear(), now.getMonth(), now.getDate());
        const yesterday = new Date(today);
        yesterday.setDate(yesterday.getDate() - 1);
        const last7Days = new Date(today);
        last7Days.setDate(last7Days.getDate() - 7);
        const last30Days = new Date(today);
        last30Days.setDate(last30Days.getDate() - 30);
        const thisYear = new Date(today);
        thisYear.setMonth(0, 1);
        return threads.reduce((groups, thread) => {
            const threadDate = new Date(thread.createdAt);
            if (threadDate >= today) {
                groups.today = [...(groups.today || []), thread];
            }
            else if (threadDate >= yesterday) {
                groups.yesterday = [...(groups.yesterday || []), thread];
            }
            else if (threadDate >= last7Days) {
                groups.last7Days = [...(groups.last7Days || []), thread];
            }
            else if (threadDate >= last30Days) {
                groups.last30Days = [...(groups.last30Days || []), thread];
            }
            else if (threadDate >= thisYear) {
                groups.thisYear = [...(groups.thisYear || []), thread];
            }
            else {
                groups.older = [...(groups.older || []), thread];
            }
            return groups;
        }, {
            today: [],
            yesterday: [],
            last7Days: [],
            last30Days: [],
            thisYear: [],
            older: [],
        });
    };
    const groupedThreads = groupThreads();
    const groupLabels = {
        today: "Today",
        yesterday: "Yesterday",
        last7Days: "Previous 7 Days",
        last30Days: "Previous 30 Days",
        thisYear: "This Year",
        older: "Older",
    };
    return (_jsx("div", { className: clsx("openui-shell-thread-list", className), children: Object.entries(groupedThreads)
            .filter(([_, groupThreads]) => groupThreads.length > 0)
            .map(([group, groupThreads]) => (_jsxs(Fragment, { children: [_jsx("div", { className: "openui-shell-thread-list-group", children: groupLabels[group] }), groupThreads.map((thread) => (_jsx(ThreadButton, { id: thread.id, title: thread.title }, thread.id)))] }, group))) }));
};
