// @openuidev/react-ui/dist/components/OpenUIChat/stories/OpenUIChat.stories.js
import { jsx as _jsx, jsxs as _jsxs, Fragment as _Fragment } from "react/jsx-runtime";
import { ChevronDown, Download, Share, Sparkles, ThumbsUp, Zap } from "lucide-react";
import { useState } from "react";
import logoUrl from "../../BottomTray/stories/thesysdev_logo.jpeg";
import { Button } from "../../Button";
import { IconButton } from "../../IconButton";
import { BottomTray } from "../ComposedBottomTray.js";
import { Copilot } from "../ComposedCopilot.js";
import { FullScreen } from "../ComposedStandalone.js";
export default {
    title: "Components/OpenUIChat",
    tags: ["dev"],
};
function mockSSEResponse(text, delayMs = 500) {
    return new Promise((resolve) => {
        setTimeout(() => {
            const events = `data: ${JSON.stringify({ type: "TEXT_MESSAGE_CONTENT", delta: text })}\n\ndata: [DONE]\n\n`;
            const stream = new ReadableStream({
                start(controller) {
                    controller.enqueue(new TextEncoder().encode(events));
                    controller.close();
                },
            });
            resolve(new Response(stream));
        }, delayMs);
    });
}
const SAMPLE_WELCOME_MESSAGE = {
    title: "Hi, I'm OpenUI Assistant",
    description: "I can help you with questions about your account, products, and more.",
    image: { url: logoUrl },
};
const SAMPLE_STARTERS = {
    variant: "short",
    options: [
        {
            displayText: "Help me get started",
            prompt: "Help me get started with OpenUI",
            icon: _jsx(Sparkles, { size: 16 }),
        },
        {
            displayText: "What can you do?",
            prompt: "What can you do?",
        },
        {
            displayText: "Tell me about features",
            prompt: "Tell me about your features",
            icon: _jsx(Zap, { size: 16 }),
        },
    ],
};
const LONG_STARTERS = {
    variant: "long",
    options: [
        {
            displayText: "Help me get started with this application and guide me through the features",
            prompt: "Help me get started",
            icon: _jsx(Sparkles, { size: 16 }),
        },
        {
            displayText: "What can you do? I'd like to know all your capabilities",
            prompt: "What can you do?",
        },
        {
            displayText: "Tell me about your advanced features and how I can use them effectively",
            prompt: "Tell me about your features",
            icon: _jsx(Zap, { size: 16 }),
        },
    ],
};
const mockGenerateShareLink = async (threadId) => {
    await new Promise((r) => setTimeout(r, 1000));
    return `https://example.com/shared/${threadId}`;
};
const mockProcessMessage = async ({ messages }) => {
    const lastMsg = messages[messages.length - 1];
    const content = lastMsg?.role === "user" && typeof lastMsg.content === "string" ? lastMsg.content : "";
    return mockSSEResponse(`You said: "${content}". This is a response from the AI assistant.`);
};
const sharedProps = {
    processMessage: mockProcessMessage,
    fetchThreadList: async () => ({
        threads: [
            { id: "1", title: "Previous Chat 1", createdAt: Date.now() },
            { id: "2", title: "Previous Chat 2", createdAt: Date.now() },
            { id: "3", title: "Previous Chat 3", createdAt: Date.now() },
        ],
    }),
    createThread: async () => ({
        id: crypto.randomUUID(),
        title: "New Chat",
        createdAt: Date.now(),
    }),
    deleteThread: async () => { },
    updateThread: async (t) => t,
    loadThread: async (threadId) => {
        if (!threadId)
            return [];
        return [
            { id: crypto.randomUUID(), role: "user", content: "Hello" },
            {
                id: crypto.randomUUID(),
                role: "assistant",
                content: "Hello! How can I help you today?",
            },
        ];
    },
    logoUrl,
    agentName: "OpenUI Assistant",
};
export const Default = {
    render: () => (_jsx("div", { style: { minHeight: "100vh", background: "var(--openui-bg-fill, #f5f5f5)" }, children: _jsx(FullScreen, { ...sharedProps }) })),
};
export const StandaloneWithWelcome = {
    render: () => (_jsx("div", { style: { minHeight: "100vh", background: "var(--openui-bg-fill, #f5f5f5)" }, children: _jsx(FullScreen, { ...sharedProps, welcomeMessage: SAMPLE_WELCOME_MESSAGE, conversationStarters: SAMPLE_STARTERS }) })),
};
export const StandaloneLongStarters = {
    render: () => (_jsx("div", { style: { minHeight: "100vh", background: "var(--openui-bg-fill, #f5f5f5)" }, children: _jsx(FullScreen, { ...sharedProps, welcomeMessage: SAMPLE_WELCOME_MESSAGE, conversationStarters: LONG_STARTERS }) })),
};
export const StandaloneWithShareLink = {
    render: () => (_jsx("div", { style: { minHeight: "100vh", background: "var(--openui-bg-fill, #f5f5f5)" }, children: _jsx(FullScreen, { ...sharedProps, welcomeMessage: SAMPLE_WELCOME_MESSAGE, conversationStarters: SAMPLE_STARTERS, generateShareLink: mockGenerateShareLink }) })),
};
export const StandaloneWithThreadHeader = {
    render: () => (_jsx("div", { style: { minHeight: "100vh", background: "var(--openui-bg-fill, #f5f5f5)" }, children: _jsx(FullScreen, { ...sharedProps, welcomeMessage: SAMPLE_WELCOME_MESSAGE, conversationStarters: SAMPLE_STARTERS, threadHeader: _jsxs(_Fragment, { children: [_jsx(Button, { iconLeft: _jsx(ChevronDown, { size: 14 }), variant: "secondary", size: "small", children: "GPT-4o" }), _jsxs("div", { style: { display: "flex", gap: "4px" }, children: [_jsx(IconButton, { icon: _jsx(ThumbsUp, { size: 16 }), "aria-label": "Feedback", size: "small", variant: "secondary" }), _jsx(IconButton, { icon: _jsx(Download, { size: 16 }), "aria-label": "Export", size: "small", variant: "secondary" }), _jsx(Button, { iconLeft: _jsx(Share, { size: 14 }), variant: "secondary", size: "small", children: "Share" })] })] }) }) })),
};
export const CopilotDefault = {
    render: () => (_jsx("div", { style: { minHeight: "100vh", background: "var(--openui-bg-fill, #f5f5f5)" }, children: _jsx(Copilot, { ...sharedProps }) })),
};
export const CopilotWithWelcome = {
    render: () => (_jsx("div", { style: { minHeight: "100vh", background: "var(--openui-bg-fill, #f5f5f5)" }, children: _jsx(Copilot, { ...sharedProps, welcomeMessage: SAMPLE_WELCOME_MESSAGE, conversationStarters: SAMPLE_STARTERS }) })),
};
export const CopilotWithShareLink = {
    render: () => (_jsx("div", { style: { minHeight: "100vh", background: "var(--openui-bg-fill, #f5f5f5)" }, children: _jsx(Copilot, { ...sharedProps, welcomeMessage: SAMPLE_WELCOME_MESSAGE, conversationStarters: SAMPLE_STARTERS, generateShareLink: mockGenerateShareLink }) })),
};
export const CopilotWithHeaderActions = {
    render: () => (_jsx("div", { style: { minHeight: "100vh", background: "var(--openui-bg-fill, #f5f5f5)" }, children: _jsx(Copilot, { ...sharedProps, welcomeMessage: SAMPLE_WELCOME_MESSAGE, headerActions: _jsxs(_Fragment, { children: [_jsx(IconButton, { icon: _jsx(Download, { size: 16 }), "aria-label": "Export", size: "small", variant: "secondary" }), _jsx(IconButton, { icon: _jsx(Share, { size: 16 }), "aria-label": "Share", size: "small", variant: "secondary" })] }) }) })),
};
const BottomTrayStory = ({ defaultOpen = false, ...rest }) => {
    const [isOpen, setIsOpen] = useState(defaultOpen);
    return (_jsxs("div", { style: { minHeight: "100vh", background: "var(--openui-bg-fill, #f5f5f5)" }, children: [_jsxs("div", { style: { padding: "2rem" }, children: [_jsx("h1", { style: { marginBottom: "1rem" }, children: "Bottom Tray Example" }), _jsx("p", { style: { marginBottom: "1rem" }, children: "The chat appears as a bottom tray. Click the pill button to open/close." }), _jsxs("button", { onClick: () => setIsOpen(!isOpen), style: {
                            padding: "0.5rem 1rem",
                            background: "var(--openui-interactive-default)",
                            border: "1px solid var(--openui-stroke-interactive-el)",
                            borderRadius: "8px",
                            cursor: "pointer",
                        }, children: [isOpen ? "Close" : "Open", " Chat"] })] }), _jsx(BottomTray, { ...sharedProps, ...rest, isOpen: isOpen, onOpenChange: setIsOpen })] }));
};
export const BottomTrayDefault = {
    render: () => _jsx(BottomTrayStory, {}),
};
export const BottomTrayOpen = {
    render: () => _jsx(BottomTrayStory, { defaultOpen: true }),
};
export const BottomTrayWithWelcome = {
    render: () => (_jsx(BottomTrayStory, { defaultOpen: true, welcomeMessage: SAMPLE_WELCOME_MESSAGE, conversationStarters: SAMPLE_STARTERS })),
};
export const BottomTrayLongStarters = {
    render: () => (_jsx(BottomTrayStory, { defaultOpen: true, welcomeMessage: SAMPLE_WELCOME_MESSAGE, conversationStarters: LONG_STARTERS })),
};
export const BottomTrayWithShareLink = {
    render: () => (_jsx(BottomTrayStory, { defaultOpen: true, welcomeMessage: SAMPLE_WELCOME_MESSAGE, conversationStarters: SAMPLE_STARTERS, generateShareLink: mockGenerateShareLink })),
};
export const BottomTrayWithHeaderActions = {
    render: () => (_jsx(BottomTrayStory, { defaultOpen: true, welcomeMessage: SAMPLE_WELCOME_MESSAGE, conversationStarters: SAMPLE_STARTERS, headerActions: _jsx(IconButton, { icon: _jsx(Share, { size: 16 }), "aria-label": "Share", size: "small", variant: "tertiary" }) })),
};
