// @openuidev/react-ui/dist/components/Calendar/stories/Calendar.stories.js
import { jsx as _jsx } from "react/jsx-runtime";
import { Calendar } from "../Calendar";
const meta = {
    title: "Components/Calendar",
    component: Calendar,
    parameters: {
        layout: "centered",
        docs: {
            description: {
                component: `
A flexible \`calendar\` component built on top of [react-day-picker](https://react-day-picker.js.org/).

\`\`\`tsx
import { Calendar } from '@openui-ui/react-ui';
\`\`\`
`,
            },
        },
    },
    decorators: [
        (Story) => (_jsx("div", { style: {
                width: "350px",
            }, children: _jsx(Story, {}) })),
    ],
    argTypes: {
        mode: {
            description: "Selection mode for the calendar",
            control: {
                type: "radio",
            },
            options: ["single", "multiple", "range"],
            defaultValue: "single",
            table: {
                defaultValue: { summary: "single" },
            },
        },
    },
    tags: ["autodocs", "!dev"],
};
export default meta;
export const SingleDate = {
    args: {
        mode: "single",
    },
};
export const MultipleDate = {
    args: {
        mode: "multiple",
        selected: [new Date(), new Date(Date.now() + 86400000)], // Today and tomorrow
    },
};
export const DateRange = {
    args: {
        mode: "range",
        selected: {
            from: new Date(),
            to: new Date(Date.now() + 86400000 * 7), // Today to 7 days later
        },
    },
};
export const Disabled = {
    args: {
        mode: "single",
        disabled: true,
    },
};
export const WithDisabledDays = {
    args: {
        mode: "single",
        disabled: [
            { from: new Date(), to: new Date(Date.now() + 86400000 * 3) }, // Disable next 3 days
        ],
    },
};
