// @openuidev/react-ui/dist/components/TagBlock/stories/TagBlock.stories.js
import { jsx as _jsx, jsxs as _jsxs } from "react/jsx-runtime";
import { Heart, Info, Star, User } from "lucide-react";
import { Tag } from "../../Tag";
import { TagBlock } from "../TagBlock";
const meta = {
    title: "Components/TagBlock",
    component: TagBlock,
    subcomponents: {
        Tag: Tag,
    },
    parameters: {
        layout: "centered",
        docs: {
            description: {
                component: "```tsx\nimport { TagBlock, Tag } from '@openui-ui/react-ui';\n```",
            },
        },
    },
    argTypes: {
        children: {
            control: false,
            description: "Accepts Tag components as children",
            table: {
                category: "Content",
                type: { summary: "ReactElement<typeof Tag> | ReactElement<typeof Tag>[]" },
            },
        },
        className: {
            control: false,
            description: "Additional CSS class name for custom styling",
            table: {
                category: "Styling",
                type: { summary: "string" },
            },
        },
        styles: {
            control: false,
            description: "Additional CSS styles for custom styling",
            table: {
                category: "Styling",
                type: { summary: "CSSProperties" },
            },
        },
    },
    tags: ["!dev", "autodocs"],
};
export default meta;
export const Default = {
    render: (args) => (_jsxs(TagBlock, { ...args, children: [_jsx(Tag, { variant: "info", icon: _jsx(Info, { size: 16 }), size: "sm", text: "Info" }), _jsx(Tag, { variant: "success", icon: _jsx(Star, { size: 16 }), size: "sm", text: "Star" }), _jsx(Tag, { variant: "info", icon: _jsx(Info, { size: 16 }), size: "md", text: "Info" }), _jsx(Tag, { variant: "success", icon: _jsx(Star, { size: 16 }), size: "md", text: "Star" }), _jsx(Tag, { variant: "warning", icon: _jsx(Heart, { size: 16 }), size: "lg", text: "Heart" }), _jsx(Tag, { variant: "danger", icon: _jsx(User, { size: 16 }), size: "lg", text: "User" })] })),
};
