// @openuidev/react-ui/dist/components/BottomTray/Thread.js
import { jsx as _jsx, jsxs as _jsxs, Fragment as _Fragment } from "react/jsx-runtime";
import { MessageProvider, useThread } from "@openuidev/react-headless";
import clsx from "clsx";
import { memo, useRef } from "react";
import { useScrollToBottom } from "../../hooks/useScrollToBottom";
import { ArtifactOverlay } from "../_shared/artifact";
import { MarkDownRenderer } from "../MarkDownRenderer";
import { MessageLoading as MessageLoadingComponent } from "../MessageLoading";
import { ToolCallComponent } from "../ToolCall";
import { ToolResult } from "../ToolResult";
export const ThreadContainer = ({ children, className, }) => {
    const isLoadingMessages = useThread((s) => s.isLoadingMessages);
    return (_jsxs("div", { className: clsx("openui-bottom-tray-thread-container", className), style: {
            visibility: isLoadingMessages ? "hidden" : undefined,
        }, children: [children, _jsx(ArtifactOverlay, {})] }));
};
export const ScrollArea = ({ children, className, scrollVariant = "user-message-anchor", userMessageSelector = ".openui-bottom-tray-thread-message-user", }) => {
    const ref = useRef(null);
    const messages = useThread((s) => s.messages);
    const isRunning = useThread((s) => s.isRunning);
    const isLoadingMessages = useThread((s) => s.isLoadingMessages);
    useScrollToBottom({
        ref,
        lastMessage: messages[messages.length - 1] || { id: "" },
        scrollVariant,
        userMessageSelector,
        isRunning,
        isLoadingMessages,
    });
    return (_jsxs("div", { className: "openui-bottom-tray-thread-scroll-container", children: [_jsx("div", { ref: ref, className: clsx("openui-bottom-tray-thread-scroll-area", {
                    "openui-bottom-tray-thread-scroll-area--user-message-anchor": scrollVariant === "user-message-anchor",
                }, className), children: children }), _jsx("div", { className: "openui-bottom-tray-thread-scroll-gradient" })] }));
};
export const AssistantMessageContainer = ({ children, className, }) => {
    return (_jsx("div", { className: clsx("openui-bottom-tray-thread-message-assistant", className), children: _jsx("div", { className: "openui-bottom-tray-thread-message-assistant__content", children: children }) }));
};
export const UserMessageContainer = ({ children, className, }) => {
    return (_jsx("div", { className: clsx("openui-bottom-tray-thread-message-user", className), children: _jsx("div", { className: "openui-bottom-tray-thread-message-user__content", children: children }) }));
};
const AssistantMessageContent = ({ message, allMessages, }) => {
    const getToolName = (toolCallId) => {
        const toolCall = message.toolCalls?.find((tc) => tc.id === toolCallId);
        return toolCall?.function.name;
    };
    const toolMessages = [];
    const msgIndex = allMessages.findIndex((m) => m.id === message.id);
    if (msgIndex !== -1) {
        for (let i = msgIndex + 1; i < allMessages.length; i++) {
            const m = allMessages[i];
            if (m && m.role === "tool") {
                toolMessages.push(m);
            }
            else {
                break;
            }
        }
    }
    return (_jsxs(_Fragment, { children: [message.content && (_jsx(MarkDownRenderer, { textMarkdown: message.content, className: "openui-bottom-tray-thread-message-assistant__text" })), message.toolCalls?.map((toolCall) => (_jsx(ToolCallComponent, { toolCall: toolCall }, toolCall.id))), toolMessages.map((tm) => (_jsx(ToolResult, { message: tm, toolName: getToolName(tm.toolCallId) }, tm.id)))] }));
};
const UserMessageContent = ({ message }) => {
    if (message.role !== "user")
        return null;
    const content = message.content;
    if (typeof content === "string") {
        return _jsx(_Fragment, { children: content });
    }
    return (_jsx(_Fragment, { children: content?.map((part, i) => {
            if (part.type === "text") {
                return _jsx("span", { children: part.text }, i);
            }
            if (part.type === "binary" && part.url) {
                return (_jsx("img", { src: part.url, alt: "", className: "openui-bottom-tray-thread-message-user__image" }, i));
            }
            return null;
        }) }));
};
export const RenderMessage = memo(({ message, className, allMessages, assistantMessage: CustomAssistantMessage, userMessage: CustomUserMessage, isStreaming, }) => {
    if (message.role === "tool") {
        return null;
    }
    if (message.role === "assistant") {
        if (CustomAssistantMessage) {
            return _jsx(CustomAssistantMessage, { message: message, isStreaming: isStreaming });
        }
        return (_jsx(AssistantMessageContainer, { className: className, children: _jsx(AssistantMessageContent, { message: message, allMessages: allMessages }) }));
    }
    if (message.role === "user") {
        if (CustomUserMessage) {
            return _jsx(CustomUserMessage, { message: message });
        }
        return (_jsx(UserMessageContainer, { className: className, children: _jsx(UserMessageContent, { message: message }) }));
    }
    return null;
});
export const MessageLoading = () => {
    return (_jsx("div", { className: "openui-bottom-tray-thread-message-loading", children: _jsx(MessageLoadingComponent, {}) }));
};
export const Messages = ({ className, loader, assistantMessage, userMessage, }) => {
    const messages = useThread((s) => s.messages);
    const isRunning = useThread((s) => s.isRunning);
    return (_jsxs("div", { className: clsx("openui-bottom-tray-thread-messages", className), children: [messages.map((message, i) => {
                return (_jsx(MessageProvider, { message: message, children: _jsx(RenderMessage, { message: message, allMessages: messages, assistantMessage: assistantMessage, userMessage: userMessage, isStreaming: isRunning && i === messages.length - 1 }) }, message.id));
            }), isRunning && _jsx("div", { children: loader })] }));
};
// Re-export Composer from components
export { Composer } from "./components";
