// @openuidev/react-ui/dist/components/_shared/store/store.js
import { jsx as _jsx } from "react/jsx-runtime";
import { createContext, useContext, useEffect, useMemo } from "react";
import { create } from "zustand";
import { useShallow } from "zustand/react/shallow";
export const createShellStore = ({ logoUrl, agentName }) => create((set) => ({
    isSidebarOpen: true,
    agentName: agentName,
    logoUrl: logoUrl,
    setIsSidebarOpen: (isOpen) => set({ isSidebarOpen: isOpen }),
    setAgentName: (name) => set({ agentName: name }),
    setLogoUrl: (url) => set({ logoUrl: url }),
}));
export const ShellStoreContext = createContext(null);
export const useShellStore = (selector) => {
    const store = useContext(ShellStoreContext);
    if (!store) {
        throw new Error("useShellStore must be used within ShellStoreProvider");
    }
    return store(useShallow(selector));
};
export const ShellStoreProvider = ({ children, agentName, logoUrl, }) => {
    const shellStore = useMemo(() => createShellStore({ agentName, logoUrl }), []);
    useEffect(() => {
        const { setAgentName, setLogoUrl } = shellStore.getState();
        setAgentName(agentName);
        setLogoUrl(logoUrl);
    }, [agentName, logoUrl]);
    return _jsx(ShellStoreContext.Provider, { value: shellStore, children: children });
};
