// @openuidev/react-ui/dist/genui-lib/SwitchGroup/index.js
"use client";
import { jsx as _jsx } from "react/jsx-runtime";
import { defineComponent, useIsStreaming, useStateField, } from "@openuidev/react-lang";
import React from "react";
import { SwitchGroup as OpenUISwitchGroup } from "../../components/SwitchGroup";
import { SwitchItem as OpenUISwitchItem } from "../../components/SwitchItem";
import { SwitchItemSchema, createSwitchGroupSchema } from "./schema";
export { SwitchItemSchema } from "./schema";
export const SwitchItem = defineComponent({
    name: "SwitchItem",
    props: SwitchItemSchema,
    description: "Individual switch toggle",
    component: () => null,
});
export const SwitchGroup = defineComponent({
    name: "SwitchGroup",
    props: createSwitchGroupSchema(SwitchItem),
    description: "Group of switch toggles",
    component: ({ props }) => {
        const isStreaming = useIsStreaming();
        const field = useStateField(props.name, props.value);
        const items = (props.items ?? []);
        // Aggregate: map of item name → checked boolean
        const getAggregate = React.useCallback(() => {
            const stored = field.value;
            const result = {};
            for (const item of items) {
                result[item.props.name] = stored?.[item.props.name] ?? item.props.defaultChecked ?? false;
            }
            return result;
        }, [field.value, items]);
        if (!items.length)
            return null;
        const aggregate = getAggregate();
        return (_jsx(OpenUISwitchGroup, { variant: props.variant || "clear", children: items.map((item, i) => (_jsx(OpenUISwitchItem, { name: item.props.name, label: item.props.label, description: item.props.description || "", checked: aggregate[item.props.name] ?? item.props.defaultChecked ?? false, onChange: (val) => {
                    const newAggregate = { ...getAggregate(), [item.props.name]: val };
                    field.setValue(newAggregate);
                }, disabled: isStreaming }, i))) }));
    },
});
