// @openuidev/react-ui/dist/components/DatePicker/helpers/components/HelperComponents.js
import { jsx as _jsx, jsxs as _jsxs } from "react/jsx-runtime";
import * as SelectPrimitive from "@radix-ui/react-select";
import clsx from "clsx";
import debounce from "lodash-es/debounce";
import { ChevronLeft, ChevronRight } from "lucide-react";
import { forwardRef, useEffect, useState, } from "react";
import { IconButton } from "../../../IconButton";
import { Select, SelectItem, SelectTrigger, SelectValue } from "../../../Select";
import { useTheme } from "../../../ThemeProvider";
import { getMonthName, getMonthNumber } from "../utils/helperFn";
export const PreviousMonthButton = (props) => {
    const { onClick, disabled, className, ...rest } = props;
    return (_jsx(IconButton, { icon: _jsx(ChevronLeft, { size: 18 }), variant: "secondary", size: "medium", onClick: onClick, disabled: disabled, className: className, ...rest }));
};
export const NextMonthButton = (props) => {
    const { onClick, disabled, className, ...rest } = props;
    return (_jsx(IconButton, { icon: _jsx(ChevronRight, { size: 18 }), variant: "secondary", size: "medium", onClick: onClick, disabled: disabled, className: className, ...rest }));
};
// this component is overriding the default SelectContent component to allow for a container prop we do not wish to give to the user
const SelectContent = forwardRef(({ className, children, position = "popper", viewportClassName, container, ...props }, ref) => {
    const { portalThemeClassName } = useTheme();
    return (_jsx(SelectPrimitive.Portal, { container: container || document.body, children: _jsx(SelectPrimitive.Content, { ref: ref, className: clsx("openui-select-content", portalThemeClassName, className), position: position, ...props, children: _jsx(SelectPrimitive.Viewport, { className: clsx("openui-select-viewport", viewportClassName), "data-position": position, children: children }) }) }));
});
export const MonthsDropdown = (props) => {
    const { className, disabled, onChange, options, value, key, "aria-label": ariaLabel, container, } = props;
    const [containerWidth, setContainerWidth] = useState(0);
    const [containerHeight, setContainerHeight] = useState(0);
    useEffect(() => {
        if (!container)
            return;
        const targetElement = container.children[0]?.children[0];
        if (!targetElement)
            return;
        const resizeObserver = new ResizeObserver(debounce((entries) => {
            const { width, height } = entries[0]?.contentRect || {};
            setContainerWidth(width || 0);
            setContainerHeight(height || 0);
        }, 100));
        resizeObserver.observe(targetElement);
        return () => {
            resizeObserver.disconnect();
        };
    }, [container]);
    return (_jsxs(Select, { onValueChange: (value) => onChange?.({
            target: { value: getMonthNumber(value) },
        }), value: getMonthName(Number(value)), disabled: disabled, "aria-label": ariaLabel, children: [_jsx(SelectTrigger, { className: clsx("openui-date-picker-select-trigger", className), children: _jsx(SelectValue, { placeholder: "Select a month" }) }), _jsx(SelectContent, { container: container, className: "openui-date-picker-select-content-months", sideOffset: 4, alignOffset: 0, style: {
                    maxHeight: `${containerHeight - 45}px`,
                    minWidth: `${containerWidth}px`,
                }, children: options?.map((option) => (_jsx(SelectItem, { value: getMonthName(option.value), disabled: option.disabled, children: option.label }, option.value))) })] }, key));
};
export const YearsDropdown = (props) => {
    const { className, disabled, onChange, options, value, key, "aria-label": ariaLabel, container, botType, } = props;
    const [containerHeight, setContainerHeight] = useState(0);
    useEffect(() => {
        if (!container)
            return;
        const targetElement = container.children[0]?.children[0];
        if (!targetElement)
            return;
        const resizeObserver = new ResizeObserver(debounce((entries) => {
            const { height } = entries[0]?.contentRect || {};
            setContainerHeight(height || 0);
        }, 100));
        resizeObserver.observe(targetElement);
        return () => {
            resizeObserver.disconnect();
        };
    }, [container]);
    return (_jsxs(Select, { onValueChange: (value) => onChange?.({
            target: { value: Number(value) },
        }), value: String(value), disabled: disabled, "aria-label": ariaLabel, children: [_jsx(SelectTrigger, { className: clsx("openui-date-picker-select-trigger", className), children: _jsx(SelectValue, { placeholder: "Select a month" }) }), _jsx(SelectContent, { container: container, className: "openui-date-picker-select-content-years", viewportClassName: clsx("openui-date-picker-select-viewport", botType === "mobile" && "openui-date-picker-select-viewport-mobile"), sideOffset: 4, alignOffset: -135, style: {
                    minHeight: `${containerHeight - 45}px`,
                    maxHeight: `${containerHeight - 45}px`,
                }, children: options?.map((option) => (_jsx(SelectItem, { value: String(option.value), disabled: option.disabled, showTick: false, className: "openui-date-picker-select-item", children: option.label }, option.value))) })] }, key));
};
