// @openuidev/react-ui/dist/genui-lib/Carousel/index.js
"use client";
import { jsx as _jsx, jsxs as _jsxs } from "react/jsx-runtime";
import { defineComponent } from "@openuidev/react-lang";
import { ChevronLeft, ChevronRight } from "lucide-react";
import { z } from "zod/v4";
import { Carousel as OpenUICarousel, CarouselContent as OpenUICarouselContent, CarouselItem as OpenUICarouselItem, CarouselNext as OpenUICarouselNext, CarouselPrevious as OpenUICarouselPrevious, } from "../../components/Carousel";
import { ContentChildUnion } from "../unions";
export const Carousel = defineComponent({
    name: "Carousel",
    props: z.object({
        children: z.array(z.array(ContentChildUnion)),
        variant: z.enum(["card", "sunk"]).optional(),
    }),
    description: "Horizontal scrollable carousel",
    component: ({ props, renderNode }) => {
        const items = props.children ?? [];
        return (_jsxs(OpenUICarousel, { showButtons: true, variant: props.variant, children: [_jsx(OpenUICarouselContent, { children: items.map((item, i) => (_jsx(OpenUICarouselItem, { children: renderNode(item) }, i))) }), _jsx(OpenUICarouselPrevious, { icon: _jsx(ChevronLeft, {}) }), _jsx(OpenUICarouselNext, { icon: _jsx(ChevronRight, {}) })] }));
    },
});
