// @openuidev/react-ui/dist/components/_shared/artifact/ArtifactPanel.js
import { jsx as _jsx, jsxs as _jsxs } from "react/jsx-runtime";
import { useArtifact, useArtifactPortalTarget } from "@openuidev/react-headless";
import clsx from "clsx";
import { X } from "lucide-react";
import { Component, forwardRef, useEffect } from "react";
import { createPortal } from "react-dom";
import { useTheme } from "../../ThemeProvider/ThemeProvider";
/** @internal */
class ArtifactErrorBoundary extends Component {
    constructor(props) {
        super(props);
        this.state = { hasError: false };
    }
    static getDerivedStateFromError() {
        return { hasError: true };
    }
    render() {
        if (this.state.hasError) {
            return this.props.fallback ?? null;
        }
        return this.props.children;
    }
}
/** @internal */
const DefaultHeader = ({ title, onClose }) => (_jsxs("div", { className: "openui-artifact-panel__header", children: [_jsx("span", { className: "openui-artifact-panel__title", children: title }), _jsx("button", { className: "openui-artifact-panel__close", onClick: onClose, "aria-label": "Close artifact panel", children: _jsx(X, { size: 16 }) })] }));
/**
 * Portals artifact content into the nearest {@link ArtifactPortalTarget}.
 *
 * Renders nothing when the artifact is inactive or no portal target is mounted.
 * Wraps children in an error boundary and applies theme-scoped class names.
 *
 * Requires `<ArtifactPortalTarget />` to be mounted in the layout.
 *
 * @category Components
 */
export const ArtifactPanel = forwardRef(({ artifactId, children, title, className, errorFallback, header = true }, ref) => {
    const { isActive, close } = useArtifact(artifactId);
    const { node: panelNode } = useArtifactPortalTarget();
    const { portalThemeClassName } = useTheme();
    useEffect(() => {
        if (!isActive || panelNode)
            return;
        const timer = setTimeout(() => {
            console.warn("[OpenUI] ArtifactPanel: artifact is active but no render target is mounted. " +
                "Ensure <ArtifactPortalTarget /> is rendered in your layout.");
        }, 100);
        return () => clearTimeout(timer);
    }, [isActive, panelNode]);
    if (!isActive || !panelNode)
        return null;
    const handleClose = () => close();
    let headerContent = null;
    if (header === true) {
        headerContent = _jsx(DefaultHeader, { title: title ?? "Artifact", onClose: handleClose });
    }
    else if (header !== false) {
        headerContent = header;
    }
    return createPortal(_jsxs("div", { ref: ref, id: `openui-artifact-panel-${artifactId}`, className: clsx("openui-artifact-panel", portalThemeClassName, className), role: "region", "aria-label": title ?? "Artifact panel", children: [headerContent, _jsx(ArtifactErrorBoundary, { fallback: errorFallback, children: children })] }), panelNode);
});
ArtifactPanel.displayName = "ArtifactPanel";
