// @openuidev/react-ui/dist/genui-lib/SectionBlock/index.js
"use client";
import { jsx as _jsx, Fragment as _Fragment, jsxs as _jsxs } from "react/jsx-runtime";
import { defineComponent, useIsStreaming } from "@openuidev/react-lang";
import React from "react";
import { z } from "zod/v4";
import { FoldableSectionContent, FoldableSectionItem, FoldableSectionRoot, FoldableSectionTrigger, } from "../../components/SectionBlock/FoldableSection";
import { SectionV2 } from "../../components/SectionBlock/SectionV2";
import { SectionItem } from "../SectionItem";
export const SectionBlock = defineComponent({
    name: "SectionBlock",
    props: z.object({
        sections: z.array(SectionItem.ref),
        isFoldable: z.boolean().optional(),
    }),
    description: "Collapsible accordion sections. Auto-opens sections as they stream in. Use SectionItem for each section.",
    component: ({ props, renderNode }) => {
        const items = (props.sections ?? []);
        const isFoldable = props.isFoldable !== false;
        const isStreaming = useIsStreaming();
        const firstItemValue = items[0]?.props?.value;
        const [openItems, setOpenItems] = React.useState([]);
        const userSelected = React.useRef(false);
        const prevLengthRef = React.useRef(0);
        const prevIsStreaming = React.useRef(isStreaming);
        // During streaming: auto-open each new section as it arrives
        React.useEffect(() => {
            if (items.length === 0)
                return;
            if (isStreaming && items.length > prevLengthRef.current && !userSelected.current) {
                const last = items[items.length - 1];
                const lastValue = last?.props?.value;
                if (lastValue) {
                    setOpenItems((prev) => (prev.includes(lastValue) ? prev : [...prev, lastValue]));
                }
            }
            else {
                setOpenItems((prev) => (prev.length === 0 && firstItemValue ? [firstItemValue] : prev));
            }
            prevLengthRef.current = items.length;
        }, [items.length, isStreaming, firstItemValue]);
        // When streaming ends: collapse back to first item (unless user interacted)
        React.useEffect(() => {
            if (prevIsStreaming.current && !isStreaming && !userSelected.current && items.length > 0) {
                setOpenItems(firstItemValue ? [firstItemValue] : []);
            }
            prevIsStreaming.current = isStreaming;
        }, [isStreaming]);
        const handleValueChange = React.useCallback((value) => {
            userSelected.current = true;
            setOpenItems(value ?? []);
        }, []);
        if (!isFoldable) {
            return (_jsx(_Fragment, { children: items.map((item, index) => (_jsx(SectionV2, { trigger: String(item?.props?.trigger ?? ""), children: renderNode(item?.props?.content) }, index))) }));
        }
        return (_jsx(FoldableSectionRoot, { type: "multiple", value: openItems, onValueChange: handleValueChange, children: items.map((item, index) => (_jsxs(FoldableSectionItem, { value: String(item?.props?.value ?? index), children: [_jsx(FoldableSectionTrigger, { text: String(item?.props?.trigger ?? "") }), _jsx(FoldableSectionContent, { children: renderNode(item?.props?.content) })] }, index))) }));
    },
});
