// @openuidev/react-ui/dist/components/Separator/stories/Separator.stories.js
import { jsx as _jsx } from "react/jsx-runtime";
import { Separator } from "../Separator";
const meta = {
    title: "Components/Separator",
    component: Separator,
    tags: ["autodocs", "!dev"],
    parameters: {
        layout: "centered",
        docs: {
            description: {
                component: "```tsx\nimport { Separator } from '@openui-ui/react-ui';\n```",
            },
        },
    },
    decorators: [
        (Story, { args }) => (_jsx("div", { style: args.orientation === "vertical" ? { height: "100px" } : { width: "100px" }, children: _jsx(Story, {}) })),
    ],
    argTypes: {
        orientation: {
            control: "radio",
            options: ["horizontal", "vertical"],
            description: "The orientation of the separator.",
            defaultValue: "horizontal",
        },
        decorative: {
            control: "boolean",
            description: "When ```true```, signifies that it is purely visual, carries no semantic meaning, and ensures it is not present in the accessibility tree.",
            defaultValue: false,
        },
    },
};
export default meta;
export const Horizontal = {
    args: {
        orientation: "horizontal",
    },
};
export const Vertical = {
    args: {
        orientation: "vertical",
    },
};
