// @openuidev/react-ui/dist/components/Table/index.cjs
Object.defineProperty(exports, Symbol.toStringTag, { value: "Module" });
const require_chunk = require("../../chunk-CKQMccvm.cjs");
const require_components_IconButton_index = require("../IconButton/index.cjs");
let clsx = require("clsx");
clsx = require_chunk.__toESM(clsx);
let lucide_react = require("lucide-react");
let react = require("react");
react = require_chunk.__toESM(react);
let react_jsx_runtime = require("react/jsx-runtime");
//#region src/components/Table/Table.tsx
const alignmentClasses = {
	left: "openui-table-align-left",
	center: "openui-table-align-center",
	right: "openui-table-align-right"
};
const Table = react.default.forwardRef(({ className, containerRef, containerClassName, containerStyle, ...props }, ref) => /* @__PURE__ */ (0, react_jsx_runtime.jsx)("div", {
	ref: containerRef,
	className: (0, clsx.default)("openui-table-container", containerClassName),
	style: containerStyle,
	children: /* @__PURE__ */ (0, react_jsx_runtime.jsx)("table", {
		ref,
		className: (0, clsx.default)("openui-table", className),
		...props
	})
}));
Table.displayName = "Table";
const TableHeader = react.default.forwardRef(({ className, ...props }, ref) => /* @__PURE__ */ (0, react_jsx_runtime.jsx)("thead", {
	ref,
	className: (0, clsx.default)("openui-table-header", className),
	...props
}));
TableHeader.displayName = "TableHeader";
const TableBody = react.default.forwardRef(({ className, ...props }, ref) => /* @__PURE__ */ (0, react_jsx_runtime.jsx)("tbody", {
	ref,
	className: (0, clsx.default)("openui-table-body", className),
	...props
}));
TableBody.displayName = "TableBody";
const TableFooter = react.default.forwardRef(({ className, ...props }, ref) => /* @__PURE__ */ (0, react_jsx_runtime.jsx)("tfoot", {
	ref,
	className: (0, clsx.default)("openui-table-footer", className),
	...props
}));
TableFooter.displayName = "TableFooter";
const TableRow = react.default.forwardRef(({ className, ...props }, ref) => /* @__PURE__ */ (0, react_jsx_runtime.jsx)("tr", {
	ref,
	className: (0, clsx.default)("openui-table-row", className),
	...props
}));
TableRow.displayName = "TableRow";
const TableHead = react.default.forwardRef(({ className, children, icon, align, ...props }, ref) => /* @__PURE__ */ (0, react_jsx_runtime.jsx)("th", {
	ref,
	className: (0, clsx.default)("openui-table-head", className),
	...props,
	children: /* @__PURE__ */ (0, react_jsx_runtime.jsxs)("div", {
		className: (0, clsx.default)("openui-table-head-content", align && alignmentClasses[align]),
		children: [icon && /* @__PURE__ */ (0, react_jsx_runtime.jsx)("div", {
			className: "openui-table-head-icon",
			children: icon
		}), /* @__PURE__ */ (0, react_jsx_runtime.jsx)("div", {
			className: "openui-table-head-label",
			children
		})]
	})
}));
TableHead.displayName = "TableHead";
const TableCell = react.default.forwardRef(({ className, align, ...props }, ref) => /* @__PURE__ */ (0, react_jsx_runtime.jsx)("td", {
	ref,
	className: (0, clsx.default)("openui-table-cell", align && alignmentClasses[align], className),
	...props
}));
TableCell.displayName = "TableCell";
const ScrollableTable = react.default.forwardRef(({ className, containerClassName, containerStyle, children, ...tableProps }, ref) => {
	const wrapperRef = (0, react.useRef)(null);
	const scrollContainerRef = (0, react.useRef)(null);
	const [isScrollable, setIsScrollable] = (0, react.useState)(false);
	const [canScrollLeft, setCanScrollLeft] = (0, react.useState)(false);
	const [canScrollRight, setCanScrollRight] = (0, react.useState)(false);
	const columnLeftsRef = (0, react.useRef)([]);
	(0, react.useEffect)(() => {
		const container = scrollContainerRef.current;
		const wrapper = wrapperRef.current;
		if (!container || !wrapper) return;
		const compute = () => {
			setIsScrollable(container.scrollWidth > container.clientWidth);
			setCanScrollLeft(container.scrollLeft > 0);
			setCanScrollRight(container.scrollLeft < container.scrollWidth - container.clientWidth - 1);
			const ths = wrapper.querySelectorAll("thead th");
			if (!ths || ths.length === 0) {
				columnLeftsRef.current = [];
				return;
			}
			const containerRect = container.getBoundingClientRect();
			const lefts = [];
			ths.forEach((th) => {
				const left = th.getBoundingClientRect().left - containerRect.left + container.scrollLeft;
				lefts.push(Math.max(0, Math.round(left)));
			});
			columnLeftsRef.current = Array.from(new Set(lefts)).sort((a, b) => a - b);
		};
		compute();
		const ro = new ResizeObserver(() => compute());
		ro.observe(container);
		const onScroll = () => {
			setCanScrollLeft(container.scrollLeft > 0);
			setCanScrollRight(container.scrollLeft < container.scrollWidth - container.clientWidth - 1);
		};
		container.addEventListener("scroll", onScroll);
		return () => {
			ro.disconnect();
			container.removeEventListener("scroll", onScroll);
		};
	}, [children]);
	const scrollToNextColumn = (0, react.useCallback)(() => {
		const container = scrollContainerRef.current;
		if (!container || columnLeftsRef.current.length === 0) return;
		const current = container.scrollLeft;
		const target = columnLeftsRef.current.find((l) => l > current + 1);
		const maxScroll = container.scrollWidth - container.clientWidth;
		const next = typeof target === "number" ? target : maxScroll;
		container.scrollTo({
			left: Math.min(next, maxScroll),
			behavior: "smooth"
		});
	}, []);
	const scrollToPrevColumn = (0, react.useCallback)(() => {
		const container = scrollContainerRef.current;
		if (!container || columnLeftsRef.current.length === 0) return;
		const current = container.scrollLeft;
		const prev = columnLeftsRef.current.filter((l) => l < current - 1).at(-1) ?? 0;
		container.scrollTo({
			left: Math.max(0, prev),
			behavior: "smooth"
		});
	}, []);
	return /* @__PURE__ */ (0, react_jsx_runtime.jsxs)("div", {
		className: "openui-scrollable-table-wrapper",
		ref: wrapperRef,
		tabIndex: 0,
		children: [isScrollable && /* @__PURE__ */ (0, react_jsx_runtime.jsxs)(react_jsx_runtime.Fragment, { children: [/* @__PURE__ */ (0, react_jsx_runtime.jsx)("div", {
			className: (0, clsx.default)("openui-scrollable-table-control", "openui-scrollable-table-control-left", !canScrollLeft && "openui-scrollable-table-control-disabled"),
			children: /* @__PURE__ */ (0, react_jsx_runtime.jsx)(require_components_IconButton_index.IconButton, {
				"aria-label": "Scroll left",
				size: "small",
				variant: "secondary",
				onClick: scrollToPrevColumn,
				disabled: !canScrollLeft,
				icon: /* @__PURE__ */ (0, react_jsx_runtime.jsx)(lucide_react.ChevronLeft, { size: 16 })
			})
		}), /* @__PURE__ */ (0, react_jsx_runtime.jsx)("div", {
			className: (0, clsx.default)("openui-scrollable-table-control", "openui-scrollable-table-control-right", !canScrollRight && "openui-scrollable-table-control-disabled"),
			children: /* @__PURE__ */ (0, react_jsx_runtime.jsx)(require_components_IconButton_index.IconButton, {
				"aria-label": "Scroll right",
				size: "small",
				variant: "secondary",
				onClick: scrollToNextColumn,
				disabled: !canScrollRight,
				icon: /* @__PURE__ */ (0, react_jsx_runtime.jsx)(lucide_react.ChevronRight, { size: 16 })
			})
		})] }), /* @__PURE__ */ (0, react_jsx_runtime.jsx)(Table, {
			ref,
			containerRef: scrollContainerRef,
			containerClassName: (0, clsx.default)("openui-scrollable-table-scroll-container", containerClassName),
			containerStyle,
			className,
			style: { width: isScrollable ? "max-content" : "100%" },
			...tableProps,
			children
		})]
	});
});
ScrollableTable.displayName = "ScrollableTable";
//#endregion
exports.ScrollableTable = ScrollableTable;
exports.Table = Table;
exports.TableBody = TableBody;
exports.TableCell = TableCell;
exports.TableFooter = TableFooter;
exports.TableHead = TableHead;
exports.TableHeader = TableHeader;
exports.TableRow = TableRow;
