// @openuidev/react-ui/dist/components/Table/Table.js
import { jsx as _jsx, jsxs as _jsxs, Fragment as _Fragment } from "react/jsx-runtime";
import clsx from "clsx";
import { ChevronLeft, ChevronRight } from "lucide-react";
import React, { useCallback, useEffect, useRef, useState } from "react";
import { IconButton } from "../IconButton";
const alignmentClasses = {
    left: "openui-table-align-left",
    center: "openui-table-align-center",
    right: "openui-table-align-right",
};
// ─── Primitives ───
const Table = React.forwardRef(({ className, containerRef, containerClassName, containerStyle, ...props }, ref) => (_jsx("div", { ref: containerRef, className: clsx("openui-table-container", containerClassName), style: containerStyle, children: _jsx("table", { ref: ref, className: clsx("openui-table", className), ...props }) })));
Table.displayName = "Table";
const TableHeader = React.forwardRef(({ className, ...props }, ref) => (_jsx("thead", { ref: ref, className: clsx("openui-table-header", className), ...props })));
TableHeader.displayName = "TableHeader";
const TableBody = React.forwardRef(({ className, ...props }, ref) => (_jsx("tbody", { ref: ref, className: clsx("openui-table-body", className), ...props })));
TableBody.displayName = "TableBody";
const TableFooter = React.forwardRef(({ className, ...props }, ref) => (_jsx("tfoot", { ref: ref, className: clsx("openui-table-footer", className), ...props })));
TableFooter.displayName = "TableFooter";
const TableRow = React.forwardRef(({ className, ...props }, ref) => (_jsx("tr", { ref: ref, className: clsx("openui-table-row", className), ...props })));
TableRow.displayName = "TableRow";
const TableHead = React.forwardRef(({ className, children, icon, align, ...props }, ref) => (_jsx("th", { ref: ref, className: clsx("openui-table-head", className), ...props, children: _jsxs("div", { className: clsx("openui-table-head-content", align && alignmentClasses[align]), children: [icon && _jsx("div", { className: "openui-table-head-icon", children: icon }), _jsx("div", { className: "openui-table-head-label", children: children })] }) })));
TableHead.displayName = "TableHead";
const TableCell = React.forwardRef(({ className, align, ...props }, ref) => (_jsx("td", { ref: ref, className: clsx("openui-table-cell", align && alignmentClasses[align], className), ...props })));
TableCell.displayName = "TableCell";
const ScrollableTable = React.forwardRef(({ className, containerClassName, containerStyle, children, ...tableProps }, ref) => {
    const wrapperRef = useRef(null);
    const scrollContainerRef = useRef(null);
    const [isScrollable, setIsScrollable] = useState(false);
    const [canScrollLeft, setCanScrollLeft] = useState(false);
    const [canScrollRight, setCanScrollRight] = useState(false);
    const columnLeftsRef = useRef([]);
    useEffect(() => {
        const container = scrollContainerRef.current;
        const wrapper = wrapperRef.current;
        if (!container || !wrapper)
            return;
        const compute = () => {
            const scrollable = container.scrollWidth > container.clientWidth;
            setIsScrollable(scrollable);
            setCanScrollLeft(container.scrollLeft > 0);
            setCanScrollRight(container.scrollLeft < container.scrollWidth - container.clientWidth - 1);
            const ths = wrapper.querySelectorAll("thead th");
            if (!ths || ths.length === 0) {
                columnLeftsRef.current = [];
                return;
            }
            const containerRect = container.getBoundingClientRect();
            const lefts = [];
            ths.forEach((th) => {
                const rect = th.getBoundingClientRect();
                const left = rect.left - containerRect.left + container.scrollLeft;
                lefts.push(Math.max(0, Math.round(left)));
            });
            columnLeftsRef.current = Array.from(new Set(lefts)).sort((a, b) => a - b);
        };
        compute();
        const ro = new ResizeObserver(() => compute());
        ro.observe(container);
        const onScroll = () => {
            setCanScrollLeft(container.scrollLeft > 0);
            setCanScrollRight(container.scrollLeft < container.scrollWidth - container.clientWidth - 1);
        };
        container.addEventListener("scroll", onScroll);
        return () => {
            ro.disconnect();
            container.removeEventListener("scroll", onScroll);
        };
    }, [children]);
    const scrollToNextColumn = useCallback(() => {
        const container = scrollContainerRef.current;
        if (!container || columnLeftsRef.current.length === 0)
            return;
        const current = container.scrollLeft;
        const target = columnLeftsRef.current.find((l) => l > current + 1);
        const maxScroll = container.scrollWidth - container.clientWidth;
        const next = typeof target === "number" ? target : maxScroll;
        container.scrollTo({ left: Math.min(next, maxScroll), behavior: "smooth" });
    }, []);
    const scrollToPrevColumn = useCallback(() => {
        const container = scrollContainerRef.current;
        if (!container || columnLeftsRef.current.length === 0)
            return;
        const current = container.scrollLeft;
        const prevs = columnLeftsRef.current.filter((l) => l < current - 1);
        const prev = prevs.at(-1) ?? 0;
        container.scrollTo({ left: Math.max(0, prev), behavior: "smooth" });
    }, []);
    return (_jsxs("div", { className: "openui-scrollable-table-wrapper", ref: wrapperRef, tabIndex: 0, children: [isScrollable && (_jsxs(_Fragment, { children: [_jsx("div", { className: clsx("openui-scrollable-table-control", "openui-scrollable-table-control-left", !canScrollLeft && "openui-scrollable-table-control-disabled"), children: _jsx(IconButton, { "aria-label": "Scroll left", size: "small", variant: "secondary", onClick: scrollToPrevColumn, disabled: !canScrollLeft, icon: _jsx(ChevronLeft, { size: 16 }) }) }), _jsx("div", { className: clsx("openui-scrollable-table-control", "openui-scrollable-table-control-right", !canScrollRight && "openui-scrollable-table-control-disabled"), children: _jsx(IconButton, { "aria-label": "Scroll right", size: "small", variant: "secondary", onClick: scrollToNextColumn, disabled: !canScrollRight, icon: _jsx(ChevronRight, { size: 16 }) }) })] })), _jsx(Table, { ref: ref, containerRef: scrollContainerRef, containerClassName: clsx("openui-scrollable-table-scroll-container", containerClassName), containerStyle: containerStyle, className: className, style: { width: isScrollable ? "max-content" : "100%" }, ...tableProps, children: children })] }));
});
ScrollableTable.displayName = "ScrollableTable";
export { ScrollableTable, Table, TableBody, TableCell, TableFooter, TableHead, TableHeader, TableRow, };
