// @openuidev/react-ui/dist/components/Callout/stories/callout.stories.js
import { Callout } from "../Callout";
const meta = {
    title: "Components/Callout",
    component: Callout,
    parameters: {
        layout: "centered",
        docs: {
            description: {
                component: "```tsx\nimport { Callout } from '@openui-ui/react-ui';\n```",
            },
        },
    },
    tags: ["!dev", "autodocs"],
    argTypes: {
        variant: {
            control: "select",
            options: ["neutral", "info", "warning", "success", "danger"],
            description: "The visual style variant of the callout",
            table: {
                category: "Appearance",
                type: { summary: "CalloutVariant" },
                defaultValue: { summary: "neutral" },
            },
        },
        title: {
            control: "text",
            description: "The title of the callout",
            table: {
                category: "Content",
                type: { summary: "React.ReactNode" },
            },
        },
        description: {
            control: "text",
            description: "The description text of the callout",
            table: {
                category: "Content",
                type: { summary: "React.ReactNode" },
            },
        },
    },
};
export default meta;
export const Default = {
    args: {
        title: "Important Information",
        description: "This is a neutral callout with some important information.",
        variant: "neutral",
    },
};
export const Info = {
    args: {
        title: "Did you know?",
        description: "This is an informational callout.",
        variant: "info",
    },
};
export const Success = {
    args: {
        title: "Success",
        description: "The operation completed successfully.",
        variant: "success",
    },
};
export const Warning = {
    args: {
        title: "Warning",
        description: "Please be careful with this action.",
        variant: "warning",
    },
};
export const Danger = {
    args: {
        title: "Error",
        description: "Something went wrong. Please try again.",
        variant: "danger",
    },
};
export const TitleOnly = {
    args: {
        title: "Title-only callout",
        variant: "info",
    },
};
export const DescriptionOnly = {
    args: {
        description: "A callout with only a description and no title.",
        variant: "neutral",
    },
};
