// @openuidev/react-ui/dist/components/Buttons/stories/buttons.stories.js
import { jsx as _jsx, jsxs as _jsxs } from "react/jsx-runtime";
import { ArrowLeft } from "lucide-react";
import { Button } from "../../Button";
import { IconButton } from "../../IconButton";
import { Buttons } from "../Buttons";
const meta = {
    title: "Components/Buttons",
    component: Buttons,
    parameters: {
        layout: "centered",
        docs: {
            description: {
                component: "```tsx\nimport { Buttons } from '@openui-ui/react-ui';\n```",
            },
        },
    },
    decorators: [
        (Story) => (_jsx("div", { style: { width: "350px" }, children: _jsx(Story, {}) })),
    ],
    argTypes: {
        variant: {
            control: "radio",
            options: ["horizontal", "vertical"],
            table: {
                category: "Appearance",
                type: { summary: "'horizontal' | 'vertical'" },
                defaultValue: { summary: "'horizontal'" },
            },
        },
        children: {
            control: false,
            description: "The child elements of the footer that accept Button and IconButton components",
            table: {
                category: "Content",
                type: {
                    summary: "ReactNode[]<ButtonProps | IconButtonProps> | ReactNode<ButtonProps | IconButtonProps>",
                },
            },
        },
        className: {
            control: false,
            description: "Additional CSS class name for custom styling",
            table: {
                category: "Styling",
                type: { summary: "string" },
            },
        },
        style: {
            control: false,
            description: "Inline CSS styles for custom styling",
            table: {
                category: "Styling",
                type: { summary: "CSSProperties" },
            },
        },
    },
    tags: ["!dev", "autodocs"],
};
export default meta;
// Basic footer stories
export const ButtonsStory = {
    args: {
        variant: "horizontal",
    },
    render: (args) => (_jsxs(Buttons, { variant: args.variant, children: [_jsx(Button, { children: "Cancel" }), _jsx(Button, { children: "Save" }), _jsx(IconButton, { icon: _jsx(ArrowLeft, { size: 16 }) })] })),
};
