// @openuidev/react-ui/dist/components/Image/index.cjs
Object.defineProperty(exports, Symbol.toStringTag, { value: "Module" });
const require_chunk = require("../../chunk-CKQMccvm.cjs");
let clsx = require("clsx");
clsx = require_chunk.__toESM(clsx);
let react = require("react");
react = require_chunk.__toESM(react);
let react_jsx_runtime = require("react/jsx-runtime");
let _radix_ui_react_aspect_ratio = require("@radix-ui/react-aspect-ratio");
_radix_ui_react_aspect_ratio = require_chunk.__toESM(_radix_ui_react_aspect_ratio);
//#region src/components/Image/Image.tsx
const aspectRatioMap = {
	"1:1": 1,
	"3:2": 3 / 2,
	"3:4": 3 / 4,
	"4:3": 4 / 3,
	"16:9": 16 / 9
};
const scaleMap = {
	fit: "openui-image-fit",
	fill: "openui-image-fill"
};
const Image = (0, react.forwardRef)((props, ref) => {
	const { src, alt, styles, className, aspectRatio = "3:2", scale = "fill", ...rest } = props;
	const image = /* @__PURE__ */ (0, react_jsx_runtime.jsx)("img", {
		ref,
		src,
		alt,
		className: (0, clsx.default)("openui-image", { [`${scaleMap[scale]}`]: scale }, className),
		style: styles,
		onError: (e) => {
			e.currentTarget.style.display = "none";
			console.error(`Failed to load image: ${src}`);
		},
		...rest
	});
	return /* @__PURE__ */ (0, react_jsx_runtime.jsx)(_radix_ui_react_aspect_ratio.Root, {
		ratio: aspectRatioMap[aspectRatio],
		children: image
	});
});
Image.displayName = "Image";
//#endregion
exports.Image = Image;
