// @openuidev/react-ui/dist/ArtifactOverlay-7Tt6OY58.cjs
const require_chunk = require("./chunk-CKQMccvm.cjs");
const require_useMultipleRefs = require("./useMultipleRefs-COMoDkic.cjs");
let clsx = require("clsx");
clsx = require_chunk.__toESM(clsx);
let react = require("react");
let react_jsx_runtime = require("react/jsx-runtime");
let zustand = require("zustand");
let zustand_react_shallow = require("zustand/react/shallow");
let _openuidev_react_headless = require("@openuidev/react-headless");
//#region src/components/_shared/store/store.tsx
const createShellStore = ({ logoUrl, agentName }) => (0, zustand.create)((set) => ({
	isSidebarOpen: true,
	agentName,
	logoUrl,
	setIsSidebarOpen: (isOpen) => set({ isSidebarOpen: isOpen }),
	setAgentName: (name) => set({ agentName: name }),
	setLogoUrl: (url) => set({ logoUrl: url })
}));
const ShellStoreContext = (0, react.createContext)(null);
const useShellStore = (selector) => {
	const store = (0, react.useContext)(ShellStoreContext);
	if (!store) throw new Error("useShellStore must be used within ShellStoreProvider");
	return store((0, zustand_react_shallow.useShallow)(selector));
};
const ShellStoreProvider = ({ children, agentName, logoUrl }) => {
	const shellStore = (0, react.useMemo)(() => createShellStore({
		agentName,
		logoUrl
	}), []);
	(0, react.useEffect)(() => {
		const { setAgentName, setLogoUrl } = shellStore.getState();
		setAgentName(agentName);
		setLogoUrl(logoUrl);
	}, [agentName, logoUrl]);
	return /* @__PURE__ */ (0, react_jsx_runtime.jsx)(ShellStoreContext.Provider, {
		value: shellStore,
		children
	});
};
//#endregion
//#region src/components/_shared/utils/index.ts
const isChatEmpty = ({ isLoadingMessages, messages }) => {
	return !isLoadingMessages && messages.length === 0;
};
//#endregion
//#region src/components/_shared/artifact/ArtifactPortalTarget.tsx
/**
* Registers a DOM node as the render target for {@link ArtifactPanel} portals.
*
* Mount exactly one instance in your layout. Renders a `<div>` with
* `display: contents` so it doesn't affect layout flow.
*
* @category Components
*/
const ArtifactPortalTarget = (0, react.forwardRef)(({ className }, ref) => {
	const { setNode } = (0, _openuidev_react_headless.useArtifactPortalTarget)();
	const forwardedRef = (0, react.useRef)(ref);
	forwardedRef.current = ref;
	return /* @__PURE__ */ (0, react_jsx_runtime.jsx)("div", {
		ref: (0, react.useCallback)((node) => {
			setNode(node);
			const fRef = forwardedRef.current;
			if (typeof fRef === "function") fRef(node);
			else if (fRef) fRef.current = node;
		}, [setNode]),
		className,
		style: { display: "contents" }
	});
});
ArtifactPortalTarget.displayName = "ArtifactPortalTarget";
//#endregion
//#region src/components/_shared/artifact/ArtifactOverlay.tsx
/**
* Shared overlay wrapper for the artifact portal target.
* Used by CopilotShell, BottomTray, and Shell (mobile) layouts.
* Renders an absolute-positioned overlay with slide-in/slide-out animations.
*
* @category Components
*/
const ArtifactOverlay = (0, react.forwardRef)(({ className }, ref) => {
	const { isArtifactActive } = (0, _openuidev_react_headless.useActiveArtifact)();
	const [shouldRender, setShouldRender] = (0, react.useState)(isArtifactActive);
	const [isExiting, setIsExiting] = (0, react.useState)(false);
	const internalRef = (0, react.useRef)(null);
	const mergedRef = require_useMultipleRefs.useMultipleRefs(ref, internalRef);
	(0, react.useEffect)(() => {
		if (isArtifactActive) {
			setShouldRender(true);
			setIsExiting(false);
		} else if (shouldRender) setIsExiting(true);
	}, [isArtifactActive]);
	const handleAnimationEnd = (0, react.useCallback)((e) => {
		if (e.target !== internalRef.current) return;
		if (isExiting) {
			setShouldRender(false);
			setIsExiting(false);
		}
	}, [isExiting]);
	if (!shouldRender) return null;
	return /* @__PURE__ */ (0, react_jsx_runtime.jsx)("div", {
		ref: mergedRef,
		className: (0, clsx.default)("openui-artifact-overlay", { "openui-artifact-overlay--exiting": isExiting }, className),
		onAnimationEnd: handleAnimationEnd,
		children: /* @__PURE__ */ (0, react_jsx_runtime.jsx)(ArtifactPortalTarget, {})
	});
});
ArtifactOverlay.displayName = "ArtifactOverlay";
//#endregion
Object.defineProperty(exports, "ArtifactOverlay", {
	enumerable: true,
	get: function() {
		return ArtifactOverlay;
	}
});
Object.defineProperty(exports, "ArtifactPortalTarget", {
	enumerable: true,
	get: function() {
		return ArtifactPortalTarget;
	}
});
Object.defineProperty(exports, "ShellStoreContext", {
	enumerable: true,
	get: function() {
		return ShellStoreContext;
	}
});
Object.defineProperty(exports, "ShellStoreProvider", {
	enumerable: true,
	get: function() {
		return ShellStoreProvider;
	}
});
Object.defineProperty(exports, "createShellStore", {
	enumerable: true,
	get: function() {
		return createShellStore;
	}
});
Object.defineProperty(exports, "isChatEmpty", {
	enumerable: true,
	get: function() {
		return isChatEmpty;
	}
});
Object.defineProperty(exports, "useShellStore", {
	enumerable: true,
	get: function() {
		return useShellStore;
	}
});
