// @openuidev/react-ui/dist/CopilotShell-CinO8LKZ.cjs
const require_chunk = require("./chunk-CKQMccvm.cjs");
const require_LayoutContext = require("./LayoutContext-qUdkWJWV.cjs");
const require_ArtifactOverlay = require("./ArtifactOverlay-7Tt6OY58.cjs");
const require_components_Separator_index = require("./components/Separator/index.cjs");
const require_components_IconButton_index = require("./components/IconButton/index.cjs");
const require_useComposerState = require("./useComposerState-CzLgzXdv.cjs");
const require_components_MarkDownRenderer_index = require("./components/MarkDownRenderer/index.cjs");
const require_components_MessageLoading_index = require("./components/MessageLoading/index.cjs");
const require_ToolCall = require("./ToolCall-BXicKrtH.cjs");
const require_ToolResult = require("./ToolResult-CeQt4t3K.cjs");
let clsx = require("clsx");
clsx = require_chunk.__toESM(clsx);
let lucide_react = require("lucide-react");
let react = require("react");
react = require_chunk.__toESM(react);
let react_jsx_runtime = require("react/jsx-runtime");
let _openuidev_react_headless = require("@openuidev/react-headless");
//#region src/components/CopilotShell/Container.tsx
const Container = ({ children, logoUrl, agentName, className }) => {
	return /* @__PURE__ */ (0, react_jsx_runtime.jsx)(require_ArtifactOverlay.ShellStoreProvider, {
		logoUrl,
		agentName,
		children: /* @__PURE__ */ (0, react_jsx_runtime.jsx)(require_LayoutContext.LayoutContextProvider, {
			layout: "tray",
			children: /* @__PURE__ */ (0, react_jsx_runtime.jsx)("div", {
				className: (0, clsx.default)("openui-copilot-shell-container", className),
				children
			})
		})
	});
};
//#endregion
//#region src/components/CopilotShell/ConversationStarter.tsx
/**
* Renders the appropriate icon based on the icon prop value
* - undefined: Show default lightbulb icon
* - ReactNode: Show the provided icon (use <></> or React.Fragment for no icon)
*/
const renderIcon = (icon) => {
	if (icon === void 0) return /* @__PURE__ */ (0, react_jsx_runtime.jsx)(lucide_react.Lightbulb, { size: 16 });
	return icon;
};
const ConversationStarterItem = ({ displayText, prompt, onClick, variant, icon }) => {
	const renderedIcon = renderIcon(icon);
	if (variant === "short") return /* @__PURE__ */ (0, react_jsx_runtime.jsxs)("button", {
		type: "button",
		className: "openui-copilot-shell-conversation-starter-item-short",
		onClick: () => onClick(prompt),
		children: [renderedIcon && /* @__PURE__ */ (0, react_jsx_runtime.jsx)("span", {
			className: "openui-copilot-shell-conversation-starter-item-short__icon",
			children: renderedIcon
		}), /* @__PURE__ */ (0, react_jsx_runtime.jsx)("span", {
			className: "openui-copilot-shell-conversation-starter-item-short__text",
			children: displayText
		})]
	});
	return /* @__PURE__ */ (0, react_jsx_runtime.jsxs)("button", {
		type: "button",
		className: "openui-copilot-shell-conversation-starter-item-long",
		onClick: () => onClick(prompt),
		children: [/* @__PURE__ */ (0, react_jsx_runtime.jsxs)("div", {
			className: "openui-copilot-shell-conversation-starter-item-long__content",
			children: [renderedIcon && /* @__PURE__ */ (0, react_jsx_runtime.jsx)("span", {
				className: "openui-copilot-shell-conversation-starter-item-long__icon",
				children: renderedIcon
			}), /* @__PURE__ */ (0, react_jsx_runtime.jsx)("span", {
				className: "openui-copilot-shell-conversation-starter-item-long__text",
				children: displayText
			})]
		}), /* @__PURE__ */ (0, react_jsx_runtime.jsx)("span", {
			className: "openui-copilot-shell-conversation-starter-item-long__arrow",
			children: /* @__PURE__ */ (0, react_jsx_runtime.jsx)(lucide_react.ArrowUp, { size: 16 })
		})]
	});
};
const ConversationStarter = ({ starters, className, variant = "short" }) => {
	const processMessage = (0, _openuidev_react_headless.useThread)((s) => s.processMessage);
	const isRunning = (0, _openuidev_react_headless.useThread)((s) => s.isRunning);
	const messages = (0, _openuidev_react_headless.useThread)((s) => s.messages);
	const isLoadingMessages = (0, _openuidev_react_headless.useThread)((s) => s.isLoadingMessages);
	const handleClick = (prompt) => {
		if (isRunning) return;
		processMessage({
			role: "user",
			content: prompt
		});
	};
	if (!require_ArtifactOverlay.isChatEmpty({
		isLoadingMessages,
		messages
	})) return null;
	if (starters.length === 0) return null;
	return /* @__PURE__ */ (0, react_jsx_runtime.jsx)("div", {
		className: (0, clsx.default)("openui-copilot-shell-conversation-starter", `openui-copilot-shell-conversation-starter--${variant}`, className),
		children: starters.map((item, index) => /* @__PURE__ */ (0, react_jsx_runtime.jsxs)(react.Fragment, { children: [/* @__PURE__ */ (0, react_jsx_runtime.jsx)(ConversationStarterItem, {
			displayText: item.displayText,
			prompt: item.prompt,
			icon: item.icon,
			onClick: handleClick,
			variant
		}), variant === "long" && index < starters.length - 1 && /* @__PURE__ */ (0, react_jsx_runtime.jsx)("div", {
			className: "openui-copilot-shell-conversation-starter__separator",
			children: /* @__PURE__ */ (0, react_jsx_runtime.jsx)(require_components_Separator_index.Separator, {})
		})] }, `${item.displayText}-${index}`))
	});
};
//#endregion
//#region src/components/CopilotShell/Header.tsx
const Header = ({ className, rightChildren }) => {
	const { logoUrl, agentName } = require_ArtifactOverlay.useShellStore((state) => ({
		logoUrl: state.logoUrl,
		agentName: state.agentName
	}));
	return /* @__PURE__ */ (0, react_jsx_runtime.jsxs)("div", {
		className: (0, clsx.default)("openui-copilot-shell-header", className),
		children: [/* @__PURE__ */ (0, react_jsx_runtime.jsxs)("div", {
			className: "openui-copilot-shell-header-logo-container",
			children: [/* @__PURE__ */ (0, react_jsx_runtime.jsx)("img", {
				className: "openui-copilot-shell-header-logo",
				src: logoUrl,
				alt: "Logo"
			}), /* @__PURE__ */ (0, react_jsx_runtime.jsx)("span", {
				className: "openui-copilot-shell-header-agent-name",
				children: agentName
			})]
		}), rightChildren && /* @__PURE__ */ (0, react_jsx_runtime.jsx)("div", {
			className: "openui-copilot-shell-header-right-content",
			children: rightChildren
		})]
	});
};
//#endregion
//#region src/components/CopilotShell/components/Composer.tsx
const Composer = ({ className, placeholder = "Type your message..." }) => {
	const { textContent, setTextContent } = require_useComposerState.useComposerState();
	const processMessage = (0, _openuidev_react_headless.useThread)((s) => s.processMessage);
	const cancelMessage = (0, _openuidev_react_headless.useThread)((s) => s.cancelMessage);
	const isRunning = (0, _openuidev_react_headless.useThread)((s) => s.isRunning);
	const isLoadingMessages = (0, _openuidev_react_headless.useThread)((s) => s.isLoadingMessages);
	const inputRef = (0, react.useRef)(null);
	const handleSubmit = () => {
		if (!textContent.trim() || isRunning || isLoadingMessages) return;
		processMessage({
			role: "user",
			content: textContent
		});
		setTextContent("");
	};
	(0, react.useLayoutEffect)(() => {
		const input = inputRef.current;
		if (!input) return;
		input.style.height = "0px";
		input.style.height = `${Math.max(input.scrollHeight, 24)}px`;
	}, [textContent]);
	return /* @__PURE__ */ (0, react_jsx_runtime.jsx)("div", {
		className: (0, clsx.default)("openui-copilot-shell-thread-composer", className),
		onClick: (e) => {
			if (!e.target.closest("button, a, [role='button']")) inputRef.current?.focus();
		},
		children: /* @__PURE__ */ (0, react_jsx_runtime.jsxs)("div", {
			className: "openui-copilot-shell-thread-composer__input-wrapper",
			children: [/* @__PURE__ */ (0, react_jsx_runtime.jsx)("textarea", {
				ref: inputRef,
				value: textContent,
				onChange: (e) => setTextContent(e.target.value),
				className: "openui-copilot-shell-thread-composer__input",
				placeholder,
				rows: 1,
				onKeyDown: (e) => {
					if (e.key === "Enter" && !e.shiftKey) {
						e.preventDefault();
						handleSubmit();
					}
				}
			}), /* @__PURE__ */ (0, react_jsx_runtime.jsx)("div", {
				className: "openui-copilot-shell-thread-composer__action-bar",
				children: /* @__PURE__ */ (0, react_jsx_runtime.jsx)(require_components_IconButton_index.IconButton, {
					onClick: isRunning ? cancelMessage : handleSubmit,
					icon: isRunning ? /* @__PURE__ */ (0, react_jsx_runtime.jsx)(lucide_react.Square, {
						size: "1em",
						fill: "currentColor"
					}) : /* @__PURE__ */ (0, react_jsx_runtime.jsx)(lucide_react.ArrowUp, { size: "1em" }),
					size: "medium",
					variant: "primary",
					className: "openui-copilot-shell-thread-composer__submit-button"
				})
			})]
		})
	});
};
//#endregion
//#region src/components/CopilotShell/Thread.tsx
const ThreadContainer = ({ children, className }) => {
	const isLoadingMessages = (0, _openuidev_react_headless.useThread)((s) => s.isLoadingMessages);
	return /* @__PURE__ */ (0, react_jsx_runtime.jsxs)("div", {
		className: (0, clsx.default)("openui-copilot-shell-thread-container", className),
		style: { visibility: isLoadingMessages ? "hidden" : void 0 },
		children: [children, /* @__PURE__ */ (0, react_jsx_runtime.jsx)(require_ArtifactOverlay.ArtifactOverlay, {})]
	});
};
const ScrollArea = ({ children, className, scrollVariant = "user-message-anchor", userMessageSelector = ".openui-copilot-shell-thread-message-user" }) => {
	const ref = (0, react.useRef)(null);
	const messages = (0, _openuidev_react_headless.useThread)((s) => s.messages);
	const isRunning = (0, _openuidev_react_headless.useThread)((s) => s.isRunning);
	const isLoadingMessages = (0, _openuidev_react_headless.useThread)((s) => s.isLoadingMessages);
	require_useComposerState.useScrollToBottom({
		ref,
		lastMessage: messages[messages.length - 1] || { id: "" },
		scrollVariant,
		userMessageSelector,
		isRunning,
		isLoadingMessages
	});
	return /* @__PURE__ */ (0, react_jsx_runtime.jsx)("div", {
		className: "openui-copilot-shell-thread-scroll-container",
		children: /* @__PURE__ */ (0, react_jsx_runtime.jsx)("div", {
			ref,
			className: (0, clsx.default)("openui-copilot-shell-thread-scroll-area", { "openui-copilot-shell-thread-scroll-area--user-message-anchor": scrollVariant === "user-message-anchor" }, className),
			children
		})
	});
};
const AssistantMessageContainer = ({ children, className }) => {
	return /* @__PURE__ */ (0, react_jsx_runtime.jsx)("div", {
		className: (0, clsx.default)("openui-copilot-shell-thread-message-assistant", className),
		children: /* @__PURE__ */ (0, react_jsx_runtime.jsx)("div", {
			className: "openui-copilot-shell-thread-message-assistant__content",
			children
		})
	});
};
const UserMessageContainer = ({ children, className }) => {
	return /* @__PURE__ */ (0, react_jsx_runtime.jsx)("div", {
		className: (0, clsx.default)("openui-copilot-shell-thread-message-user", className),
		children: /* @__PURE__ */ (0, react_jsx_runtime.jsx)("div", {
			className: "openui-copilot-shell-thread-message-user__content",
			children
		})
	});
};
const AssistantMessageContent = ({ message, allMessages }) => {
	const getToolName = (toolCallId) => {
		return (message.toolCalls?.find((tc) => tc.id === toolCallId))?.function.name;
	};
	const toolMessages = [];
	const msgIndex = allMessages.findIndex((m) => m.id === message.id);
	if (msgIndex !== -1) for (let i = msgIndex + 1; i < allMessages.length; i++) {
		const m = allMessages[i];
		if (m && m.role === "tool") toolMessages.push(m);
		else break;
	}
	return /* @__PURE__ */ (0, react_jsx_runtime.jsxs)(react_jsx_runtime.Fragment, { children: [
		message.content && /* @__PURE__ */ (0, react_jsx_runtime.jsx)(require_components_MarkDownRenderer_index.MarkDownRenderer, {
			textMarkdown: message.content,
			className: "openui-copilot-shell-thread-message-assistant__text"
		}),
		message.toolCalls?.map((toolCall) => /* @__PURE__ */ (0, react_jsx_runtime.jsx)(require_ToolCall.ToolCallComponent, { toolCall }, toolCall.id)),
		toolMessages.map((tm) => /* @__PURE__ */ (0, react_jsx_runtime.jsx)(require_ToolResult.ToolResult, {
			message: tm,
			toolName: getToolName(tm.toolCallId)
		}, tm.id))
	] });
};
const UserMessageContent = ({ message }) => {
	if (message.role !== "user") return null;
	const content = message.content;
	if (typeof content === "string") return /* @__PURE__ */ (0, react_jsx_runtime.jsx)(react_jsx_runtime.Fragment, { children: content });
	return /* @__PURE__ */ (0, react_jsx_runtime.jsx)(react_jsx_runtime.Fragment, { children: content?.map((part, i) => {
		if (part.type === "text") return /* @__PURE__ */ (0, react_jsx_runtime.jsx)("span", { children: part.text }, i);
		if (part.type === "binary" && part.url) return /* @__PURE__ */ (0, react_jsx_runtime.jsx)("img", {
			src: part.url,
			alt: "",
			className: "openui-copilot-shell-thread-message-user__image"
		}, i);
		return null;
	}) });
};
const RenderMessage = (0, react.memo)(({ message, className, allMessages, assistantMessage: CustomAssistantMessage, userMessage: CustomUserMessage, isStreaming }) => {
	if (message.role === "tool") return null;
	if (message.role === "assistant") {
		if (CustomAssistantMessage) return /* @__PURE__ */ (0, react_jsx_runtime.jsx)(CustomAssistantMessage, {
			message,
			isStreaming
		});
		return /* @__PURE__ */ (0, react_jsx_runtime.jsx)(AssistantMessageContainer, {
			className,
			children: /* @__PURE__ */ (0, react_jsx_runtime.jsx)(AssistantMessageContent, {
				message,
				allMessages
			})
		});
	}
	if (message.role === "user") {
		if (CustomUserMessage) return /* @__PURE__ */ (0, react_jsx_runtime.jsx)(CustomUserMessage, { message });
		return /* @__PURE__ */ (0, react_jsx_runtime.jsx)(UserMessageContainer, {
			className,
			children: /* @__PURE__ */ (0, react_jsx_runtime.jsx)(UserMessageContent, { message })
		});
	}
	return null;
});
const MessageLoading = () => {
	return /* @__PURE__ */ (0, react_jsx_runtime.jsx)("div", {
		className: "openui-copilot-shell-thread-message-loading",
		children: /* @__PURE__ */ (0, react_jsx_runtime.jsx)(require_components_MessageLoading_index.MessageLoading, {})
	});
};
const Messages = ({ className, loader, assistantMessage, userMessage }) => {
	const messages = (0, _openuidev_react_headless.useThread)((s) => s.messages);
	const isRunning = (0, _openuidev_react_headless.useThread)((s) => s.isRunning);
	return /* @__PURE__ */ (0, react_jsx_runtime.jsxs)("div", {
		className: (0, clsx.default)("openui-copilot-shell-thread-messages", className),
		children: [messages.map((message, i) => {
			return /* @__PURE__ */ (0, react_jsx_runtime.jsx)(_openuidev_react_headless.MessageProvider, {
				message,
				children: /* @__PURE__ */ (0, react_jsx_runtime.jsx)(RenderMessage, {
					message,
					allMessages: messages,
					assistantMessage,
					userMessage,
					isStreaming: isRunning && i === messages.length - 1
				})
			}, message.id);
		}), isRunning && /* @__PURE__ */ (0, react_jsx_runtime.jsx)("div", { children: loader })]
	});
};
//#endregion
//#region src/components/CopilotShell/WelcomeScreen.tsx
/**
* Type guard to check if image is a URL object
*/
const isImageUrl = (image) => {
	return typeof image === "object" && image !== null && "url" in image;
};
const WelcomeScreen = (props) => {
	const { className } = props;
	const messages = (0, _openuidev_react_headless.useThread)((s) => s.messages);
	if (!require_ArtifactOverlay.isChatEmpty({
		isLoadingMessages: (0, _openuidev_react_headless.useThread)((s) => s.isLoadingMessages),
		messages
	})) return null;
	if ("children" in props && props.children) return /* @__PURE__ */ (0, react_jsx_runtime.jsx)("div", {
		className: (0, clsx.default)("openui-copilot-shell-welcome-screen", className),
		children: props.children
	});
	const { title, description, image } = props;
	const renderImage = () => {
		if (!image) return null;
		if (isImageUrl(image)) return /* @__PURE__ */ (0, react_jsx_runtime.jsx)("img", {
			src: image.url,
			alt: title || "",
			className: "openui-copilot-shell-welcome-screen__image"
		});
		return image;
	};
	return /* @__PURE__ */ (0, react_jsx_runtime.jsxs)("div", {
		className: (0, clsx.default)("openui-copilot-shell-welcome-screen", className),
		children: [image && /* @__PURE__ */ (0, react_jsx_runtime.jsx)("div", {
			className: "openui-copilot-shell-welcome-screen__image-container",
			children: renderImage()
		}), (title || description) && /* @__PURE__ */ (0, react_jsx_runtime.jsxs)("div", {
			className: "openui-copilot-shell-welcome-screen__content",
			children: [title && /* @__PURE__ */ (0, react_jsx_runtime.jsx)("h2", {
				className: "openui-copilot-shell-welcome-screen__title",
				children: title
			}), description && /* @__PURE__ */ (0, react_jsx_runtime.jsx)("p", {
				className: "openui-copilot-shell-welcome-screen__description",
				children: description
			})]
		})]
	});
};
//#endregion
Object.defineProperty(exports, "AssistantMessageContainer", {
	enumerable: true,
	get: function() {
		return AssistantMessageContainer;
	}
});
Object.defineProperty(exports, "Composer", {
	enumerable: true,
	get: function() {
		return Composer;
	}
});
Object.defineProperty(exports, "Container", {
	enumerable: true,
	get: function() {
		return Container;
	}
});
Object.defineProperty(exports, "ConversationStarter", {
	enumerable: true,
	get: function() {
		return ConversationStarter;
	}
});
Object.defineProperty(exports, "Header", {
	enumerable: true,
	get: function() {
		return Header;
	}
});
Object.defineProperty(exports, "MessageLoading", {
	enumerable: true,
	get: function() {
		return MessageLoading;
	}
});
Object.defineProperty(exports, "Messages", {
	enumerable: true,
	get: function() {
		return Messages;
	}
});
Object.defineProperty(exports, "RenderMessage", {
	enumerable: true,
	get: function() {
		return RenderMessage;
	}
});
Object.defineProperty(exports, "ScrollArea", {
	enumerable: true,
	get: function() {
		return ScrollArea;
	}
});
Object.defineProperty(exports, "ThreadContainer", {
	enumerable: true,
	get: function() {
		return ThreadContainer;
	}
});
Object.defineProperty(exports, "UserMessageContainer", {
	enumerable: true,
	get: function() {
		return UserMessageContainer;
	}
});
Object.defineProperty(exports, "WelcomeScreen", {
	enumerable: true,
	get: function() {
		return WelcomeScreen;
	}
});
