// @openuidev/react-ui/dist/Shell-CXCVPlSX.cjs
const require_chunk = require("./chunk-CKQMccvm.cjs");
const require_LayoutContext = require("./LayoutContext-qUdkWJWV.cjs");
const require_ArtifactOverlay = require("./ArtifactOverlay-7Tt6OY58.cjs");
const require_components_Separator_index = require("./components/Separator/index.cjs");
const require_components_IconButton_index = require("./components/IconButton/index.cjs");
const require_useComposerState = require("./useComposerState-CzLgzXdv.cjs");
const require_components_MarkDownRenderer_index = require("./components/MarkDownRenderer/index.cjs");
const require_components_MessageLoading_index = require("./components/MessageLoading/index.cjs");
const require_ToolCall = require("./ToolCall-BXicKrtH.cjs");
const require_ToolResult = require("./ToolResult-CeQt4t3K.cjs");
const require_components_Button_index = require("./components/Button/index.cjs");
const require_components_Callout_index = require("./components/Callout/index.cjs");
let clsx = require("clsx");
clsx = require_chunk.__toESM(clsx);
let lucide_react = require("lucide-react");
let react = require("react");
react = require_chunk.__toESM(react);
let react_jsx_runtime = require("react/jsx-runtime");
let _openuidev_react_headless = require("@openuidev/react-headless");
let _radix_ui_react_dropdown_menu = require("@radix-ui/react-dropdown-menu");
_radix_ui_react_dropdown_menu = require_chunk.__toESM(_radix_ui_react_dropdown_menu);
//#region src/utils/contentParser.ts
const openTag = (tag) => `<${tag}>`;
const closeTag = (tag) => `</${tag}>`;
function wrapContent(text) {
	return `${openTag("content")}${text}${closeTag("content")}`;
}
function wrapContext(json) {
	return `${openTag("context")}${json}${closeTag("context")}`;
}
/**
* Separate openui-lang code from <context> tag in a message.
* Returns { content: the message/code, contextString: raw JSON or null }
*/
function separateContentAndContext(raw) {
	const contextMatch = raw.match(/<context>([\s\S]*)<\/context>\s*$/);
	let content = raw;
	let contextString = null;
	if (contextMatch) {
		contextString = contextMatch[1] ?? null;
		content = raw.slice(0, contextMatch.index).trimEnd();
	}
	const contentMatch = content.match(/^<content[^>]*>([\s\S]*)<\/content>\s*$/);
	if (contentMatch) content = contentMatch[1] ?? content;
	return {
		content,
		contextString
	};
}
//#endregion
//#region src/components/Shell/components/Composer.tsx
const Composer = ({ className, placeholder = "Type your query here" }) => {
	const { textContent, setTextContent } = require_useComposerState.useComposerState();
	const processMessage = (0, _openuidev_react_headless.useThread)((s) => s.processMessage);
	const cancelMessage = (0, _openuidev_react_headless.useThread)((s) => s.cancelMessage);
	const isRunning = (0, _openuidev_react_headless.useThread)((s) => s.isRunning);
	const isLoadingMessages = (0, _openuidev_react_headless.useThread)((s) => s.isLoadingMessages);
	const inputRef = (0, react.useRef)(null);
	const handleSubmit = () => {
		if (!textContent.trim() || isRunning || isLoadingMessages) return;
		processMessage({
			role: "user",
			content: textContent
		});
		setTextContent("");
	};
	(0, react.useLayoutEffect)(() => {
		const input = inputRef.current;
		if (!input) return;
		input.style.height = "0px";
		input.style.height = `${Math.max(input.scrollHeight, 24)}px`;
	}, [textContent]);
	return /* @__PURE__ */ (0, react_jsx_runtime.jsx)("div", {
		className: (0, clsx.default)("openui-shell-thread-composer", className),
		onClick: (e) => {
			if (!e.target.closest("button, a, [role='button']")) inputRef.current?.focus();
		},
		children: /* @__PURE__ */ (0, react_jsx_runtime.jsxs)("div", {
			className: "openui-shell-thread-composer__input-wrapper",
			children: [/* @__PURE__ */ (0, react_jsx_runtime.jsx)("textarea", {
				ref: inputRef,
				value: textContent,
				onChange: (e) => setTextContent(e.target.value),
				className: "openui-shell-thread-composer__input",
				placeholder,
				rows: 1,
				onKeyDown: (e) => {
					if (e.key === "Enter" && !e.shiftKey) {
						e.preventDefault();
						handleSubmit();
					}
				}
			}), /* @__PURE__ */ (0, react_jsx_runtime.jsx)("div", {
				className: "openui-shell-thread-composer__action-bar",
				children: /* @__PURE__ */ (0, react_jsx_runtime.jsx)(require_components_IconButton_index.IconButton, {
					onClick: isRunning ? cancelMessage : handleSubmit,
					icon: isRunning ? /* @__PURE__ */ (0, react_jsx_runtime.jsx)(lucide_react.Square, {
						size: "1em",
						fill: "currentColor"
					}) : /* @__PURE__ */ (0, react_jsx_runtime.jsx)(lucide_react.ArrowUp, { size: "1em" }),
					size: "medium",
					variant: "primary",
					className: "openui-shell-thread-composer__submit-button"
				})
			})]
		})
	});
};
//#endregion
//#region src/components/Shell/components/DesktopWelcomeComposer.tsx
const DesktopWelcomeComposer = ({ className, placeholder = "Type your query here" }) => {
	const { textContent, setTextContent } = require_useComposerState.useComposerState();
	const processMessage = (0, _openuidev_react_headless.useThread)((s) => s.processMessage);
	const cancelMessage = (0, _openuidev_react_headless.useThread)((s) => s.cancelMessage);
	const isRunning = (0, _openuidev_react_headless.useThread)((s) => s.isRunning);
	const isLoadingMessages = (0, _openuidev_react_headless.useThread)((s) => s.isLoadingMessages);
	const inputRef = (0, react.useRef)(null);
	const handleSubmit = () => {
		if (!textContent.trim() || isRunning || isLoadingMessages) return;
		processMessage({
			role: "user",
			content: textContent
		});
		setTextContent("");
	};
	(0, react.useLayoutEffect)(() => {
		const input = inputRef.current;
		if (!input) return;
		input.style.height = "0px";
		input.style.height = `${Math.max(input.scrollHeight, 24)}px`;
	}, [textContent]);
	return /* @__PURE__ */ (0, react_jsx_runtime.jsxs)("div", {
		className: (0, clsx.default)("openui-shell-desktop-welcome-composer", className),
		children: [/* @__PURE__ */ (0, react_jsx_runtime.jsx)("textarea", {
			ref: inputRef,
			value: textContent,
			onChange: (e) => setTextContent(e.target.value),
			className: "openui-shell-desktop-welcome-composer__input",
			placeholder,
			rows: 1,
			onKeyDown: (e) => {
				if (e.key === "Enter" && !e.shiftKey) {
					e.preventDefault();
					handleSubmit();
				}
			}
		}), /* @__PURE__ */ (0, react_jsx_runtime.jsx)("div", {
			className: "openui-shell-desktop-welcome-composer__action-bar",
			children: /* @__PURE__ */ (0, react_jsx_runtime.jsx)(require_components_IconButton_index.IconButton, {
				onClick: isRunning ? cancelMessage : handleSubmit,
				disabled: !textContent.trim() && !isRunning,
				"aria-label": isRunning ? "Cancel" : "Send",
				icon: isRunning ? /* @__PURE__ */ (0, react_jsx_runtime.jsx)(lucide_react.Square, {
					size: "1em",
					fill: "currentColor"
				}) : /* @__PURE__ */ (0, react_jsx_runtime.jsx)(lucide_react.ArrowUp, { size: "1em" }),
				size: "medium",
				variant: "primary",
				className: "openui-shell-desktop-welcome-composer__submit-button"
			})
		})]
	});
};
//#endregion
//#region src/hooks/useElementSize.ts
const useElementSize = ({ ref }) => {
	const [size, setSize] = (0, react.useState)({
		width: 0,
		height: 0
	});
	(0, react.useEffect)(() => {
		if (!ref.current) return;
		const element = ref.current;
		const handleResize = () => {
			setSize({
				width: element.clientWidth,
				height: element.clientHeight
			});
		};
		const resizeObserver = new ResizeObserver(handleResize);
		resizeObserver.observe(element);
		handleResize();
		return () => {
			resizeObserver.disconnect();
		};
	}, [ref]);
	return size;
};
//#endregion
//#region src/components/Shell/Container.tsx
const Container = ({ children, logoUrl, agentName, className }) => {
	const ref = (0, react.useRef)(null);
	const { width } = useElementSize({ ref }) || {};
	const isMobile = width > 0 && width < 768;
	return /* @__PURE__ */ (0, react_jsx_runtime.jsx)(require_ArtifactOverlay.ShellStoreProvider, {
		logoUrl,
		agentName,
		children: /* @__PURE__ */ (0, react_jsx_runtime.jsx)(require_LayoutContext.LayoutContextProvider, {
			layout: isMobile ? "mobile" : width > 768 ? "fullscreen" : "tray",
			children: /* @__PURE__ */ (0, react_jsx_runtime.jsx)("div", {
				className: (0, clsx.default)("openui-shell-container", { "openui-shell-container--mobile": isMobile }, className),
				ref,
				children
			})
		})
	});
};
//#endregion
//#region src/components/Shell/ConversationStarter.tsx
/**
* Renders the appropriate icon based on the icon prop value
* - undefined: Show default lightbulb icon
* - ReactNode: Show the provided icon (use <></> or React.Fragment for no icon)
*/
const renderIcon = (icon) => {
	if (icon === void 0) return /* @__PURE__ */ (0, react_jsx_runtime.jsx)(lucide_react.Lightbulb, { size: 16 });
	return icon;
};
const ConversationStarterItem = ({ displayText, prompt, onClick, variant, icon }) => {
	const renderedIcon = renderIcon(icon);
	if (variant === "short") return /* @__PURE__ */ (0, react_jsx_runtime.jsxs)("button", {
		type: "button",
		className: "openui-shell-conversation-starter-item-short",
		onClick: () => onClick(prompt),
		children: [renderedIcon && /* @__PURE__ */ (0, react_jsx_runtime.jsx)("span", {
			className: "openui-shell-conversation-starter-item-short__icon",
			children: renderedIcon
		}), /* @__PURE__ */ (0, react_jsx_runtime.jsx)("span", {
			className: "openui-shell-conversation-starter-item-short__text",
			children: displayText
		})]
	});
	return /* @__PURE__ */ (0, react_jsx_runtime.jsxs)("button", {
		type: "button",
		className: "openui-shell-conversation-starter-item-long",
		onClick: () => onClick(prompt),
		children: [/* @__PURE__ */ (0, react_jsx_runtime.jsxs)("div", {
			className: "openui-shell-conversation-starter-item-long__content",
			children: [renderedIcon && /* @__PURE__ */ (0, react_jsx_runtime.jsx)("span", {
				className: "openui-shell-conversation-starter-item-long__icon",
				children: renderedIcon
			}), /* @__PURE__ */ (0, react_jsx_runtime.jsx)("span", {
				className: "openui-shell-conversation-starter-item-long__text",
				children: displayText
			})]
		}), /* @__PURE__ */ (0, react_jsx_runtime.jsx)("span", {
			className: "openui-shell-conversation-starter-item-long__arrow",
			children: /* @__PURE__ */ (0, react_jsx_runtime.jsx)(lucide_react.ArrowUp, { size: 16 })
		})]
	});
};
const ConversationStarter = ({ starters, className, variant = "short" }) => {
	const processMessage = (0, _openuidev_react_headless.useThread)((s) => s.processMessage);
	const isRunning = (0, _openuidev_react_headless.useThread)((s) => s.isRunning);
	const messages = (0, _openuidev_react_headless.useThread)((s) => s.messages);
	const isLoadingMessages = (0, _openuidev_react_headless.useThread)((s) => s.isLoadingMessages);
	const handleClick = (prompt) => {
		if (isRunning) return;
		processMessage({
			role: "user",
			content: prompt
		});
	};
	if (!require_ArtifactOverlay.isChatEmpty({
		isLoadingMessages,
		messages
	})) return null;
	if (starters.length === 0) return null;
	return /* @__PURE__ */ (0, react_jsx_runtime.jsx)("div", {
		className: (0, clsx.default)("openui-shell-conversation-starter", `openui-shell-conversation-starter--${variant}`, className),
		children: starters.map((item, index) => /* @__PURE__ */ (0, react_jsx_runtime.jsxs)(react.Fragment, { children: [/* @__PURE__ */ (0, react_jsx_runtime.jsx)(ConversationStarterItem, {
			displayText: item.displayText,
			prompt: item.prompt,
			icon: item.icon,
			onClick: handleClick,
			variant
		}), variant === "long" && index < starters.length - 1 && /* @__PURE__ */ (0, react_jsx_runtime.jsx)("div", {
			className: "openui-shell-conversation-starter__separator",
			children: /* @__PURE__ */ (0, react_jsx_runtime.jsx)(require_components_Separator_index.Separator, {})
		})] }, `${item.displayText}-${index}`))
	});
};
//#endregion
//#region src/components/Shell/MobileHeader.tsx
const MobileHeader = ({ className, rightChildren }) => {
	const switchToNewThread = (0, _openuidev_react_headless.useThreadList)((s) => s.switchToNewThread);
	const { logoUrl, agentName, setIsSidebarOpen } = require_ArtifactOverlay.useShellStore((state) => ({
		logoUrl: state.logoUrl,
		agentName: state.agentName,
		setIsSidebarOpen: state.setIsSidebarOpen
	}));
	return /* @__PURE__ */ (0, react_jsx_runtime.jsxs)("div", {
		className: (0, clsx.default)("openui-shell-mobile-header", className),
		children: [
			/* @__PURE__ */ (0, react_jsx_runtime.jsx)(require_components_IconButton_index.IconButton, {
				size: "medium",
				icon: /* @__PURE__ */ (0, react_jsx_runtime.jsx)(lucide_react.Menu, { size: "1em" }),
				onClick: () => setIsSidebarOpen(true),
				variant: "secondary"
			}),
			/* @__PURE__ */ (0, react_jsx_runtime.jsxs)("div", {
				className: "openui-shell-mobile-header-logo-container",
				children: [/* @__PURE__ */ (0, react_jsx_runtime.jsx)("img", {
					className: "openui-shell-mobile-header-logo",
					src: logoUrl,
					alt: "Logo"
				}), /* @__PURE__ */ (0, react_jsx_runtime.jsx)("span", {
					className: "openui-shell-mobile-header-agent-name",
					children: agentName
				})]
			}),
			/* @__PURE__ */ (0, react_jsx_runtime.jsxs)("div", {
				className: "openui-shell-mobile-header-actions",
				children: [rightChildren, /* @__PURE__ */ (0, react_jsx_runtime.jsx)(require_components_IconButton_index.IconButton, {
					size: "medium",
					icon: /* @__PURE__ */ (0, react_jsx_runtime.jsx)(lucide_react.Plus, { size: "1em" }),
					onClick: switchToNewThread,
					variant: "secondary"
				})]
			})
		]
	});
};
//#endregion
//#region src/components/Shell/NewChatButton.tsx
const NewChatButton = ({ className }) => {
	const switchToNewThread = (0, _openuidev_react_headless.useThreadList)((s) => s.switchToNewThread);
	const { isSidebarOpen } = require_ArtifactOverlay.useShellStore((state) => ({ isSidebarOpen: state.isSidebarOpen }));
	if (!isSidebarOpen) return /* @__PURE__ */ (0, react_jsx_runtime.jsx)(require_components_IconButton_index.IconButton, {
		icon: /* @__PURE__ */ (0, react_jsx_runtime.jsx)(lucide_react.SquarePen, { size: "1em" }),
		onClick: switchToNewThread,
		variant: "primary",
		size: "small",
		className: (0, clsx.default)("openui-shell-new-chat-button_collapsed", className)
	});
	return /* @__PURE__ */ (0, react_jsx_runtime.jsx)(require_components_Button_index.Button, {
		className: (0, clsx.default)("openui-shell-new-chat-button", className),
		iconRight: /* @__PURE__ */ (0, react_jsx_runtime.jsx)(lucide_react.Plus, {}),
		variant: "primary",
		size: "small",
		onClick: switchToNewThread,
		children: "New Chat"
	});
};
//#endregion
//#region src/components/Shell/Sidebar.tsx
const SidebarContainer = ({ children, className }) => {
	const { isSidebarOpen, setIsSidebarOpen } = require_ArtifactOverlay.useShellStore((state) => ({
		isSidebarOpen: state.isSidebarOpen,
		setIsSidebarOpen: state.setIsSidebarOpen
	}));
	const { isArtifactActive } = (0, _openuidev_react_headless.useActiveArtifact)();
	const { layout } = require_LayoutContext.useLayoutContext() || {};
	const isMobile = layout === "mobile";
	(0, react.useEffect)(() => {
		if (isMobile) setIsSidebarOpen(false);
		else setIsSidebarOpen(true);
	}, [isMobile]);
	return /* @__PURE__ */ (0, react_jsx_runtime.jsxs)(react_jsx_runtime.Fragment, { children: [isMobile && /* @__PURE__ */ (0, react_jsx_runtime.jsx)("div", {
		className: (0, clsx.default)("openui-shell-sidebar-container__overlay", { "openui-shell-sidebar-container__overlay--collapsed": !isSidebarOpen }),
		onClick: () => {
			setIsSidebarOpen(false);
		}
	}), /* @__PURE__ */ (0, react_jsx_runtime.jsx)("div", {
		className: (0, clsx.default)("openui-shell-sidebar-container", {
			"openui-shell-sidebar-container--collapsed": !isSidebarOpen,
			"openui-shell-sidebar-container--hidden": isArtifactActive && !isMobile
		}, className),
		children
	})] });
};
const SidebarHeader = ({ children, className }) => {
	const { agentName, logoUrl, setIsSidebarOpen, isSidebarOpen } = require_ArtifactOverlay.useShellStore((state) => ({
		agentName: state.agentName,
		logoUrl: state.logoUrl,
		setIsSidebarOpen: state.setIsSidebarOpen,
		isSidebarOpen: state.isSidebarOpen
	}));
	return /* @__PURE__ */ (0, react_jsx_runtime.jsxs)("div", {
		className: (0, clsx.default)("openui-shell-sidebar-header", { "openui-shell-sidebar-header--collapsed": !isSidebarOpen }, className),
		children: [/* @__PURE__ */ (0, react_jsx_runtime.jsxs)("div", {
			className: "openui-shell-sidebar-header__top-row",
			children: [
				/* @__PURE__ */ (0, react_jsx_runtime.jsx)("img", {
					src: logoUrl,
					alt: agentName,
					className: "openui-shell-sidebar-header__logo"
				}),
				/* @__PURE__ */ (0, react_jsx_runtime.jsx)("div", {
					className: "openui-shell-sidebar-header__agent-name",
					children: agentName
				}),
				/* @__PURE__ */ (0, react_jsx_runtime.jsx)(require_components_IconButton_index.IconButton, {
					icon: isSidebarOpen ? /* @__PURE__ */ (0, react_jsx_runtime.jsx)(lucide_react.ArrowLeftFromLine, { size: "1em" }) : /* @__PURE__ */ (0, react_jsx_runtime.jsx)(lucide_react.ArrowRightFromLine, { size: "1em" }),
					onClick: () => {
						setIsSidebarOpen(!isSidebarOpen);
					},
					size: "small",
					variant: "secondary",
					className: "openui-shell-sidebar-header__toggle-button"
				})
			]
		}), children]
	});
};
const SidebarContent = ({ children, className }) => {
	const { isSidebarOpen } = require_ArtifactOverlay.useShellStore((state) => ({ isSidebarOpen: state.isSidebarOpen }));
	return /* @__PURE__ */ (0, react_jsx_runtime.jsx)("div", {
		className: (0, clsx.default)("openui-shell-sidebar-content", className, { "openui-shell-sidebar-content--collapsed": !isSidebarOpen }),
		children
	});
};
const SidebarSeparator = () => {
	return /* @__PURE__ */ (0, react_jsx_runtime.jsx)("div", { className: "openui-shell-sidebar-separator" });
};
//#endregion
//#region src/components/Shell/ResizableSeparator.tsx
/**
* A draggable vertical separator for resizing panels.
* Used between chat and artifact panels in desktop mode.
*/
const ResizableSeparator = ({ onResize, onDragStart, onDragEnd, className }) => {
	const isDraggingRef = (0, react.useRef)(false);
	const onResizeRef = (0, react.useRef)(onResize);
	const onDragStartRef = (0, react.useRef)(onDragStart);
	const onDragEndRef = (0, react.useRef)(onDragEnd);
	(0, react.useEffect)(() => {
		onResizeRef.current = onResize;
		onDragStartRef.current = onDragStart;
		onDragEndRef.current = onDragEnd;
	}, [
		onResize,
		onDragStart,
		onDragEnd
	]);
	(0, react.useEffect)(() => {
		const handleMouseMove = (e) => {
			if (isDraggingRef.current) {
				e.preventDefault();
				onResizeRef.current(e.clientX);
			}
		};
		const handleMouseUp = () => {
			if (isDraggingRef.current) {
				isDraggingRef.current = false;
				onDragEndRef.current();
				document.body.style.cursor = "";
				document.body.style.userSelect = "";
			}
		};
		document.addEventListener("mousemove", handleMouseMove);
		document.addEventListener("mouseup", handleMouseUp);
		return () => {
			document.removeEventListener("mousemove", handleMouseMove);
			document.removeEventListener("mouseup", handleMouseUp);
		};
	}, []);
	const handleMouseDown = () => {
		isDraggingRef.current = true;
		onDragStartRef.current();
		document.body.style.cursor = "col-resize";
		document.body.style.userSelect = "none";
	};
	return /* @__PURE__ */ (0, react_jsx_runtime.jsx)("div", {
		className: (0, clsx.default)("openui-shell-resizable-separator", className),
		onMouseDown: handleMouseDown,
		children: /* @__PURE__ */ (0, react_jsx_runtime.jsx)("div", { className: "openui-shell-resizable-separator__handle" })
	});
};
//#endregion
//#region src/components/Shell/useArtifactResize.ts
const INITIAL_CHAT_WIDTH = 420;
const MIN_CHAT_WIDTH = 420;
const MAX_CHAT_WIDTH_RATIO = .8;
/**
* Custom hook to manage artifact panel resizing logic (desktop only).
* Handles:
* - Chat panel width constraints
* - Resize drag events
* - Sidebar state when artifact is active/inactive
*/
const useArtifactResize = ({ isArtifactActive, isMobile, setIsSidebarOpen }) => {
	const [isDragging, setIsDragging] = (0, react.useState)(false);
	const containerRef = (0, react.useRef)(null);
	const chatPanelRef = (0, react.useRef)(null);
	const artifactPanelRef = (0, react.useRef)(null);
	(0, react.useEffect)(() => {
		if (isMobile) return;
		if (isArtifactActive) {
			setIsSidebarOpen(false);
			if (chatPanelRef.current) chatPanelRef.current.style.width = `${INITIAL_CHAT_WIDTH}px`;
		} else {
			setIsSidebarOpen(true);
			if (chatPanelRef.current) chatPanelRef.current.style.width = "100%";
		}
	}, [
		isArtifactActive,
		isMobile,
		setIsSidebarOpen
	]);
	return {
		containerRef,
		chatPanelRef,
		artifactPanelRef,
		isDragging,
		handleResize: (0, react.useCallback)((clientX) => {
			if (!containerRef.current || !chatPanelRef.current) return;
			const containerRect = containerRef.current.getBoundingClientRect();
			const newWidthPx = clientX - containerRect.left;
			const maxWidthPx = containerRect.width * MAX_CHAT_WIDTH_RATIO;
			const constrainedWidth = Math.min(Math.max(newWidthPx, MIN_CHAT_WIDTH), maxWidthPx);
			chatPanelRef.current.style.width = `${constrainedWidth}px`;
		}, []),
		handleDragStart: (0, react.useCallback)(() => {
			setIsDragging(true);
		}, []),
		handleDragEnd: (0, react.useCallback)(() => {
			setIsDragging(false);
		}, [])
	};
};
//#endregion
//#region src/components/Shell/Thread.tsx
const ThreadContainer = ({ children, className }) => {
	const { layout } = require_LayoutContext.useLayoutContext();
	const isMobile = layout === "mobile";
	const { isArtifactActive } = (0, _openuidev_react_headless.useActiveArtifact)();
	const { setIsSidebarOpen } = require_ArtifactOverlay.useShellStore((state) => ({ setIsSidebarOpen: state.setIsSidebarOpen }));
	const isLoadingMessages = (0, _openuidev_react_headless.useThread)((s) => s.isLoadingMessages);
	const { containerRef, chatPanelRef, artifactPanelRef, isDragging, handleResize, handleDragStart, handleDragEnd } = useArtifactResize({
		isArtifactActive,
		isMobile,
		setIsSidebarOpen
	});
	return /* @__PURE__ */ (0, react_jsx_runtime.jsx)("div", {
		className: (0, clsx.default)("openui-shell-thread-container", className, { "openui-shell-thread-container--artifact-active": isArtifactActive }),
		style: { visibility: isLoadingMessages ? "hidden" : void 0 },
		children: /* @__PURE__ */ (0, react_jsx_runtime.jsxs)("div", {
			className: "openui-shell-thread-wrapper",
			ref: containerRef,
			children: [/* @__PURE__ */ (0, react_jsx_runtime.jsxs)("div", {
				ref: chatPanelRef,
				className: (0, clsx.default)("openui-shell-thread-chat-panel", { "openui-shell-thread-chat-panel--animating": !isDragging }),
				children: [children, isMobile && /* @__PURE__ */ (0, react_jsx_runtime.jsx)(require_ArtifactOverlay.ArtifactOverlay, {})]
			}), !isMobile && isArtifactActive && /* @__PURE__ */ (0, react_jsx_runtime.jsxs)(react_jsx_runtime.Fragment, { children: [/* @__PURE__ */ (0, react_jsx_runtime.jsx)(ResizableSeparator, {
				onResize: handleResize,
				onDragStart: handleDragStart,
				onDragEnd: handleDragEnd
			}), /* @__PURE__ */ (0, react_jsx_runtime.jsx)("div", {
				ref: artifactPanelRef,
				className: (0, clsx.default)("openui-shell-thread-artifact-panel", { "openui-shell-thread-artifact-panel--animating": !isDragging }),
				children: /* @__PURE__ */ (0, react_jsx_runtime.jsx)(require_ArtifactOverlay.ArtifactPortalTarget, {})
			})] })]
		})
	});
};
const ScrollArea = ({ children, className, scrollVariant = "user-message-anchor", userMessageSelector }) => {
	const ref = (0, react.useRef)(null);
	const messages = (0, _openuidev_react_headless.useThread)((s) => s.messages);
	const isRunning = (0, _openuidev_react_headless.useThread)((s) => s.isRunning);
	const isLoadingMessages = (0, _openuidev_react_headless.useThread)((s) => s.isLoadingMessages);
	require_useComposerState.useScrollToBottom({
		ref,
		lastMessage: messages[messages.length - 1] || { id: "" },
		scrollVariant,
		userMessageSelector,
		isRunning,
		isLoadingMessages
	});
	return /* @__PURE__ */ (0, react_jsx_runtime.jsx)("div", {
		className: "openui-shell-thread-scroll-container",
		children: /* @__PURE__ */ (0, react_jsx_runtime.jsx)("div", {
			ref,
			className: (0, clsx.default)("openui-shell-thread-scroll-area", { "openui-shell-thread-scroll-area--user-message-anchor": scrollVariant === "user-message-anchor" }, className),
			children
		})
	});
};
const AssistantMessageContainer = ({ children, className }) => {
	const { logoUrl } = require_ArtifactOverlay.useShellStore((store) => ({ logoUrl: store.logoUrl }));
	return /* @__PURE__ */ (0, react_jsx_runtime.jsxs)("div", {
		className: (0, clsx.default)("openui-shell-thread-message-assistant", className),
		children: [/* @__PURE__ */ (0, react_jsx_runtime.jsx)("img", {
			src: logoUrl,
			alt: "Assistant",
			className: "openui-shell-thread-message-assistant__logo"
		}), /* @__PURE__ */ (0, react_jsx_runtime.jsx)("div", {
			className: "openui-shell-thread-message-assistant__content",
			children
		})]
	});
};
const UserMessageContainer = ({ children, className }) => {
	return /* @__PURE__ */ (0, react_jsx_runtime.jsx)("div", {
		className: (0, clsx.default)("openui-shell-thread-message-user", className),
		children: /* @__PURE__ */ (0, react_jsx_runtime.jsx)("div", {
			className: "openui-shell-thread-message-user__content",
			children
		})
	});
};
const AssistantMessageContent = ({ message, allMessages }) => {
	const getToolName = (toolCallId) => {
		return (message.toolCalls?.find((tc) => tc.id === toolCallId))?.function.name;
	};
	const toolMessages = [];
	const msgIndex = allMessages.findIndex((m) => m.id === message.id);
	if (msgIndex !== -1) for (let i = msgIndex + 1; i < allMessages.length; i++) {
		const m = allMessages[i];
		if (m && m.role === "tool") toolMessages.push(m);
		else break;
	}
	return /* @__PURE__ */ (0, react_jsx_runtime.jsxs)(react_jsx_runtime.Fragment, { children: [
		message.content && /* @__PURE__ */ (0, react_jsx_runtime.jsx)(require_components_MarkDownRenderer_index.MarkDownRenderer, {
			textMarkdown: message.content,
			className: "openui-shell-thread-message-assistant__text"
		}),
		message.toolCalls?.map((toolCall) => /* @__PURE__ */ (0, react_jsx_runtime.jsx)(require_ToolCall.ToolCallComponent, { toolCall }, toolCall.id)),
		toolMessages.map((tm) => /* @__PURE__ */ (0, react_jsx_runtime.jsx)(require_ToolResult.ToolResult, {
			message: tm,
			toolName: getToolName(tm.toolCallId)
		}, tm.id))
	] });
};
const UserMessageContent = ({ message }) => {
	if (message.role !== "user") return null;
	const content = message.content;
	if (typeof content === "string") {
		const { content: humanText } = separateContentAndContext(content);
		return /* @__PURE__ */ (0, react_jsx_runtime.jsx)(react_jsx_runtime.Fragment, { children: humanText });
	}
	return /* @__PURE__ */ (0, react_jsx_runtime.jsx)(react_jsx_runtime.Fragment, { children: content?.map((part, i) => {
		if (part.type === "text") return /* @__PURE__ */ (0, react_jsx_runtime.jsx)("span", { children: part.text }, i);
		if (part.type === "binary" && part.url) return /* @__PURE__ */ (0, react_jsx_runtime.jsx)("img", {
			src: part.url,
			alt: "",
			className: "openui-shell-thread-message-user__image"
		}, i);
		return null;
	}) });
};
const RenderMessage = (0, react.memo)(({ message, className, allMessages, assistantMessage: CustomAssistantMessage, userMessage: CustomUserMessage, isStreaming }) => {
	if (message.role === "tool") return null;
	if (message.role === "assistant") {
		if (CustomAssistantMessage) return /* @__PURE__ */ (0, react_jsx_runtime.jsx)(CustomAssistantMessage, {
			message,
			isStreaming
		});
		return /* @__PURE__ */ (0, react_jsx_runtime.jsx)(AssistantMessageContainer, {
			className,
			children: /* @__PURE__ */ (0, react_jsx_runtime.jsx)(AssistantMessageContent, {
				message,
				allMessages
			})
		});
	}
	if (message.role === "user") {
		if (CustomUserMessage) return /* @__PURE__ */ (0, react_jsx_runtime.jsx)(CustomUserMessage, { message });
		return /* @__PURE__ */ (0, react_jsx_runtime.jsx)(UserMessageContainer, {
			className,
			children: /* @__PURE__ */ (0, react_jsx_runtime.jsx)(UserMessageContent, { message })
		});
	}
	return null;
});
const MessageLoading = () => {
	return /* @__PURE__ */ (0, react_jsx_runtime.jsx)("div", {
		className: "openui-shell-thread-message-loading",
		children: /* @__PURE__ */ (0, react_jsx_runtime.jsx)(require_components_MessageLoading_index.MessageLoading, {})
	});
};
const ThreadError = () => {
	const threadError = (0, _openuidev_react_headless.useThread)((s) => s.threadError);
	if (!threadError) return null;
	return /* @__PURE__ */ (0, react_jsx_runtime.jsx)("div", {
		className: "openui-shell-thread-error",
		children: /* @__PURE__ */ (0, react_jsx_runtime.jsx)(require_components_Callout_index.Callout, {
			variant: "danger",
			title: "Something went wrong",
			description: threadError.message || "An unexpected error occurred. Please try again."
		})
	});
};
const Messages = ({ className, loader, assistantMessage, userMessage }) => {
	const messages = (0, _openuidev_react_headless.useThread)((s) => s.messages);
	const isRunning = (0, _openuidev_react_headless.useThread)((s) => s.isRunning);
	const threadError = (0, _openuidev_react_headless.useThread)((s) => s.threadError);
	return /* @__PURE__ */ (0, react_jsx_runtime.jsxs)("div", {
		className: (0, clsx.default)("openui-shell-thread-messages", className),
		children: [
			messages.map((message, i) => {
				return /* @__PURE__ */ (0, react_jsx_runtime.jsx)(_openuidev_react_headless.MessageProvider, {
					message,
					children: /* @__PURE__ */ (0, react_jsx_runtime.jsx)(RenderMessage, {
						message,
						allMessages: messages,
						assistantMessage,
						userMessage,
						isStreaming: isRunning && i === messages.length - 1
					})
				}, message.id);
			}),
			isRunning && /* @__PURE__ */ (0, react_jsx_runtime.jsx)("div", { children: loader }),
			!isRunning && threadError && /* @__PURE__ */ (0, react_jsx_runtime.jsx)(ThreadError, {})
		]
	});
};
const ThreadHeader = ({ children, className }) => {
	return /* @__PURE__ */ (0, react_jsx_runtime.jsx)("div", {
		className: (0, clsx.default)("openui-shell-thread-header", className),
		children
	});
};
//#endregion
//#region src/components/Shell/ThreadList.tsx
const ThreadButton = ({ id, title, className }) => {
	const selectThread = (0, _openuidev_react_headless.useThreadList)((s) => s.selectThread);
	const deleteThread = (0, _openuidev_react_headless.useThreadList)((s) => s.deleteThread);
	const selectedThreadId = (0, _openuidev_react_headless.useThreadList)((s) => s.selectedThreadId);
	const { isSidebarOpen, setIsSidebarOpen } = require_ArtifactOverlay.useShellStore((state) => ({
		isSidebarOpen: state.isSidebarOpen,
		setIsSidebarOpen: state.setIsSidebarOpen
	}));
	const { layout } = require_LayoutContext.useLayoutContext();
	return /* @__PURE__ */ (0, react_jsx_runtime.jsxs)("div", {
		className: (0, clsx.default)("openui-shell-thread-button", { "openui-shell-thread-button--selected": selectedThreadId === id }, className),
		children: [/* @__PURE__ */ (0, react_jsx_runtime.jsx)("button", {
			className: "openui-shell-thread-button-title",
			onClick: () => {
				if (layout === "mobile") setIsSidebarOpen(!isSidebarOpen);
				selectThread(id);
			},
			children: title
		}), /* @__PURE__ */ (0, react_jsx_runtime.jsxs)(_radix_ui_react_dropdown_menu.Root, { children: [/* @__PURE__ */ (0, react_jsx_runtime.jsx)(_radix_ui_react_dropdown_menu.Trigger, {
			asChild: true,
			children: /* @__PURE__ */ (0, react_jsx_runtime.jsx)("button", {
				className: "openui-shell-thread-button-dropdown-trigger",
				children: /* @__PURE__ */ (0, react_jsx_runtime.jsx)(lucide_react.EllipsisVerticalIcon, { size: 14 })
			})
		}), /* @__PURE__ */ (0, react_jsx_runtime.jsx)(_radix_ui_react_dropdown_menu.Portal, { children: /* @__PURE__ */ (0, react_jsx_runtime.jsx)(_radix_ui_react_dropdown_menu.Content, {
			className: "openui-shell-thread-button-dropdown-menu",
			side: "bottom",
			align: "start",
			sideOffset: 2,
			children: /* @__PURE__ */ (0, react_jsx_runtime.jsxs)(_radix_ui_react_dropdown_menu.Item, {
				className: "openui-shell-thread-button-dropdown-menu-item",
				onSelect: () => {
					deleteThread(id);
				},
				children: [/* @__PURE__ */ (0, react_jsx_runtime.jsx)(lucide_react.Trash2Icon, {
					size: 14,
					className: "openui-shell-thread-button-dropdown-menu-item-icon"
				}), "Delete"]
			})
		}) })] })]
	});
};
const ThreadList = ({ className }) => {
	const threads = (0, _openuidev_react_headless.useThreadList)((s) => s.threads);
	const loadThreads = (0, _openuidev_react_headless.useThreadList)((s) => s.loadThreads);
	(0, react.useEffect)(() => {
		loadThreads();
	}, []);
	const groupThreads = () => {
		const now = /* @__PURE__ */ new Date();
		const today = new Date(now.getFullYear(), now.getMonth(), now.getDate());
		const yesterday = new Date(today);
		yesterday.setDate(yesterday.getDate() - 1);
		const last7Days = new Date(today);
		last7Days.setDate(last7Days.getDate() - 7);
		const last30Days = new Date(today);
		last30Days.setDate(last30Days.getDate() - 30);
		const thisYear = new Date(today);
		thisYear.setMonth(0, 1);
		return threads.reduce((groups, thread) => {
			const threadDate = new Date(thread.createdAt);
			if (threadDate >= today) groups.today = [...groups.today || [], thread];
			else if (threadDate >= yesterday) groups.yesterday = [...groups.yesterday || [], thread];
			else if (threadDate >= last7Days) groups.last7Days = [...groups.last7Days || [], thread];
			else if (threadDate >= last30Days) groups.last30Days = [...groups.last30Days || [], thread];
			else if (threadDate >= thisYear) groups.thisYear = [...groups.thisYear || [], thread];
			else groups.older = [...groups.older || [], thread];
			return groups;
		}, {
			today: [],
			yesterday: [],
			last7Days: [],
			last30Days: [],
			thisYear: [],
			older: []
		});
	};
	const groupedThreads = groupThreads();
	const groupLabels = {
		today: "Today",
		yesterday: "Yesterday",
		last7Days: "Previous 7 Days",
		last30Days: "Previous 30 Days",
		thisYear: "This Year",
		older: "Older"
	};
	return /* @__PURE__ */ (0, react_jsx_runtime.jsx)("div", {
		className: (0, clsx.default)("openui-shell-thread-list", className),
		children: Object.entries(groupedThreads).filter(([_, groupThreads]) => groupThreads.length > 0).map(([group, groupThreads]) => /* @__PURE__ */ (0, react_jsx_runtime.jsxs)(react.Fragment, { children: [/* @__PURE__ */ (0, react_jsx_runtime.jsx)("div", {
			className: "openui-shell-thread-list-group",
			children: groupLabels[group]
		}), groupThreads.map((thread) => /* @__PURE__ */ (0, react_jsx_runtime.jsx)(ThreadButton, {
			id: thread.id,
			title: thread.title
		}, thread.id))] }, group))
	});
};
//#endregion
//#region src/components/Shell/WelcomeScreen.tsx
/**
* Type guard to check if image is a URL object
*/
const isImageUrl = (image) => {
	return typeof image === "object" && image !== null && "url" in image;
};
const WelcomeScreen = (props) => {
	const { className, starters = [], starterVariant = "long" } = props;
	const messages = (0, _openuidev_react_headless.useThread)((s) => s.messages);
	if (!require_ArtifactOverlay.isChatEmpty({
		isLoadingMessages: (0, _openuidev_react_headless.useThread)((s) => s.isLoadingMessages),
		messages
	})) return null;
	if ("children" in props && props.children) return /* @__PURE__ */ (0, react_jsx_runtime.jsx)("div", {
		className: (0, clsx.default)("openui-shell-welcome-screen", className),
		children: props.children
	});
	const { title, description, image } = props;
	const renderImage = () => {
		if (!image) return null;
		if (isImageUrl(image)) return /* @__PURE__ */ (0, react_jsx_runtime.jsx)("img", {
			src: image.url,
			alt: title || "",
			className: "openui-shell-welcome-screen__image"
		});
		return image;
	};
	return /* @__PURE__ */ (0, react_jsx_runtime.jsxs)("div", {
		className: (0, clsx.default)("openui-shell-welcome-screen", "openui-shell-welcome-screen--with-composer", className),
		children: [/* @__PURE__ */ (0, react_jsx_runtime.jsxs)("div", {
			className: "openui-shell-welcome-screen__header",
			children: [image && /* @__PURE__ */ (0, react_jsx_runtime.jsx)("div", {
				className: "openui-shell-welcome-screen__image-container",
				children: renderImage()
			}), (title || description) && /* @__PURE__ */ (0, react_jsx_runtime.jsxs)("div", {
				className: "openui-shell-welcome-screen__content",
				children: [title && /* @__PURE__ */ (0, react_jsx_runtime.jsx)("h2", {
					className: "openui-shell-welcome-screen__title",
					children: title
				}), description && /* @__PURE__ */ (0, react_jsx_runtime.jsx)("p", {
					className: "openui-shell-welcome-screen__description",
					children: description
				})]
			})]
		}), /* @__PURE__ */ (0, react_jsx_runtime.jsxs)("div", {
			className: "openui-shell-welcome-screen__composer-starters-container",
			children: [/* @__PURE__ */ (0, react_jsx_runtime.jsx)("div", {
				className: "openui-shell-welcome-screen__desktop-composer",
				children: /* @__PURE__ */ (0, react_jsx_runtime.jsx)(DesktopWelcomeComposer, {})
			}), starters.length > 0 && /* @__PURE__ */ (0, react_jsx_runtime.jsx)("div", {
				className: "openui-shell-welcome-screen__desktop-starters",
				children: /* @__PURE__ */ (0, react_jsx_runtime.jsx)(ConversationStarter, {
					starters,
					variant: starterVariant
				})
			})]
		})]
	});
};
//#endregion
Object.defineProperty(exports, "AssistantMessageContainer", {
	enumerable: true,
	get: function() {
		return AssistantMessageContainer;
	}
});
Object.defineProperty(exports, "Composer", {
	enumerable: true,
	get: function() {
		return Composer;
	}
});
Object.defineProperty(exports, "Container", {
	enumerable: true,
	get: function() {
		return Container;
	}
});
Object.defineProperty(exports, "ConversationStarter", {
	enumerable: true,
	get: function() {
		return ConversationStarter;
	}
});
Object.defineProperty(exports, "DesktopWelcomeComposer", {
	enumerable: true,
	get: function() {
		return DesktopWelcomeComposer;
	}
});
Object.defineProperty(exports, "MessageLoading", {
	enumerable: true,
	get: function() {
		return MessageLoading;
	}
});
Object.defineProperty(exports, "Messages", {
	enumerable: true,
	get: function() {
		return Messages;
	}
});
Object.defineProperty(exports, "MobileHeader", {
	enumerable: true,
	get: function() {
		return MobileHeader;
	}
});
Object.defineProperty(exports, "NewChatButton", {
	enumerable: true,
	get: function() {
		return NewChatButton;
	}
});
Object.defineProperty(exports, "RenderMessage", {
	enumerable: true,
	get: function() {
		return RenderMessage;
	}
});
Object.defineProperty(exports, "ScrollArea", {
	enumerable: true,
	get: function() {
		return ScrollArea;
	}
});
Object.defineProperty(exports, "SidebarContainer", {
	enumerable: true,
	get: function() {
		return SidebarContainer;
	}
});
Object.defineProperty(exports, "SidebarContent", {
	enumerable: true,
	get: function() {
		return SidebarContent;
	}
});
Object.defineProperty(exports, "SidebarHeader", {
	enumerable: true,
	get: function() {
		return SidebarHeader;
	}
});
Object.defineProperty(exports, "SidebarSeparator", {
	enumerable: true,
	get: function() {
		return SidebarSeparator;
	}
});
Object.defineProperty(exports, "ThreadButton", {
	enumerable: true,
	get: function() {
		return ThreadButton;
	}
});
Object.defineProperty(exports, "ThreadContainer", {
	enumerable: true,
	get: function() {
		return ThreadContainer;
	}
});
Object.defineProperty(exports, "ThreadError", {
	enumerable: true,
	get: function() {
		return ThreadError;
	}
});
Object.defineProperty(exports, "ThreadHeader", {
	enumerable: true,
	get: function() {
		return ThreadHeader;
	}
});
Object.defineProperty(exports, "ThreadList", {
	enumerable: true,
	get: function() {
		return ThreadList;
	}
});
Object.defineProperty(exports, "UserMessageContainer", {
	enumerable: true,
	get: function() {
		return UserMessageContainer;
	}
});
Object.defineProperty(exports, "WelcomeScreen", {
	enumerable: true,
	get: function() {
		return WelcomeScreen;
	}
});
Object.defineProperty(exports, "separateContentAndContext", {
	enumerable: true,
	get: function() {
		return separateContentAndContext;
	}
});
Object.defineProperty(exports, "wrapContent", {
	enumerable: true,
	get: function() {
		return wrapContent;
	}
});
Object.defineProperty(exports, "wrapContext", {
	enumerable: true,
	get: function() {
		return wrapContext;
	}
});
