// @openuidev/react-ui/dist/ToolCall-BXicKrtH.cjs
const require_chunk = require("./chunk-CKQMccvm.cjs");
let clsx = require("clsx");
clsx = require_chunk.__toESM(clsx);
let lucide_react = require("lucide-react");
let react = require("react");
let react_jsx_runtime = require("react/jsx-runtime");
//#region src/components/ToolCall/BehindTheScenes.tsx
const BehindTheScenes = ({ isStreaming, toolCallsComplete, children }) => {
	const [userOverride, setUserOverride] = (0, react.useState)(null);
	const hasCompletedOnce = (0, react.useRef)(false);
	const prevStreaming = (0, react.useRef)(isStreaming);
	(0, react.useEffect)(() => {
		if (isStreaming && !prevStreaming.current) {
			setUserOverride(null);
			hasCompletedOnce.current = false;
		}
		prevStreaming.current = isStreaming;
	}, [isStreaming]);
	if (toolCallsComplete && !hasCompletedOnce.current) hasCompletedOnce.current = true;
	const toolsActive = !!isStreaming && !hasCompletedOnce.current;
	const isExpanded = userOverride !== null ? userOverride : toolsActive;
	const toggle = () => {
		setUserOverride((prev) => prev !== null ? !prev : !isExpanded);
	};
	return /* @__PURE__ */ (0, react_jsx_runtime.jsxs)("div", {
		className: "openui-behind-the-scenes",
		children: [/* @__PURE__ */ (0, react_jsx_runtime.jsxs)("button", {
			className: "openui-behind-the-scenes__toggle",
			onClick: toggle,
			type: "button",
			children: [isExpanded ? /* @__PURE__ */ (0, react_jsx_runtime.jsx)(lucide_react.ChevronUp, {
				size: 14,
				className: "openui-behind-the-scenes__toggle-icon"
			}) : /* @__PURE__ */ (0, react_jsx_runtime.jsx)(lucide_react.ChevronDown, {
				size: 14,
				className: "openui-behind-the-scenes__toggle-icon"
			}), toolsActive ? "Working..." : "Behind the scenes"]
		}), isExpanded && /* @__PURE__ */ (0, react_jsx_runtime.jsx)("div", {
			className: "openui-behind-the-scenes__items",
			children
		})]
	});
};
//#endregion
//#region src/components/ToolCall/ToolCall.tsx
const ToolCallComponent = ({ toolCall, isStreaming, toolsDone, isLast = false, className }) => {
	const isRunning = !!isStreaming && !toolsDone;
	const actionLabel = isRunning ? `Calling the ${toolCall.function.name} tool` : `Called the ${toolCall.function.name} tool`;
	let parsedArgs = null;
	try {
		parsedArgs = JSON.parse(toolCall.function.arguments);
	} catch {}
	const hasRequest = parsedArgs && parsedArgs._request != null;
	const hasResponse = parsedArgs && parsedArgs._response != null;
	const requestStr = hasRequest ? JSON.stringify(parsedArgs._request, null, 2) : null;
	const responseStr = hasResponse ? JSON.stringify(parsedArgs._response, null, 2) : null;
	const plainArgs = !hasRequest && !hasResponse && toolCall.function.arguments ? (() => {
		try {
			return JSON.stringify(JSON.parse(toolCall.function.arguments), null, 2);
		} catch {
			return toolCall.function.arguments;
		}
	})() : null;
	return /* @__PURE__ */ (0, react_jsx_runtime.jsxs)("div", {
		className: (0, clsx.default)("openui-tool-call", className),
		children: [/* @__PURE__ */ (0, react_jsx_runtime.jsxs)("div", {
			className: "openui-tool-call__title-row",
			children: [/* @__PURE__ */ (0, react_jsx_runtime.jsx)("span", {
				className: (0, clsx.default)("openui-tool-call__icon-wrapper", { "openui-tool-call__icon--blinking": isRunning && isLast }),
				children: /* @__PURE__ */ (0, react_jsx_runtime.jsx)(lucide_react.SquareCode, {
					size: 14,
					className: "openui-tool-call__icon"
				})
			}), /* @__PURE__ */ (0, react_jsx_runtime.jsx)("span", {
				className: (0, clsx.default)("openui-tool-call__name", { "openui-tool-call__name--shimmer": isRunning && isLast }),
				children: actionLabel
			})]
		}), /* @__PURE__ */ (0, react_jsx_runtime.jsx)("div", {
			className: (0, clsx.default)("openui-tool-call__connector", { "openui-tool-call__connector--last": isLast }),
			children: /* @__PURE__ */ (0, react_jsx_runtime.jsxs)("div", {
				className: "openui-tool-call__args-block",
				children: [
					requestStr && /* @__PURE__ */ (0, react_jsx_runtime.jsx)(ToolCodeBlock, {
						type: "request",
						code: requestStr,
						isRunning: isRunning && !hasResponse,
						toolName: toolCall.function.name
					}),
					responseStr && /* @__PURE__ */ (0, react_jsx_runtime.jsx)(ToolCodeBlock, {
						type: "response",
						code: responseStr,
						isRunning: isRunning && isLast,
						toolName: toolCall.function.name
					}),
					plainArgs && /* @__PURE__ */ (0, react_jsx_runtime.jsx)(ToolCodeBlock, {
						type: "request",
						code: plainArgs,
						isRunning,
						toolName: toolCall.function.name
					})
				]
			})
		})]
	});
};
const ToolCodeBlock = ({ type, code, isRunning = false, toolName }) => {
	const [isExpanded, setIsExpanded] = (0, react.useState)(false);
	const label = type === "request" ? "Tool Request" : "Tool Response";
	const runningLabel = type === "request" ? `Sending request to ${toolName}...` : `Awaiting response from ${toolName}...`;
	return /* @__PURE__ */ (0, react_jsx_runtime.jsxs)("div", {
		className: "openui-tool-code-block",
		children: [/* @__PURE__ */ (0, react_jsx_runtime.jsxs)("button", {
			className: "openui-tool-code-block__header",
			onClick: () => setIsExpanded((v) => !v),
			type: "button",
			children: [/* @__PURE__ */ (0, react_jsx_runtime.jsx)("span", {
				className: (0, clsx.default)("openui-tool-code-block__label", { "openui-tool-code-block__label--loading": isRunning }),
				children: isRunning ? runningLabel : label
			}), /* @__PURE__ */ (0, react_jsx_runtime.jsx)(lucide_react.ChevronDown, {
				size: 14,
				className: (0, clsx.default)("openui-tool-code-block__chevron", { "openui-tool-code-block__chevron--expanded": isExpanded })
			})]
		}), isExpanded && /* @__PURE__ */ (0, react_jsx_runtime.jsx)("div", {
			className: "openui-tool-code-block__content",
			children: /* @__PURE__ */ (0, react_jsx_runtime.jsx)("pre", {
				className: "openui-tool-code-block__code",
				children: code
			})
		})]
	});
};
//#endregion
Object.defineProperty(exports, "BehindTheScenes", {
	enumerable: true,
	get: function() {
		return BehindTheScenes;
	}
});
Object.defineProperty(exports, "ToolCallComponent", {
	enumerable: true,
	get: function() {
		return ToolCallComponent;
	}
});
