// @openuidev/react-ui/dist/components/Button/stories/Button.stories.js
import { jsx as _jsx } from "react/jsx-runtime";
import { ArrowRight, Download } from "lucide-react";
import { Button } from "../Button";
// Basic button stories
export const Primary = {
    args: {
        children: "Primary Button",
        variant: "primary",
        size: "medium",
        disabled: false,
        buttonType: "normal",
    },
    render: (args) => (_jsx(Button, { variant: args.variant, size: args.size, disabled: args.disabled, buttonType: args.buttonType, children: args.children })),
};
export const Secondary = {
    args: {
        children: "Secondary",
        variant: "secondary",
        buttonType: "normal",
    },
};
export const Tertiary = {
    name: "Tertiary (Ghost)",
    args: {
        children: "Tertiary",
        variant: "tertiary",
        buttonType: "normal",
    },
};
// With icons
export const WithLeftIcon = {
    args: {
        children: "Download",
        variant: "primary",
        iconLeft: _jsx(Download, { size: 18 }),
    },
    render: (args) => (_jsx(Button, { variant: args.variant, size: args.size, disabled: args.disabled, iconLeft: args.iconLeft, children: args.children })),
};
export const WithRightIcon = {
    args: {
        children: "Next",
        variant: "primary",
        iconRight: _jsx(ArrowRight, { size: 18 }),
    },
    render: (args) => (_jsx(Button, { variant: args.variant, size: args.size, disabled: args.disabled, iconRight: args.iconRight, children: args.children })),
};
const meta = {
    title: "Components/Button",
    component: Button,
    parameters: {
        layout: "centered",
        docs: {
            description: {
                component: "```tsx\nimport { Button } from '@openui-ui/react-ui';\n```",
            },
        },
    },
    argTypes: {
        variant: {
            control: "radio",
            options: ["primary", "secondary", "tertiary"],
            table: {
                category: "Appearance",
            },
        },
        size: {
            control: "radio",
            options: ["extra-small", "small", "medium", "large"],
            table: {
                category: "Appearance",
            },
        },
        disabled: {
            control: "boolean",
            table: {
                category: "State",
            },
        },
        iconLeft: {
            control: false,
            description: "Any react icon component",
            table: {
                category: "Icons",
            },
        },
        iconRight: {
            control: false,
            description: "Any react icon component",
            table: {
                category: "Icons",
            },
        },
        buttonType: {
            control: "radio",
            options: ["normal", "destructive"],
            table: {
                category: "Appearance",
            },
        },
    },
    tags: ["autodocs", "!dev"],
};
export default meta;
