// @openuidev/react-ui/dist/genui-lib/Button/index.js
"use client";
import { jsx as _jsx } from "react/jsx-runtime";
import { ACTION_STEPS, defineComponent, useFormName, useFormValidation, useIsStreaming, useTriggerAction, } from "@openuidev/react-lang";
import { Button as OpenUIButton } from "../../components/Button";
import { ButtonSchema } from "./schema";
export { ButtonSchema } from "./schema";
const variantMap = {
    primary: "primary",
    secondary: "secondary",
    ghost: "tertiary",
    tertiary: "tertiary",
};
export const Button = defineComponent({
    name: "Button",
    props: ButtonSchema,
    description: "Clickable button",
    component: ({ props }) => {
        const triggerAction = useTriggerAction();
        const formName = useFormName();
        const isStreaming = useIsStreaming();
        const formValidation = useFormValidation();
        const label = props.label;
        return (_jsx(OpenUIButton, { variant: variantMap[props.variant] || "primary", size: props.size || "medium", buttonType: props.type, disabled: isStreaming, onClick: () => {
                const action = props.action;
                const variant = props.variant || "primary";
                // Validate form for primary buttons before firing action
                if (formValidation && variant === "primary") {
                    if (action?.steps) {
                        // v0.5 ActionPlan — validate if any step is ToAssistant or mutation
                        const needsValidation = action.steps.some((s) => s.type === ACTION_STEPS.ToAssistant ||
                            (s.type === ACTION_STEPS.Run && s.refType === "mutation"));
                        if (needsValidation && !formValidation.validateForm())
                            return;
                    }
                    else {
                        // v0.1 legacy or no action — always validate for primary buttons
                        if (!formValidation.validateForm())
                            return;
                    }
                }
                triggerAction(label, formName, action);
            }, children: label }));
    },
});
