// @openuidev/react-headless/dist/index.cjs
Object.defineProperty(exports, Symbol.toStringTag, { value: "Module" });
let react = require("react");
let zustand = require("zustand");
let zustand_react_shallow = require("zustand/react/shallow");
let react_jsx_runtime = require("react/jsx-runtime");
let zustand_middleware = require("zustand/middleware");
let _ag_ui_core = require("@ag-ui/core");
//#region src/store/ArtifactContext.ts
/** @internal React context holding the artifact Zustand store. Provided by `ChatProvider`. */
const ArtifactContext = (0, react.createContext)(null);
/**
* Returns the raw artifact Zustand store for advanced use cases.
*
* Prefer {@link useArtifact} or {@link useActiveArtifact} for most cases —
* this hook is an escape hatch when you need direct store access.
*
* @category Hooks
* @returns The Zustand `StoreApi<ArtifactStore>` instance
* @throws Error if called outside a `<ChatProvider>`
*/
const useArtifactStore = () => {
	const store = (0, react.useContext)(ArtifactContext);
	if (!store) throw new Error("useArtifactStore must be used within a <ChatProvider>");
	return store;
};
//#endregion
//#region src/hooks/useActiveArtifact.ts
/**
* Returns global artifact activation state — whether *any* artifact is open,
* and a close action that dismisses it.
*
* Use this in layout components that react to artifact presence (resizing panels,
* showing overlays) without needing to know *which* artifact is active.
* For per-artifact state and actions, use {@link useArtifact} instead.
*
* Must be called within a `<ChatProvider>`.
*
* @category Hooks
* @returns {@link UseActiveArtifactReturn}
*/
function useActiveArtifact() {
	const store = useArtifactStore();
	const activeArtifactId = (0, zustand.useStore)(store, (s) => s.activeArtifactId);
	return {
		isArtifactActive: activeArtifactId !== null,
		activeArtifactId,
		closeArtifact: (0, react.useCallback)(() => {
			const state = store.getState();
			if (state.activeArtifactId) state.closeArtifact(state.activeArtifactId);
		}, [store])
	};
}
//#endregion
//#region src/hooks/useArtifact.ts
/**
* Binds a component to a specific artifact by ID, providing activation state
* and actions (open, close, toggle).
*
* Multiple `useArtifact` hooks with different IDs can coexist —
* only one artifact is active at a time.
*
* Must be called within a `<ChatProvider>`.
*
* @category Hooks
* @param artifactId - Unique identifier for the artifact
* @returns {@link UseArtifactReturn}
*
* @example
* ```tsx
* function PreviewButton({ id }: { id: string }) {
*   const { isActive, toggle } = useArtifact(id);
*   return (
*     <button onClick={() => toggle()}>
*       {isActive ? "Hide" : "Show"} Preview
*     </button>
*   );
* }
* ```
*/
function useArtifact(artifactId) {
	const store = useArtifactStore();
	return {
		isActive: (0, zustand.useStore)(store, (s) => s.activeArtifactId === artifactId),
		open: (0, react.useCallback)(() => {
			store.getState().openArtifact(artifactId);
		}, [store, artifactId]),
		close: (0, react.useCallback)(() => {
			store.getState().closeArtifact(artifactId);
		}, [store, artifactId]),
		toggle: (0, react.useCallback)(() => {
			const state = store.getState();
			if (state.activeArtifactId === artifactId) state.closeArtifact(artifactId);
			else state.openArtifact(artifactId);
		}, [store, artifactId])
	};
}
//#endregion
//#region src/hooks/useArtifactPortalTarget.ts
/**
* Provides access to the artifact portal target DOM node.
*
* This hook serves two roles:
* - **Registering a portal target:** Call `setNode` from a ref callback to
*   designate a DOM element as the render target for artifact content.
*   Only one target should be registered at a time.
* - **Reading the portal target:** Read `node` to get the current target
*   element for use with `createPortal()`.
*
* Must be called within a `<ChatProvider>`.
*
* @category Hooks
* @returns `{ setNode, node }` — setter for registration, getter for portal rendering
*
* @example
* ```tsx
* // Registering a portal target
* function MyPortalTarget() {
*   const { setNode } = useArtifactPortalTarget();
*   return <div ref={setNode} />;
* }
*
* // Building a custom artifact panel
* function MyArtifactPanel({ artifactId, children }) {
*   const { isActive } = useArtifact(artifactId);
*   const { node } = useArtifactPortalTarget();
*   if (!isActive || !node) return null;
*   return createPortal(<div>{children}</div>, node);
* }
* ```
*/
function useArtifactPortalTarget() {
	const store = useArtifactStore();
	const node = (0, zustand.useStore)(store, (s) => s._artifactPanelNode);
	return {
		setNode: (0, react.useCallback)((node) => {
			store.getState()._setArtifactPanelNode(node);
		}, [store]),
		node
	};
}
//#endregion
//#region src/hooks/useMessage.tsx
/**
* @category Contexts
*/
const MessageContext = (0, react.createContext)(null);
/**
* @category Hooks
* @returns The current message. See {@link Message} for more information.
*/
const useMessage = () => {
	const context = (0, react.useContext)(MessageContext);
	if (!context) throw new Error("useMessage must be used within a MessageProvider");
	return context;
};
/**
* @category Components
*/
const MessageProvider = ({ message, children }) => {
	const ctxValue = (0, zustand_react_shallow.useShallow)((_s) => ({ message }))();
	return /* @__PURE__ */ (0, react_jsx_runtime.jsx)(MessageContext.Provider, {
		value: ctxValue,
		children
	});
};
//#endregion
//#region src/store/ChatContext.ts
const ChatContext = (0, react.createContext)(null);
const useChatStore = () => {
	const store = (0, react.useContext)(ChatContext);
	if (!store) throw new Error("useChatStore must be used within a <ChatProvider>");
	return store;
};
//#endregion
//#region src/hooks/useThread.ts
const threadSelector = (s) => ({
	messages: s.messages,
	isRunning: s.isRunning,
	isLoadingMessages: s.isLoadingMessages,
	threadError: s.threadError,
	processMessage: s.processMessage,
	appendMessages: s.appendMessages,
	updateMessage: s.updateMessage,
	setMessages: s.setMessages,
	deleteMessage: s.deleteMessage,
	cancelMessage: s.cancelMessage
});
const threadListSelector = (s) => ({
	threads: s.threads,
	isLoadingThreads: s.isLoadingThreads,
	threadListError: s.threadListError,
	selectedThreadId: s.selectedThreadId,
	hasMoreThreads: s.hasMoreThreads,
	loadThreads: s.loadThreads,
	loadMoreThreads: s.loadMoreThreads,
	switchToNewThread: s.switchToNewThread,
	createThread: s.createThread,
	selectThread: s.selectThread,
	updateThread: s.updateThread,
	deleteThread: s.deleteThread
});
function useThread(selector) {
	const store = useChatStore();
	if (selector) return (0, zustand.useStore)(store, (s) => selector(threadSelector(s)));
	return (0, zustand.useStore)(store, (0, zustand_react_shallow.useShallow)(threadSelector));
}
function useThreadList(selector) {
	const store = useChatStore();
	if (selector) return (0, zustand.useStore)(store, (s) => selector(threadListSelector(s)));
	return (0, zustand.useStore)(store, (0, zustand_react_shallow.useShallow)(threadListSelector));
}
//#endregion
//#region src/store/createArtifactStore.ts
/**
* Creates a Zustand store managing artifact state.
* Instantiated once by `ChatProvider` — consumers should not call this directly.
*
* @internal
*/
const createArtifactStore = () => {
	return (0, zustand.createStore)()((0, zustand_middleware.subscribeWithSelector)((set, get) => ({
		activeArtifactId: null,
		openArtifact: (id) => {
			set({ activeArtifactId: id });
		},
		closeArtifact: (id) => {
			if (get().activeArtifactId === id) set({ activeArtifactId: null });
		},
		resetArtifacts: () => {
			set({ activeArtifactId: null });
		},
		_artifactPanelNode: null,
		_setArtifactPanelNode: (node) => {
			if (process.env["NODE_ENV"] !== "production" && node && get()._artifactPanelNode && get()._artifactPanelNode !== node) console.warn("[OpenUI] Multiple ArtifactPortalTarget instances detected. Only one should be mounted at a time.");
			set({ _artifactPanelNode: node });
		}
	})));
};
//#endregion
//#region src/types/messageFormat.ts
/**
* Default identity message format — no conversion.
* Messages are sent and received as-is in AG-UI format.
*/
const identityMessageFormat = {
	toApi: (messages) => messages,
	fromApi: (data) => data
};
//#endregion
//#region src/stream/adapters/ag-ui.ts
const agUIAdapter = () => ({ async *parse(response) {
	const reader = response.body?.getReader();
	if (!reader) throw new Error("No response body");
	const decoder = new TextDecoder();
	while (true) {
		const { done, value } = await reader.read();
		if (done) break;
		const lines = decoder.decode(value, { stream: true }).split("\n");
		for (const line of lines) {
			if (!line.startsWith("data: ")) continue;
			const data = line.slice(6).trim();
			if (!data || data === "[DONE]") continue;
			try {
				yield JSON.parse(data);
			} catch (e) {
				console.error("Failed to parse SSE event", e);
			}
		}
	}
} });
//#endregion
//#region src/stream/adapters/langgraph.ts
/**
* Adapter for LangGraph streaming responses.
*
* LangGraph uses named SSE events (`event: <type>\ndata: <json>\n\n`)
* rather than the `data:`-only format used by OpenAI. The adapter handles
* the `messages`, `metadata`, `updates`, and `error` event types and maps
* them to AG-UI events.
*
* Usage:
* ```tsx
* <ChatProvider
*   apiUrl="/api/langgraph"
*   streamProtocol={langGraphAdapter()}
* />
* ```
*/
const langGraphAdapter = (options) => ({ async *parse(response) {
	const reader = response.body?.getReader();
	if (!reader) throw new Error("No response body");
	const decoder = new TextDecoder();
	const messageId = crypto.randomUUID();
	const toolCallIds = {};
	let messageStarted = false;
	let buffer = "";
	while (true) {
		const { done, value } = await reader.read();
		if (done) break;
		buffer += decoder.decode(value, { stream: true });
		const blocks = buffer.split("\n\n");
		buffer = blocks.pop() ?? "";
		for (const block of blocks) {
			const trimmed = block.trim();
			if (!trimmed) continue;
			const { event, data } = parseSSEBlock(trimmed);
			if (!data) continue;
			let parsed;
			try {
				parsed = JSON.parse(data);
			} catch (e) {
				console.error("Failed to parse LangGraph SSE data", e);
				continue;
			}
			switch (event) {
				case "metadata": break;
				case "messages": {
					const tuple = parsed;
					const msg = Array.isArray(tuple) ? tuple[0] : tuple;
					if (msg.type !== "ai" && msg.type !== "AIMessageChunk" && msg.type !== "AIMessage") break;
					if (!messageStarted) {
						yield {
							type: _ag_ui_core.EventType.TEXT_MESSAGE_START,
							messageId,
							role: "assistant"
						};
						messageStarted = true;
					}
					const textContent = extractTextContent$1(msg.content);
					if (textContent) yield {
						type: _ag_ui_core.EventType.TEXT_MESSAGE_CONTENT,
						messageId,
						delta: textContent
					};
					if (msg.tool_call_chunks) for (const chunk of msg.tool_call_chunks) {
						const index = chunk.index ?? 0;
						if (chunk.id && !toolCallIds[index]) {
							toolCallIds[index] = chunk.id;
							yield {
								type: _ag_ui_core.EventType.TOOL_CALL_START,
								toolCallId: chunk.id,
								toolCallName: chunk.name || ""
							};
						}
						if (chunk.args) {
							const toolCallId = toolCallIds[index];
							if (toolCallId) yield {
								type: _ag_ui_core.EventType.TOOL_CALL_ARGS,
								toolCallId,
								delta: chunk.args
							};
						}
					}
					if (msg.tool_calls && msg.tool_calls.length > 0) for (let i = 0; i < msg.tool_calls.length; i++) {
						const tc = msg.tool_calls[i];
						if (!tc) continue;
						const toolCallId = tc.id || crypto.randomUUID();
						if (!Object.values(toolCallIds).includes(toolCallId)) {
							yield {
								type: _ag_ui_core.EventType.TOOL_CALL_START,
								toolCallId,
								toolCallName: tc.name
							};
							const argsStr = typeof tc.args === "string" ? tc.args : JSON.stringify(tc.args);
							yield {
								type: _ag_ui_core.EventType.TOOL_CALL_ARGS,
								toolCallId,
								delta: argsStr
							};
							yield {
								type: _ag_ui_core.EventType.TOOL_CALL_END,
								toolCallId
							};
						}
					}
					break;
				}
				case "updates": {
					const updates = parsed;
					if ("__interrupt__" in updates && options?.onInterrupt) options.onInterrupt(updates["__interrupt__"]);
					break;
				}
				case "error": {
					const err = parsed;
					yield {
						type: _ag_ui_core.EventType.RUN_ERROR,
						message: err.message || err.error || "Unknown error",
						code: err.error ?? void 0
					};
					break;
				}
				case "end":
					if (messageStarted) {
						for (const toolCallId of Object.values(toolCallIds)) yield {
							type: _ag_ui_core.EventType.TOOL_CALL_END,
							toolCallId
						};
						yield {
							type: _ag_ui_core.EventType.TEXT_MESSAGE_END,
							messageId
						};
						messageStarted = false;
					}
					break;
				default: break;
			}
		}
	}
	if (messageStarted) {
		for (const toolCallId of Object.values(toolCallIds)) yield {
			type: _ag_ui_core.EventType.TOOL_CALL_END,
			toolCallId
		};
		yield {
			type: _ag_ui_core.EventType.TEXT_MESSAGE_END,
			messageId
		};
	}
} });
/**
* Parse an SSE block into its event name and data payload.
*
* LangGraph SSE format:
* ```
* event: messages
* data: [{"content": "Hello", ...}, {"langgraph_node": "agent"}]
* ```
*/
function parseSSEBlock(block) {
	let event = "";
	const dataLines = [];
	for (const line of block.split("\n")) if (line.startsWith("event:")) event = line.slice(6).trim();
	else if (line.startsWith("data:")) dataLines.push(line.slice(5).trim());
	return {
		event,
		data: dataLines.join("\n")
	};
}
/**
* Extract text content from a LangGraph message content field.
* Content can be a plain string or an array of typed content blocks.
*/
function extractTextContent$1(content) {
	if (typeof content === "string") return content;
	return content.filter((block) => block.type === "text" && block.text).map((block) => block.text).join("");
}
//#endregion
//#region src/stream/adapters/openai-completions.ts
const openAIAdapter = () => ({ async *parse(response) {
	const reader = response.body?.getReader();
	if (!reader) throw new Error("No response body");
	const decoder = new TextDecoder();
	const messageId = crypto.randomUUID();
	const toolCallIds = {};
	let messageStarted = false;
	while (true) {
		const { done, value } = await reader.read();
		if (done) break;
		const lines = decoder.decode(value, { stream: true }).split("\n");
		for (const line of lines) {
			if (!line.startsWith("data: ")) continue;
			const data = line.slice(6).trim();
			if (!data || data === "[DONE]") continue;
			try {
				const choice = JSON.parse(data).choices?.[0];
				const delta = choice?.delta;
				if (!delta) continue;
				if (!messageStarted && (delta.content || delta.role)) {
					yield {
						type: _ag_ui_core.EventType.TEXT_MESSAGE_START,
						messageId,
						role: "assistant"
					};
					messageStarted = true;
				}
				if (delta.content) yield {
					type: _ag_ui_core.EventType.TEXT_MESSAGE_CONTENT,
					messageId,
					delta: delta.content
				};
				if (delta.tool_calls) for (const toolCall of delta.tool_calls) {
					const index = toolCall.index;
					if (toolCall.id) {
						toolCallIds[index] = toolCall.id;
						yield {
							type: _ag_ui_core.EventType.TOOL_CALL_START,
							toolCallId: toolCall.id,
							toolCallName: toolCall.function?.name || ""
						};
					}
					if (toolCall.function?.arguments) {
						const toolCallId = toolCallIds[index];
						if (toolCallId) yield {
							type: _ag_ui_core.EventType.TOOL_CALL_ARGS,
							toolCallId,
							delta: toolCall.function.arguments
						};
					}
				}
				if (choice?.finish_reason === "stop") yield {
					type: _ag_ui_core.EventType.TEXT_MESSAGE_END,
					messageId
				};
				else if (choice?.finish_reason === "tool_calls") for (const toolCallId of Object.values(toolCallIds)) yield {
					type: _ag_ui_core.EventType.TOOL_CALL_END,
					toolCallId
				};
			} catch (e) {
				console.error("Failed to parse OpenAI SSE event", e);
			}
		}
	}
} });
//#endregion
//#region src/stream/adapters/openai-readable-stream.ts
/**
* Adapter for streams produced by the OpenAI SDK's `Stream.toReadableStream()`.
* That method emits NDJSON (one JSON object per line, no `data: ` SSE prefix),
* which differs from the raw SSE format that `openAIAdapter` expects.
*/
const openAIReadableStreamAdapter = () => ({ async *parse(response) {
	const reader = response.body?.getReader();
	if (!reader) throw new Error("No response body");
	const decoder = new TextDecoder();
	const messageId = crypto.randomUUID();
	const toolCallIds = {};
	let messageStarted = false;
	let buffer = "";
	while (true) {
		const { done, value } = await reader.read();
		if (done) break;
		buffer += decoder.decode(value, { stream: true });
		const lines = buffer.split("\n");
		buffer = lines.pop() ?? "";
		for (const line of lines) {
			const data = line.trim();
			if (!data) continue;
			try {
				const choice = JSON.parse(data).choices?.[0];
				const delta = choice?.delta;
				if (!delta) continue;
				if (!messageStarted && (delta.content || delta.role)) {
					yield {
						type: _ag_ui_core.EventType.TEXT_MESSAGE_START,
						messageId,
						role: "assistant"
					};
					messageStarted = true;
				}
				if (delta.content) yield {
					type: _ag_ui_core.EventType.TEXT_MESSAGE_CONTENT,
					messageId,
					delta: delta.content
				};
				if (delta.tool_calls) for (const toolCall of delta.tool_calls) {
					const index = toolCall.index;
					if (toolCall.id) {
						toolCallIds[index] = toolCall.id;
						yield {
							type: _ag_ui_core.EventType.TOOL_CALL_START,
							toolCallId: toolCall.id,
							toolCallName: toolCall.function?.name || ""
						};
					}
					if (toolCall.function?.arguments) {
						const toolCallId = toolCallIds[index];
						if (toolCallId) yield {
							type: _ag_ui_core.EventType.TOOL_CALL_ARGS,
							toolCallId,
							delta: toolCall.function.arguments
						};
					}
				}
				if (choice?.finish_reason === "stop") yield {
					type: _ag_ui_core.EventType.TEXT_MESSAGE_END,
					messageId
				};
				else if (choice?.finish_reason === "tool_calls") for (const toolCallId of Object.values(toolCallIds)) yield {
					type: _ag_ui_core.EventType.TOOL_CALL_END,
					toolCallId
				};
			} catch (e) {
				console.error("Failed to parse OpenAI NDJSON chunk", e);
			}
		}
	}
} });
//#endregion
//#region src/stream/adapters/openai-responses.ts
const openAIResponsesAdapter = () => ({ async *parse(response) {
	const reader = response.body?.getReader();
	if (!reader) throw new Error("No response body");
	const decoder = new TextDecoder();
	const itemIdToCallId = {};
	while (true) {
		const { done, value } = await reader.read();
		if (done) break;
		const lines = decoder.decode(value, { stream: true }).split("\n");
		for (const line of lines) {
			if (!line.startsWith("data: ")) continue;
			const data = line.slice(6).trim();
			if (!data || data === "[DONE]") continue;
			try {
				const event = JSON.parse(data);
				switch (event.type) {
					case "response.output_item.added": {
						const item = event.item;
						if (item.type === "message" && item.role === "assistant") yield {
							type: _ag_ui_core.EventType.TEXT_MESSAGE_START,
							messageId: item.id,
							role: "assistant"
						};
						else if (item.type === "function_call") {
							itemIdToCallId[item.id ?? item.call_id] = item.call_id;
							yield {
								type: _ag_ui_core.EventType.TOOL_CALL_START,
								toolCallId: item.call_id,
								toolCallName: item.name
							};
						}
						break;
					}
					case "response.output_text.delta":
						yield {
							type: _ag_ui_core.EventType.TEXT_MESSAGE_CONTENT,
							messageId: event.item_id,
							delta: event.delta
						};
						break;
					case "response.output_text.done":
						yield {
							type: _ag_ui_core.EventType.TEXT_MESSAGE_END,
							messageId: event.item_id
						};
						break;
					case "response.function_call_arguments.delta": {
						const callId = itemIdToCallId[event.item_id] ?? event.item_id;
						yield {
							type: _ag_ui_core.EventType.TOOL_CALL_ARGS,
							toolCallId: callId,
							delta: event.delta
						};
						break;
					}
					case "response.function_call_arguments.done": {
						const callId = itemIdToCallId[event.item_id] ?? event.item_id;
						yield {
							type: _ag_ui_core.EventType.TOOL_CALL_END,
							toolCallId: callId
						};
						break;
					}
					case "error":
						yield {
							type: _ag_ui_core.EventType.RUN_ERROR,
							message: event.message,
							code: event.code ?? void 0
						};
						break;
					case "response.failed":
						yield {
							type: _ag_ui_core.EventType.RUN_ERROR,
							message: event.response?.error?.message ?? "Response failed",
							code: event.response?.error?.code ?? void 0
						};
						break;
					default: break;
				}
			} catch (e) {
				console.error("Failed to parse OpenAI Responses SSE event", e);
			}
		}
	}
} });
//#endregion
//#region src/stream/processStreamedMessage.ts
/**
* @category Utilities
*/
const processStreamedMessage = async ({ response, createMessage, updateMessage, deleteMessage, adapter = agUIAdapter() }) => {
	let currentMessage = {
		id: crypto.randomUUID(),
		role: "assistant",
		content: "",
		toolCalls: []
	};
	let isFirst = true;
	let rafId = null;
	const debouncedUpdate = (msg) => {
		if (rafId !== null) cancelAnimationFrame(rafId);
		rafId = requestAnimationFrame(() => {
			updateMessage(msg);
			rafId = null;
		});
	};
	for await (const event of adapter.parse(response)) {
		switch (event.type) {
			case _ag_ui_core.EventType.TEXT_MESSAGE_CHUNK:
			case _ag_ui_core.EventType.TEXT_MESSAGE_CONTENT:
				currentMessage = {
					...currentMessage,
					content: (currentMessage.content || "") + event.delta
				};
				break;
			case _ag_ui_core.EventType.TOOL_CALL_START:
				currentMessage = {
					...currentMessage,
					toolCalls: [...currentMessage.toolCalls || [], {
						id: event.toolCallId,
						type: "function",
						function: {
							name: event.toolCallName,
							arguments: ""
						}
					}]
				};
				break;
			case _ag_ui_core.EventType.TOOL_CALL_ARGS:
				if (currentMessage.toolCalls) {
					const toolCalls = [...currentMessage.toolCalls];
					const toolCallIndex = toolCalls.findIndex((tc) => tc.id === event.toolCallId);
					if (toolCallIndex !== -1) {
						const currentToolCall = toolCalls[toolCallIndex];
						if (currentToolCall) {
							toolCalls[toolCallIndex] = {
								id: currentToolCall.id,
								type: "function",
								function: {
									name: currentToolCall.function.name,
									arguments: currentToolCall.function.arguments + event.delta
								}
							};
							currentMessage = {
								...currentMessage,
								toolCalls
							};
						}
					}
				}
				break;
			case _ag_ui_core.EventType.TEXT_MESSAGE_START:
				if (event.messageId !== currentMessage.id) {
					deleteMessage(currentMessage.id);
					currentMessage = {
						...currentMessage,
						id: event.messageId
					};
					isFirst = true;
				}
				break;
			case _ag_ui_core.EventType.RUN_ERROR: {
				const msg = event.message || event.error || "Stream error";
				throw new Error(typeof msg === "string" ? msg : JSON.stringify(msg));
			}
		}
		if (isFirst) {
			createMessage(currentMessage);
			isFirst = false;
		} else debouncedUpdate(currentMessage);
	}
	if (rafId !== null) {
		cancelAnimationFrame(rafId);
		updateMessage(currentMessage);
	}
	return currentMessage;
};
//#endregion
//#region src/store/createChatStore.ts
const mergeThreadList = (existing, incoming) => Array.from(new Map([...existing, ...incoming].map((t) => [t.id, t])).values()).sort((a, b) => new Date(b.createdAt).getTime() - new Date(a.createdAt).getTime());
const createChatStore = (config) => {
	const { apiUrl, threadApiUrl, processMessage: userProcessMessage, fetchThreadList: userFetchThreadList, createThread: userCreateThread, deleteThread: userDeleteThread, updateThread: userUpdateThread, loadThread: userLoadThread, streamProtocol, messageFormat = identityMessageFormat } = config;
	const fetchThreadList = userFetchThreadList ?? (async (cursor) => {
		if (!threadApiUrl) return { threads: [] };
		const url = cursor ? `${threadApiUrl}/get?cursor=${cursor}` : `${threadApiUrl}/get`;
		return (await fetch(url)).json();
	});
	const createThread = userCreateThread ?? (async (firstMessage) => {
		if (!threadApiUrl) throw new Error("threadApiUrl or createThread required");
		return (await fetch(`${threadApiUrl}/create`, {
			method: "POST",
			headers: { "Content-Type": "application/json" },
			body: JSON.stringify({ messages: messageFormat.toApi([firstMessage]) })
		})).json();
	});
	const deleteThreadFn = userDeleteThread ?? (async (id) => {
		if (!threadApiUrl) return;
		await fetch(`${threadApiUrl}/delete/${id}`, { method: "DELETE" });
	});
	const updateThreadFn = userUpdateThread ?? (async (updated) => {
		if (!threadApiUrl) return updated;
		return (await fetch(`${threadApiUrl}/update/${updated.id}`, {
			method: "PATCH",
			headers: { "Content-Type": "application/json" },
			body: JSON.stringify(updated)
		})).json();
	});
	const loadThread = userLoadThread ?? (async (threadId) => {
		if (!threadApiUrl) return [];
		const raw = await (await fetch(`${threadApiUrl}/get/${threadId}`)).json();
		return messageFormat.fromApi(raw);
	});
	const sendMessage = userProcessMessage ?? (async ({ threadId, messages, abortController }) => {
		if (!apiUrl) throw new Error("apiUrl or processMessage required");
		return fetch(apiUrl, {
			method: "POST",
			headers: { "Content-Type": "application/json" },
			body: JSON.stringify({
				threadId,
				messages: messageFormat.toApi(messages)
			}),
			signal: abortController.signal
		});
	});
	return (0, zustand.createStore)()((0, zustand_middleware.subscribeWithSelector)((set, get) => ({
		threads: [],
		isLoadingThreads: false,
		threadListError: null,
		selectedThreadId: null,
		hasMoreThreads: false,
		_nextCursor: void 0,
		messages: [],
		isRunning: false,
		isLoadingMessages: false,
		threadError: null,
		_abortController: null,
		loadThreads: () => {
			set({
				isLoadingThreads: true,
				threadListError: null
			});
			fetchThreadList(void 0).then(({ threads = [], nextCursor }) => {
				set({
					threads,
					isLoadingThreads: false,
					_nextCursor: nextCursor,
					hasMoreThreads: nextCursor !== void 0
				});
			}).catch((e) => {
				set({
					isLoadingThreads: false,
					threadListError: e
				});
			});
		},
		loadMoreThreads: () => {
			const cursor = get()._nextCursor;
			if (cursor === void 0) return;
			fetchThreadList(cursor).then(({ threads = [], nextCursor }) => {
				set((s) => ({
					threads: mergeThreadList(s.threads, threads),
					_nextCursor: nextCursor,
					hasMoreThreads: nextCursor !== void 0
				}));
			}).catch((e) => {
				set({ threadListError: e });
			});
		},
		switchToNewThread: () => {
			get().cancelMessage();
			set({
				selectedThreadId: null,
				messages: [],
				threadError: null
			});
		},
		createThread: async (firstMessage) => {
			const thread = await createThread(firstMessage);
			set((s) => ({ threads: mergeThreadList(s.threads, [thread]) }));
			return thread;
		},
		selectThread: (threadId) => {
			get().cancelMessage();
			set({
				selectedThreadId: threadId,
				messages: [],
				isLoadingMessages: true,
				threadError: null
			});
			loadThread(threadId).then((messages) => set({
				messages,
				isLoadingMessages: false
			})).catch((e) => set({
				threadError: e,
				isLoadingMessages: false
			}));
		},
		updateThread: (thread) => {
			const setPending = (id, isPending) => set((s) => ({ threads: s.threads.map((t) => t.id === id ? {
				...t,
				isPending
			} : t) }));
			setPending(thread.id, true);
			updateThreadFn(thread).then((updated) => {
				set((s) => ({ threads: s.threads.map((t) => t.id === updated.id ? updated : t) }));
			}).catch(() => setPending(thread.id, false));
		},
		deleteThread: (threadId) => {
			const setPending = (id, isPending) => set((s) => ({ threads: s.threads.map((t) => t.id === id ? {
				...t,
				isPending
			} : t) }));
			setPending(threadId, true);
			deleteThreadFn(threadId).then(() => {
				const state = get();
				set({ threads: state.threads.filter((t) => t.id !== threadId) });
				if (state.selectedThreadId === threadId) state.switchToNewThread();
			}).catch(() => setPending(threadId, false));
		},
		processMessage: async (message) => {
			if (get().isRunning) return;
			const abortController = new AbortController();
			const optimisticMessage = {
				...message,
				id: crypto.randomUUID(),
				role: "user"
			};
			set({
				_abortController: abortController,
				isRunning: true,
				threadError: null
			});
			set((s) => ({ messages: [...s.messages, optimisticMessage] }));
			abortController.signal.addEventListener("abort", () => {
				set({
					_abortController: null,
					isRunning: false
				});
			});
			try {
				let threadId = get().selectedThreadId;
				if (!threadId) if (userCreateThread || threadApiUrl) {
					threadId = (await get().createThread(optimisticMessage)).id;
					set({ selectedThreadId: threadId });
				} else threadId = "ephemeral";
				const response = await sendMessage({
					threadId,
					messages: get().messages,
					abortController
				});
				if (response instanceof Response && !response.ok) throw new Error(`Request failed: ${response.status} ${response.statusText}`);
				await processStreamedMessage({
					response,
					createMessage: (msg) => set((s) => ({ messages: [...s.messages, msg] })),
					updateMessage: (msg) => set((s) => ({ messages: s.messages.map((m) => m.id === msg.id ? msg : m) })),
					deleteMessage: (id) => set((s) => ({ messages: s.messages.filter((m) => m.id !== id) })),
					adapter: streamProtocol
				});
			} catch (e) {
				if (!abortController.signal.aborted) set({ threadError: e instanceof Error ? e : new Error(String(e)) });
			} finally {
				set({
					_abortController: null,
					isRunning: false
				});
			}
		},
		appendMessages: (...newMessages) => {
			set((s) => ({ messages: [...s.messages, ...newMessages] }));
		},
		updateMessage: (message) => {
			set((s) => ({ messages: s.messages.map((m) => m.id === message.id ? message : m) }));
		},
		setMessages: (messages) => {
			set({ messages });
		},
		deleteMessage: (messageId) => {
			set((s) => ({ messages: s.messages.filter((m) => m.id !== messageId) }));
		},
		cancelMessage: () => {
			get()._abortController?.abort();
		}
	})));
};
//#endregion
//#region src/store/ChatProvider.tsx
const ChatProvider = ({ children, ...config }) => {
	const [chatStore] = (0, react.useState)(() => createChatStore(config));
	const [artifactStore] = (0, react.useState)(() => createArtifactStore());
	(0, react.useEffect)(() => {
		return chatStore.subscribe((state) => state.selectedThreadId, () => artifactStore.getState().resetArtifacts());
	}, [chatStore, artifactStore]);
	return /* @__PURE__ */ (0, react_jsx_runtime.jsx)(ChatContext.Provider, {
		value: chatStore,
		children: /* @__PURE__ */ (0, react_jsx_runtime.jsx)(ArtifactContext.Provider, {
			value: artifactStore,
			children
		})
	});
};
//#endregion
//#region src/stream/formats/langgraph-message-format.ts
function toLangChainMessage(message) {
	switch (message.role) {
		case "user": return {
			type: "human",
			content: message.content ?? ""
		};
		case "assistant": {
			const result = {
				type: "ai",
				content: message.content ?? ""
			};
			if (message.toolCalls?.length) result.tool_calls = message.toolCalls.map((tc) => ({
				id: tc.id,
				name: tc.function.name,
				args: safeParseArgs(tc.function.arguments)
			}));
			return result;
		}
		case "tool": return {
			type: "tool",
			content: message.content,
			tool_call_id: message.toolCallId
		};
		case "system": return {
			type: "system",
			content: message.content
		};
		case "developer": return {
			type: "system",
			content: message.content
		};
		default: return {
			type: "system",
			content: ""
		};
	}
}
function fromLangChainMessage(msg) {
	const id = msg.id ?? crypto.randomUUID();
	switch (msg.type) {
		case "human": return {
			id,
			role: "user",
			content: extractContent(msg.content)
		};
		case "ai": {
			const result = {
				id,
				role: "assistant",
				content: extractContent(msg.content)
			};
			if (msg.tool_calls?.length) result.toolCalls = msg.tool_calls.map((tc) => ({
				id: tc.id,
				type: "function",
				function: {
					name: tc.name,
					arguments: typeof tc.args === "string" ? tc.args : JSON.stringify(tc.args)
				}
			}));
			return result;
		}
		case "tool": return {
			id,
			role: "tool",
			content: extractContent(msg.content),
			toolCallId: msg.tool_call_id ?? ""
		};
		case "system":
		case "developer": return {
			id,
			role: "system",
			content: extractContent(msg.content)
		};
		default: return {
			id,
			role: "system",
			content: extractContent(msg.content)
		};
	}
}
function extractContent(content) {
	if (typeof content === "string") return content;
	return content.filter((block) => block.type === "text" && block.text).map((block) => block.text).join("");
}
function safeParseArgs(args) {
	try {
		return JSON.parse(args);
	} catch {
		return args;
	}
}
/**
* Converts between AG-UI message format and LangGraph's LangChain-style
* message format.
*
* LangGraph uses `type` discriminators (`"human"`, `"ai"`, `"tool"`,
* `"system"`) instead of `role`, and tool call arguments are objects
* rather than JSON strings.
*
* AG-UI → LangGraph (toApi):
*   - Maps `role` to `type` (`"user"` → `"human"`, `"assistant"` → `"ai"`)
*   - Converts `toolCalls[].function.arguments` from JSON string to object
*   - Converts `toolCallId` → `tool_call_id`
*
* LangGraph → AG-UI (fromApi):
*   - Maps `type` to `role` (`"human"` → `"user"`, `"ai"` → `"assistant"`)
*   - Converts tool call `args` object to JSON string
*   - Generates `id` via `crypto.randomUUID()` if not present
*/
const langGraphMessageFormat = {
	toApi(messages) {
		return messages.map(toLangChainMessage);
	},
	fromApi(data) {
		return data.map(fromLangChainMessage);
	}
};
//#endregion
//#region src/stream/formats/openai-conversation-message-format.ts
/**
* Both the Responses API and the Conversations API accept
* `ResponseInputItem[]` as input, so a single outbound conversion
* works for both.
*
* Tool calls are *sibling items* of the assistant message, not nested
* inside it.  We flatten each AG-UI assistant message into:
*   1. An `EasyInputMessage` with `role: "assistant"` (text part)
*   2. One `ResponseFunctionToolCall` per tool call
*
* Tool result messages become `FunctionCallOutput` items.
*/
function toItems(message) {
	switch (message.role) {
		case "user": return [toInputMessage(message)];
		case "assistant": {
			const items = [];
			if (message.content) items.push({
				role: "assistant",
				content: message.content,
				type: "message"
			});
			if (message.toolCalls?.length) for (const tc of message.toolCalls) items.push({
				type: "function_call",
				call_id: tc.id,
				name: tc.function.name,
				arguments: tc.function.arguments
			});
			return items;
		}
		case "tool": return [{
			type: "function_call_output",
			call_id: message.toolCallId,
			output: message.content
		}];
		case "system": return [{
			role: "system",
			content: message.content,
			type: "message"
		}];
		case "developer": return [{
			role: "developer",
			content: message.content,
			type: "message"
		}];
		default: return [];
	}
}
function toInputMessage(message) {
	const content = message.content;
	if (typeof content === "string") return {
		role: "user",
		content,
		type: "message"
	};
	return {
		role: "user",
		content: content?.map((part) => {
			if (part.type === "text") return {
				type: "input_text",
				text: part.text
			};
			if (part.type === "binary") return {
				type: "input_image",
				image_url: part.url ?? `data:${part.mimeType};base64,${part.data ?? ""}`,
				detail: "auto"
			};
			return {
				type: "input_text",
				text: ""
			};
		}) ?? [],
		type: "message"
	};
}
/**
* Converts `ConversationItem[]` (from the Conversations API) or
* `ResponseItem[]` (from the Responses API) into AG-UI `Message[]`.
*
* Both APIs return items as a flat list with `type` discriminator:
*   - `"message"` → user / assistant / system / developer / tool message
*   - `"function_call"` → tool call (grouped into the preceding assistant)
*   - `"function_call_output"` → tool result
*
* The Conversations API `Message` type is a superset that also handles
* `reasoning_text`, `summary_text`, and `text` content parts, as well
* as additional roles like `tool` and `critic`.
*/
function fromItems(items) {
	const messages = [];
	let currentAssistant = null;
	for (const item of items) switch (item.type) {
		case "message": {
			if (currentAssistant) {
				messages.push(currentAssistant);
				currentAssistant = null;
			}
			const msg = item;
			if (msg.role === "assistant") currentAssistant = {
				id: msg.id,
				role: "assistant",
				content: extractTextContent(msg) || void 0
			};
			else if (msg.role === "tool") messages.push({
				id: msg.id,
				role: "tool",
				content: extractTextContent(msg),
				toolCallId: ""
			});
			else if (msg.role === "user") messages.push(fromUserMessage(msg));
			else {
				const role = msg.role === "developer" ? "developer" : "system";
				messages.push({
					id: msg.id,
					role,
					content: extractTextContent(msg)
				});
			}
			break;
		}
		case "function_call": {
			const tc = item;
			if (!currentAssistant) currentAssistant = {
				id: crypto.randomUUID(),
				role: "assistant"
			};
			currentAssistant = {
				...currentAssistant,
				toolCalls: [...currentAssistant.toolCalls ?? [], {
					id: tc.call_id,
					type: "function",
					function: {
						name: tc.name,
						arguments: tc.arguments
					}
				}]
			};
			break;
		}
		case "function_call_output": {
			if (currentAssistant) {
				messages.push(currentAssistant);
				currentAssistant = null;
			}
			const output = item;
			messages.push({
				id: output.id,
				role: "tool",
				content: typeof output.output === "string" ? output.output : JSON.stringify(output.output),
				toolCallId: output.call_id
			});
			break;
		}
		default: break;
	}
	if (currentAssistant) messages.push(currentAssistant);
	return messages;
}
/**
* Extract text content from a Conversations API `Message`.
*
* Handles the full content union:
*   - `output_text` (assistant output)
*   - `input_text` (user/system/developer input)
*   - `text` (generic text)
*   - `summary_text`, `reasoning_text` (appended)
*   - `refusal` (mapped to text)
*/
function extractTextContent(msg) {
	return msg.content.map((part) => {
		switch (part.type) {
			case "output_text": return part.text;
			case "input_text": return part.text;
			case "text": return part.text;
			case "summary_text": return part.text;
			case "reasoning_text": return part.text;
			case "refusal": return `[Refusal]: ${part.refusal}`;
			default: return "";
		}
	}).filter(Boolean).join("");
}
function fromUserMessage(msg) {
	if (!msg.content.some((part) => part.type === "input_image" || part.type === "input_file")) return {
		id: msg.id,
		role: "user",
		content: extractTextContent(msg)
	};
	const parts = msg.content.map((part) => {
		if (part.type === "input_text" || part.type === "text") return {
			type: "text",
			text: part.text
		};
		if (part.type === "input_image" && part.image_url) return {
			type: "binary",
			url: part.image_url,
			mimeType: "image/*"
		};
		return {
			type: "text",
			text: ""
		};
	});
	return {
		id: msg.id,
		role: "user",
		content: parts
	};
}
/**
* Converts between AG-UI message format and OpenAI's item-based format,
* compatible with both the **Responses API** and **Conversations API**.
*
* AG-UI → OpenAI (toApi):
*   - Returns `ResponseInputItem[]` — works for both `responses.create({ input })`
*     and `conversations.items.create({ items })`
*   - Flattens assistant messages: text → `EasyInputMessage`, tool calls → `ResponseFunctionToolCall`
*
* OpenAI → AG-UI (fromApi):
*   - Accepts `ConversationItem[]` (or `ResponseItem[]`, which is a subset)
*   - Groups adjacent assistant messages + function_calls into `AssistantMessage`
*   - Handles Conversations-specific content types (`reasoning_text`, `summary_text`, etc.)
*/
const openAIConversationMessageFormat = {
	toApi(messages) {
		return messages.flatMap(toItems);
	},
	fromApi(data) {
		return fromItems(data);
	}
};
//#endregion
//#region src/stream/formats/openai-message-format.ts
function toOpenAIUserMessage(message) {
	const content = message.content;
	if (typeof content === "string") return {
		role: "user",
		content
	};
	return {
		role: "user",
		content: content?.map((part) => {
			if (part.type === "text") return {
				type: "text",
				text: part.text
			};
			if (part.type === "binary") return {
				type: "image_url",
				image_url: { url: part.url ?? `data:${part.mimeType};base64,${part.data ?? ""}` }
			};
			return {
				type: "text",
				text: ""
			};
		}) ?? []
	};
}
function toOpenAIAssistantMessage(message) {
	const result = {
		role: "assistant",
		content: message.content ?? null
	};
	if (message.toolCalls?.length) result.tool_calls = message.toolCalls.map((tc) => ({
		id: tc.id,
		type: "function",
		function: {
			name: tc.function.name,
			arguments: tc.function.arguments
		}
	}));
	return result;
}
function toOpenAIToolMessage(message) {
	return {
		role: "tool",
		content: message.content,
		tool_call_id: message.toolCallId
	};
}
function toOpenAI(message) {
	switch (message.role) {
		case "user": return toOpenAIUserMessage(message);
		case "assistant": return toOpenAIAssistantMessage(message);
		case "tool": return toOpenAIToolMessage(message);
		case "system": return {
			role: "system",
			content: message.content
		};
		case "developer": return {
			role: "developer",
			content: message.content
		};
		default: return {
			role: "system",
			content: ""
		};
	}
}
function fromOpenAIAssistant(msg) {
	const content = typeof msg.content === "string" ? msg.content : void 0;
	const result = {
		id: crypto.randomUUID(),
		role: "assistant",
		content
	};
	if (msg.tool_calls?.length) result.toolCalls = msg.tool_calls.filter((tc) => tc.type === "function").map((tc) => ({
		id: tc.id,
		type: "function",
		function: {
			name: tc.function.name,
			arguments: tc.function.arguments
		}
	}));
	return result;
}
function fromOpenAIUser(msg) {
	if (typeof msg.content === "string") return {
		id: crypto.randomUUID(),
		role: "user",
		content: msg.content
	};
	const content = msg.content.map((part) => {
		if (part.type === "text") return {
			type: "text",
			text: part.text
		};
		return {
			type: "text",
			text: ""
		};
	});
	return {
		id: crypto.randomUUID(),
		role: "user",
		content
	};
}
function fromOpenAITool(msg) {
	const content = typeof msg.content === "string" ? msg.content : msg.content.map((p) => p.text).join("");
	return {
		id: crypto.randomUUID(),
		role: "tool",
		content,
		toolCallId: msg.tool_call_id
	};
}
function fromOpenAI(data) {
	switch (data.role) {
		case "user": return fromOpenAIUser(data);
		case "assistant": return fromOpenAIAssistant(data);
		case "tool": return fromOpenAITool(data);
		case "system": return {
			id: crypto.randomUUID(),
			role: "system",
			content: typeof data.content === "string" ? data.content : ""
		};
		case "developer": return {
			id: crypto.randomUUID(),
			role: "developer",
			content: typeof data.content === "string" ? data.content : ""
		};
		default: return {
			id: crypto.randomUUID(),
			role: "system",
			content: ""
		};
	}
}
/**
* Converts between AG-UI message format and OpenAI **Chat Completions**
* message format (`ChatCompletionMessageParam`).
*
* This is a 1-to-1 mapping — each AG-UI message becomes exactly one
* `ChatCompletionMessageParam` and vice versa.
*
* AG-UI → OpenAI (toApi):
*   - Strips `id` (OpenAI doesn't use message IDs)
*   - Converts `toolCalls` → `tool_calls`
*   - Converts `toolCallId` → `tool_call_id`
*   - Converts multipart `content` arrays to OpenAI content format
*
* OpenAI → AG-UI (fromApi):
*   - Generates `id` via `crypto.randomUUID()`
*   - Converts `tool_calls` → `toolCalls`
*   - Converts `tool_call_id` → `toolCallId`
*/
const openAIMessageFormat = {
	toApi(messages) {
		return messages.map(toOpenAI);
	},
	fromApi(data) {
		return data.map(fromOpenAI);
	}
};
//#endregion
exports.ArtifactContext = ArtifactContext;
exports.ChatProvider = ChatProvider;
Object.defineProperty(exports, "EventType", {
	enumerable: true,
	get: function() {
		return _ag_ui_core.EventType;
	}
});
exports.MessageContext = MessageContext;
exports.MessageProvider = MessageProvider;
exports.agUIAdapter = agUIAdapter;
exports.identityMessageFormat = identityMessageFormat;
exports.langGraphAdapter = langGraphAdapter;
exports.langGraphMessageFormat = langGraphMessageFormat;
exports.openAIAdapter = openAIAdapter;
exports.openAIConversationMessageFormat = openAIConversationMessageFormat;
exports.openAIMessageFormat = openAIMessageFormat;
exports.openAIReadableStreamAdapter = openAIReadableStreamAdapter;
exports.openAIResponsesAdapter = openAIResponsesAdapter;
exports.processStreamedMessage = processStreamedMessage;
exports.useActiveArtifact = useActiveArtifact;
exports.useArtifact = useArtifact;
exports.useArtifactPortalTarget = useArtifactPortalTarget;
exports.useArtifactStore = useArtifactStore;
exports.useMessage = useMessage;
exports.useThread = useThread;
exports.useThreadList = useThreadList;
