// @openuidev/react-ui/dist/genui-lib/Modal/index.js
"use client";
import { jsx as _jsx } from "react/jsx-runtime";
import { defineComponent, useStateField } from "@openuidev/react-lang";
import { Modal as OpenUIModal } from "../../components/Modal";
import { ModalSchema } from "./schema";
export { ModalSchema } from "./schema";
export const Modal = defineComponent({
    name: "Modal",
    props: ModalSchema,
    description: "Modal dialog. open is a reactive $boolean binding — set to true to open, X/Escape/backdrop auto-closes. Put Form with buttons inside children.",
    component: ({ props, renderNode }) => {
        const field = useStateField("open", props.open);
        const isOpen = field.value === true || field.value === "true";
        const handleOpenChange = (open) => {
            if (!open) {
                field.setValue(false);
            }
        };
        return (_jsx(OpenUIModal, { title: props.title, open: isOpen, onOpenChange: handleOpenChange, size: props.size ?? "md", children: renderNode(props.children) }));
    },
});
