// @openuidev/react-ui/dist/genui-lib/CheckBoxGroup/index.js
"use client";
import { jsx as _jsx } from "react/jsx-runtime";
import { defineComponent, parseStructuredRules, useFormValidation, useIsStreaming, useStateField, } from "@openuidev/react-lang";
import React from "react";
import { CheckBoxGroup as OpenUICheckBoxGroup } from "../../components/CheckBoxGroup";
import { CheckBoxItem as OpenUICheckBoxItem } from "../../components/CheckBoxItem";
import { CheckBoxItemSchema, createCheckBoxGroupSchema } from "./schema";
export { CheckBoxItemSchema } from "./schema";
export const CheckBoxItem = defineComponent({
    name: "CheckBoxItem",
    props: CheckBoxItemSchema,
    description: "",
    component: () => null,
});
export const CheckBoxGroup = defineComponent({
    name: "CheckBoxGroup",
    props: createCheckBoxGroupSchema(CheckBoxItem),
    description: "",
    component: ({ props }) => {
        const isStreaming = useIsStreaming();
        const formValidation = useFormValidation();
        const field = useStateField(props.name, props.value);
        const rules = React.useMemo(() => parseStructuredRules(props.rules), [props.rules]);
        const hasRules = rules.length > 0;
        const items = (props.items ?? []);
        // Aggregate: map of item name → checked boolean
        const getAggregate = React.useCallback(() => {
            const stored = field.value;
            const result = {};
            for (const item of items) {
                result[item.props.name] = stored?.[item.props.name] ?? item.props.defaultChecked ?? false;
            }
            return result;
        }, [field.value, items]);
        React.useEffect(() => {
            if (!isStreaming && hasRules && formValidation) {
                formValidation.registerField(field.name, rules, () => field.value);
                return () => formValidation.unregisterField(field.name);
            }
            return undefined;
        }, [field.name, field.value, formValidation, hasRules, isStreaming, rules]);
        if (!items.length)
            return null;
        const aggregate = getAggregate();
        return (_jsx(OpenUICheckBoxGroup, { children: items.map((item, i) => (_jsx(OpenUICheckBoxItem, { name: item.props.name, label: item.props.label, description: item.props.description || "", checked: aggregate[item.props.name] ?? item.props.defaultChecked ?? false, onChange: (val) => {
                    const newAggregate = { ...getAggregate(), [item.props.name]: val };
                    field.setValue(newAggregate);
                    if (hasRules) {
                        formValidation?.validateField(field.name, newAggregate, rules);
                    }
                }, disabled: isStreaming }, i))) }));
    },
});
