// @openuidev/react-ui/dist/components/Tabs/index.cjs
Object.defineProperty(exports, Symbol.toStringTag, { value: "Module" });
const require_chunk = require("../../chunk-CKQMccvm.cjs");
const require_components_IconButton_index = require("../IconButton/index.cjs");
let clsx = require("clsx");
clsx = require_chunk.__toESM(clsx);
let lucide_react = require("lucide-react");
let react = require("react");
react = require_chunk.__toESM(react);
let react_jsx_runtime = require("react/jsx-runtime");
let _radix_ui_react_tabs = require("@radix-ui/react-tabs");
_radix_ui_react_tabs = require_chunk.__toESM(_radix_ui_react_tabs);
//#region src/components/Tabs/Tabs.tsx
const tabsVariants = { clear: "openui-tabs-clear" };
const Tabs = (0, react.forwardRef)(({ className, style, variant = "clear", ...props }, ref) => /* @__PURE__ */ (0, react_jsx_runtime.jsx)(_radix_ui_react_tabs.Root, {
	ref,
	className: (0, clsx.default)("openui-tabs", tabsVariants[variant], className),
	style,
	...props
}));
Tabs.displayName = "Tabs";
const tabsListVariants = {
	title: "openui-tabs-list--title",
	iconTitle: "openui-tabs-list--icon-title",
	iconTitleSubtext: "openui-tabs-list--icon-title-subtext",
	imageTitle: "openui-tabs-list--image-title",
	imageTitleSubtext: "openui-tabs-list--image-title-subtext"
};
const TabsList = (0, react.forwardRef)(({ className, style, variant = "title", ...props }, ref) => {
	const listRef = (0, react.useRef)(null);
	const indicatorRef = (0, react.useRef)(null);
	const [showLeftButton, setShowLeftButton] = (0, react.useState)(false);
	const [showRightButton, setShowRightButton] = (0, react.useState)(false);
	const updateIndicator = (0, react.useCallback)((animated) => {
		const list = listRef.current;
		const indicator = indicatorRef.current;
		if (!list || !indicator) return;
		const activeTab = list.querySelector("[role=\"tab\"][data-state=\"active\"]");
		if (!activeTab) return;
		const left = activeTab.offsetLeft - list.scrollLeft;
		const width = activeTab.offsetWidth;
		indicator.style.transition = animated ? "transform 0.25s cubic-bezier(0.4, 0, 0.2, 1), width 0.25s cubic-bezier(0.4, 0, 0.2, 1)" : "none";
		indicator.style.width = `${width}px`;
		indicator.style.transform = `translateX(${left}px)`;
		indicator.style.opacity = "1";
	}, []);
	(0, react.useEffect)(() => {
		const checkScroll = () => {
			if (listRef.current) {
				const { scrollLeft, scrollWidth, clientWidth } = listRef.current;
				setShowLeftButton(scrollLeft > 0);
				setShowRightButton(scrollLeft < scrollWidth - clientWidth - 1);
			}
		};
		const handleScroll = () => {
			checkScroll();
			updateIndicator(false);
		};
		checkScroll();
		updateIndicator(false);
		const currentRef = listRef.current;
		if (currentRef) {
			currentRef.addEventListener("scroll", handleScroll);
			const resizeObserver = new ResizeObserver(() => {
				checkScroll();
				updateIndicator(false);
			});
			resizeObserver.observe(currentRef);
			return () => {
				currentRef.removeEventListener("scroll", handleScroll);
				resizeObserver.disconnect();
			};
		}
		return () => {};
	}, [updateIndicator]);
	(0, react.useEffect)(() => {
		const list = listRef.current;
		if (!list) return;
		const observer = new MutationObserver(() => updateIndicator(true));
		observer.observe(list, {
			attributes: true,
			subtree: true,
			attributeFilter: ["data-state"]
		});
		return () => observer.disconnect();
	}, [updateIndicator]);
	(0, react.useEffect)(() => {
		const list = listRef.current;
		if (!list) return;
		const handleClick = (e) => {
			const trigger = e.target.closest("[role=\"tab\"]");
			if (!trigger) return;
			const listWidth = list.clientWidth;
			const triggerLeft = trigger.offsetLeft;
			const triggerWidth = trigger.offsetWidth;
			const scrollTo = triggerLeft - listWidth / 2 + triggerWidth / 2;
			list.scrollTo({
				left: scrollTo,
				behavior: "smooth"
			});
		};
		list.addEventListener("click", handleClick);
		return () => list.removeEventListener("click", handleClick);
	}, []);
	const scrollLeft = () => {
		if (listRef.current) listRef.current.scrollBy({
			left: -120,
			behavior: "smooth"
		});
	};
	const scrollRight = () => {
		if (listRef.current) listRef.current.scrollBy({
			left: 120,
			behavior: "smooth"
		});
	};
	return /* @__PURE__ */ (0, react_jsx_runtime.jsxs)("div", {
		className: "openui-tabs-list-container",
		children: [
			showLeftButton && /* @__PURE__ */ (0, react_jsx_runtime.jsx)("div", {
				className: "openui-tabs-scroll-button-container-left",
				children: /* @__PURE__ */ (0, react_jsx_runtime.jsx)(require_components_IconButton_index.IconButton, {
					className: "openui-tabs-scroll-button openui-tabs-scroll-left",
					onClick: scrollLeft,
					"aria-label": "Scroll tabs left",
					icon: /* @__PURE__ */ (0, react_jsx_runtime.jsx)(lucide_react.ChevronLeft, {}),
					variant: "secondary",
					size: "small"
				})
			}),
			/* @__PURE__ */ (0, react_jsx_runtime.jsx)(_radix_ui_react_tabs.List, {
				ref: (node) => {
					if (typeof ref === "function") ref(node);
					else if (ref) ref.current = node;
					listRef.current = node;
				},
				className: (0, clsx.default)("openui-tabs-list", tabsListVariants[variant], className),
				style,
				...props
			}),
			/* @__PURE__ */ (0, react_jsx_runtime.jsx)("div", {
				className: "openui-tabs-indicator",
				ref: indicatorRef
			}),
			showRightButton && /* @__PURE__ */ (0, react_jsx_runtime.jsx)("div", {
				className: "openui-tabs-scroll-button-container-right",
				children: /* @__PURE__ */ (0, react_jsx_runtime.jsx)(require_components_IconButton_index.IconButton, {
					className: "openui-tabs-scroll-button openui-tabs-scroll-right",
					onClick: scrollRight,
					"aria-label": "Scroll tabs right",
					icon: /* @__PURE__ */ (0, react_jsx_runtime.jsx)(lucide_react.ChevronRight, {}),
					variant: "secondary",
					size: "small"
				})
			})
		]
	});
});
TabsList.displayName = "TabsList";
const TabsTrigger = (0, react.forwardRef)(({ className, style, icon, text, subtext, image, value, ...props }, ref) => /* @__PURE__ */ (0, react_jsx_runtime.jsxs)(_radix_ui_react_tabs.Trigger, {
	ref,
	className: (0, clsx.default)("openui-tabs-trigger", className),
	style,
	value,
	...props,
	children: [
		image && /* @__PURE__ */ (0, react_jsx_runtime.jsx)("span", {
			className: "openui-tabs-trigger-image",
			children: /* @__PURE__ */ (0, react_jsx_runtime.jsx)("img", {
				src: image,
				alt: ""
			})
		}),
		icon && /* @__PURE__ */ (0, react_jsx_runtime.jsx)("span", {
			className: "openui-tabs-trigger-icon",
			children: icon
		}),
		/* @__PURE__ */ (0, react_jsx_runtime.jsxs)("span", {
			className: "openui-tabs-trigger-content",
			children: [text && /* @__PURE__ */ (0, react_jsx_runtime.jsx)("span", {
				className: "openui-tabs-trigger-text",
				children: text
			}), subtext && /* @__PURE__ */ (0, react_jsx_runtime.jsx)("span", {
				className: "openui-tabs-trigger-subtext",
				children: subtext
			})]
		})
	]
}));
TabsTrigger.displayName = "TabsTrigger";
const TabsContent = (0, react.forwardRef)(({ className, style, children, ...props }, ref) => /* @__PURE__ */ (0, react_jsx_runtime.jsx)(_radix_ui_react_tabs.Content, {
	ref,
	className: (0, clsx.default)("openui-tabs-content", className),
	style,
	...props,
	children: /* @__PURE__ */ (0, react_jsx_runtime.jsx)("div", {
		className: "openui-tabs-content-inner",
		children
	})
}));
TabsContent.displayName = "TabsContent";
//#endregion
exports.Tabs = Tabs;
exports.TabsContent = TabsContent;
exports.TabsList = TabsList;
exports.TabsTrigger = TabsTrigger;
