// @openuidev/react-ui/dist/components/Tabs/Tabs.js
import { jsx as _jsx, jsxs as _jsxs } from "react/jsx-runtime";
import * as TabsPrimitive from "@radix-ui/react-tabs";
import clsx from "clsx";
import { ChevronLeft, ChevronRight } from "lucide-react";
import { forwardRef, useCallback, useEffect, useRef, useState } from "react";
import { IconButton } from "../IconButton";
const tabsVariants = {
    clear: "openui-tabs-clear",
};
export const Tabs = forwardRef(({ className, style, variant = "clear", ...props }, ref) => (_jsx(TabsPrimitive.Root, { ref: ref, className: clsx("openui-tabs", tabsVariants[variant], className), style: style, ...props })));
Tabs.displayName = "Tabs";
const tabsListVariants = {
    title: "openui-tabs-list--title",
    iconTitle: "openui-tabs-list--icon-title",
    iconTitleSubtext: "openui-tabs-list--icon-title-subtext",
    imageTitle: "openui-tabs-list--image-title",
    imageTitleSubtext: "openui-tabs-list--image-title-subtext",
};
export const TabsList = forwardRef(({ className, style, variant = "title", ...props }, ref) => {
    const listRef = useRef(null);
    const indicatorRef = useRef(null);
    const [showLeftButton, setShowLeftButton] = useState(false);
    const [showRightButton, setShowRightButton] = useState(false);
    const updateIndicator = useCallback((animated) => {
        const list = listRef.current;
        const indicator = indicatorRef.current;
        if (!list || !indicator)
            return;
        const activeTab = list.querySelector('[role="tab"][data-state="active"]');
        if (!activeTab)
            return;
        const left = activeTab.offsetLeft - list.scrollLeft;
        const width = activeTab.offsetWidth;
        indicator.style.transition = animated
            ? "transform 0.25s cubic-bezier(0.4, 0, 0.2, 1), width 0.25s cubic-bezier(0.4, 0, 0.2, 1)"
            : "none";
        indicator.style.width = `${width}px`;
        indicator.style.transform = `translateX(${left}px)`;
        indicator.style.opacity = "1";
    }, []);
    // Check scroll state and keep indicator in sync while scrolling
    useEffect(() => {
        const checkScroll = () => {
            if (listRef.current) {
                const { scrollLeft, scrollWidth, clientWidth } = listRef.current;
                setShowLeftButton(scrollLeft > 0);
                setShowRightButton(scrollLeft < scrollWidth - clientWidth - 1);
            }
        };
        const handleScroll = () => {
            checkScroll();
            updateIndicator(false);
        };
        checkScroll();
        updateIndicator(false);
        const currentRef = listRef.current;
        if (currentRef) {
            currentRef.addEventListener("scroll", handleScroll);
            const resizeObserver = new ResizeObserver(() => {
                checkScroll();
                updateIndicator(false);
            });
            resizeObserver.observe(currentRef);
            return () => {
                currentRef.removeEventListener("scroll", handleScroll);
                resizeObserver.disconnect();
            };
        }
        return () => { };
    }, [updateIndicator]);
    // Watch for active tab changes and animate indicator
    useEffect(() => {
        const list = listRef.current;
        if (!list)
            return;
        const observer = new MutationObserver(() => updateIndicator(true));
        observer.observe(list, {
            attributes: true,
            subtree: true,
            attributeFilter: ["data-state"],
        });
        return () => observer.disconnect();
    }, [updateIndicator]);
    // Center the clicked trigger in the list
    useEffect(() => {
        const list = listRef.current;
        if (!list)
            return;
        const handleClick = (e) => {
            const trigger = e.target.closest('[role="tab"]');
            if (!trigger)
                return;
            const listWidth = list.clientWidth;
            const triggerLeft = trigger.offsetLeft;
            const triggerWidth = trigger.offsetWidth;
            const scrollTo = triggerLeft - listWidth / 2 + triggerWidth / 2;
            list.scrollTo({ left: scrollTo, behavior: "smooth" });
        };
        list.addEventListener("click", handleClick);
        return () => list.removeEventListener("click", handleClick);
    }, []);
    const scrollLeft = () => {
        if (listRef.current) {
            listRef.current.scrollBy({ left: -120, behavior: "smooth" });
        }
    };
    const scrollRight = () => {
        if (listRef.current) {
            listRef.current.scrollBy({ left: 120, behavior: "smooth" });
        }
    };
    return (_jsxs("div", { className: "openui-tabs-list-container", children: [showLeftButton && (_jsx("div", { className: "openui-tabs-scroll-button-container-left", children: _jsx(IconButton, { className: "openui-tabs-scroll-button openui-tabs-scroll-left", onClick: scrollLeft, "aria-label": "Scroll tabs left", icon: _jsx(ChevronLeft, {}), variant: "secondary", size: "small" }) })), _jsx(TabsPrimitive.List, { ref: (node) => {
                    if (typeof ref === "function") {
                        ref(node);
                    }
                    else if (ref) {
                        ref.current = node;
                    }
                    listRef.current = node;
                }, className: clsx("openui-tabs-list", tabsListVariants[variant], className), style: style, ...props }), _jsx("div", { className: "openui-tabs-indicator", ref: indicatorRef }), showRightButton && (_jsx("div", { className: "openui-tabs-scroll-button-container-right", children: _jsx(IconButton, { className: "openui-tabs-scroll-button openui-tabs-scroll-right", onClick: scrollRight, "aria-label": "Scroll tabs right", icon: _jsx(ChevronRight, {}), variant: "secondary", size: "small" }) }))] }));
});
TabsList.displayName = "TabsList";
export const TabsTrigger = forwardRef(({ className, style, icon, text, subtext, image, value, ...props }, ref) => (_jsxs(TabsPrimitive.Trigger, { ref: ref, className: clsx("openui-tabs-trigger", className), style: style, value: value, ...props, children: [image && (_jsx("span", { className: "openui-tabs-trigger-image", children: _jsx("img", { src: image, alt: "" }) })), icon && _jsx("span", { className: "openui-tabs-trigger-icon", children: icon }), _jsxs("span", { className: "openui-tabs-trigger-content", children: [text && _jsx("span", { className: "openui-tabs-trigger-text", children: text }), subtext && _jsx("span", { className: "openui-tabs-trigger-subtext", children: subtext })] })] })));
TabsTrigger.displayName = "TabsTrigger";
export const TabsContent = forwardRef(({ className, style, children, ...props }, ref) => (_jsx(TabsPrimitive.Content, { ref: ref, className: clsx("openui-tabs-content", className), style: style, ...props, children: _jsx("div", { className: "openui-tabs-content-inner", children: children }) })));
TabsContent.displayName = "TabsContent";
