// @openuidev/react-ui/dist/components/FollowUpBlock/stories/FollowUpBlock.stories.js
import { jsx as _jsx, jsxs as _jsxs } from "react/jsx-runtime";
import { Plus } from "lucide-react";
import { FollowUpItem } from "../../FollowUpItem";
import { FollowUpBlock } from "../FollowUpBlock";
const meta = {
    title: "Components/FollowUpBlock",
    component: FollowUpBlock,
    subcomponents: { FollowUpItem },
    tags: ["!dev", "autodocs"],
    parameters: {
        layout: "centered",
        docs: {
            description: {
                component: "```tsx\nimport { FollowUpBlock, FollowUpItem } from '@openui-ui/react-ui';\n```",
            },
        },
    },
    argTypes: {
        children: {
            control: false,
            description: "The child elements of the follow up block that accept FollowUpItem components",
            table: {
                category: "Content",
                type: { summary: "ReactNode[] | ReactNode" },
                expanded: false,
            },
        },
        className: {
            control: false,
            description: "Additional CSS class name for custom styling",
            table: {
                category: "Styling",
                type: { summary: "string" },
                expanded: false,
            },
        },
        style: {
            control: false,
            description: "Inline CSS styles for custom styling",
            table: {
                category: "Styling",
                type: { summary: "CSSProperties" },
                expanded: false,
            },
        },
    },
};
export default meta;
export const FollowUpBlockWithItems = {
    render: (args) => (_jsxs(FollowUpBlock, { children: [_jsx(FollowUpItem, { text: "What is machine learning?", icon: _jsx(Plus, { size: 16 }) }), _jsx(FollowUpItem, { text: "what is the weather in tokyo?", icon: _jsx(Plus, { size: 16 }) }), _jsx(FollowUpItem, { text: "How to leverage AI to improve customer service?", icon: _jsx(Plus, { size: 16 }) })] })),
};
