// @openuidev/react-ui/dist/components/CheckBoxGroup/stories/CheckBoxGroup.stories.js
import { jsx as _jsx } from "react/jsx-runtime";
import { CheckBoxItem } from "../../CheckBoxItem";
import { CheckBoxGroup } from "../CheckBoxGroup";
const meta = {
    title: "Components/CheckBoxGroup",
    component: CheckBoxGroup,
    subcomponents: { CheckBoxItem },
    parameters: {
        layout: "centered",
        docs: {
            description: {
                component: "```tsx\nimport { CheckBoxGroup, CheckBoxItem } from '@openuidev/react-ui';\n```",
            },
        },
    },
    argTypes: {
        variant: {
            control: "radio",
            options: ["clear", "card", "sunk"],
            description: "The visual style variant of the checkbox group",
            table: {
                category: "Appearance",
                type: {
                    summary: "'clear' | 'card' | 'sunk'",
                },
                defaultValue: { summary: "clear" },
            },
        },
        children: {
            control: false,
            description: "The child elements of the checkbox group that accept CheckBoxItem components",
            table: {
                category: "Content",
                type: {
                    summary: "ReactNode[] | ReactNode",
                },
                expanded: false,
            },
        },
        className: {
            control: false,
            description: "Additional CSS class name for custom styling",
            table: {
                category: "Styling",
                type: {
                    summary: "string",
                },
                expanded: false,
            },
        },
        style: {
            control: false,
            description: "Inline CSS styles for custom styling",
            table: {
                category: "Styling",
                type: {
                    summary: "CSSProperties",
                },
                expanded: false,
            },
        },
    },
    tags: ["!dev", "autodocs"],
};
export default meta;
export const ClearVariant = {
    args: {
        variant: "card",
        children: [
            _jsx(CheckBoxItem, { label: "Option 1", description: "This is a description" }, "1"),
            _jsx(CheckBoxItem, { label: "Option 2", description: "This is a description" }, "2"),
            _jsx(CheckBoxItem, { label: "Option 3", description: "This is a description" }, "3"),
        ],
    },
};
export const WithLongDescription = {
    args: {
        variant: "card",
        children: [
            _jsx(CheckBoxItem, { label: "First option with a long description", description: "This is a long description that elaborates on the choice in great detail so that readers can fully understand the implications of selecting this option. It provides context, examples, and any caveats that might be relevant when making a selection." }, "long1"),
            _jsx(CheckBoxItem, { label: "Second option with a long description", description: "Another extended description that spans multiple sentences to showcase how the component behaves with verbose content. It should wrap correctly and remain readable without breaking the layout of the checkbox group." }, "long2"),
            _jsx(CheckBoxItem, { label: "Third option with a long description", description: "A very long explanation that includes more nuance about the option, possible trade-offs, and guidance for when this might be preferred. This helps validate the design for accessibility and usability with real-world copy lengths." }, "long3"),
        ],
    },
};
