// @openuidev/react-ui/dist/components/Button/Button.js
import { jsxs as _jsxs } from "react/jsx-runtime";
import clsx from "clsx";
import { forwardRef } from "react";
const normalVariantMap = {
    primary: "openui-button-base-primary",
    secondary: "openui-button-base-secondary",
    tertiary: "openui-button-base-tertiary",
};
const destructiveVariantMap = {
    primary: "openui-button-base-destructive-primary",
    secondary: "openui-button-base-destructive-secondary",
    tertiary: "openui-button-base-destructive-tertiary",
};
const sizeMap = {
    "extra-small": "openui-button-base-extra-small",
    small: "openui-button-base-small",
    medium: "openui-button-base-medium",
    large: "openui-button-base-large",
};
export const Button = forwardRef(({ children, variant = "primary", size = "medium", iconLeft, iconRight, className, buttonType = "normal", ...props }, ref) => {
    const variantMap = buttonType === "destructive" ? destructiveVariantMap : normalVariantMap;
    return (_jsxs("button", { ref: ref, className: clsx("openui-button-base", variantMap[variant], sizeMap[size], className), ...props, children: [iconLeft, children, iconRight] }));
});
Button.displayName = "Button";
