// @openuidev/react-ui/dist/components/Button/index.cjs
Object.defineProperty(exports, Symbol.toStringTag, { value: "Module" });
const require_chunk = require("../../chunk-CKQMccvm.cjs");
let clsx = require("clsx");
clsx = require_chunk.__toESM(clsx);
let react = require("react");
let react_jsx_runtime = require("react/jsx-runtime");
//#region src/components/Button/Button.tsx
const normalVariantMap = {
	primary: "openui-button-base-primary",
	secondary: "openui-button-base-secondary",
	tertiary: "openui-button-base-tertiary"
};
const destructiveVariantMap = {
	primary: "openui-button-base-destructive-primary",
	secondary: "openui-button-base-destructive-secondary",
	tertiary: "openui-button-base-destructive-tertiary"
};
const sizeMap = {
	"extra-small": "openui-button-base-extra-small",
	small: "openui-button-base-small",
	medium: "openui-button-base-medium",
	large: "openui-button-base-large"
};
const Button = (0, react.forwardRef)(({ children, variant = "primary", size = "medium", iconLeft, iconRight, className, buttonType = "normal", ...props }, ref) => {
	return /* @__PURE__ */ (0, react_jsx_runtime.jsxs)("button", {
		ref,
		className: (0, clsx.default)("openui-button-base", (buttonType === "destructive" ? destructiveVariantMap : normalVariantMap)[variant], sizeMap[size], className),
		...props,
		children: [
			iconLeft,
			children,
			iconRight
		]
	});
});
Button.displayName = "Button";
//#endregion
exports.Button = Button;
