// @openuidev/react-ui/dist/components/Charts/MiniLineChart/MiniLineChart.js
import { jsx as _jsx, jsxs as _jsxs } from "react/jsx-runtime";
import clsx from "clsx";
import { useEffect, useMemo, useRef, useState } from "react";
import { Line, LineChart as RechartsLineChart, XAxis } from "recharts";
import { usePrintContext } from "../../../context/PrintContext";
import { ChartContainer } from "../Charts";
import { DATA_KEY, getRecentDataThatFits, transformDataForChart, } from "../utils/AreaAndLine/MiniAreaAndLineUtils";
import { getLineType } from "../utils/AreaAndLine/common";
import { useChartPalette } from "../utils/PalletUtils";
import { get2dChartConfig } from "../utils/dataUtils";
export const MiniLineChart = ({ data, theme = "ocean", customPalette, variant: lineChartVariant = "natural", strokeWidth = 2, isAnimationActive = true, onLineClick, size = "100%", className, lineColor, }) => {
    const printContext = usePrintContext();
    isAnimationActive = printContext ? false : isAnimationActive;
    const containerRef = useRef(null);
    const [containerWidth, setContainerWidth] = useState(0);
    const variant = getLineType(lineChartVariant);
    useEffect(() => {
        if (!containerRef.current) {
            return () => { };
        }
        const resizeObserver = new ResizeObserver((entries) => {
            for (const entry of entries) {
                setContainerWidth(entry.contentRect.width);
            }
        });
        resizeObserver.observe(containerRef.current);
        return () => {
            resizeObserver.disconnect();
        };
    }, []);
    // Get the most recent data that fits in the container
    const filteredData = useMemo(() => {
        return getRecentDataThatFits(data, containerWidth);
    }, [data, containerWidth]);
    // Transform the filtered data to a consistent format for recharts
    const chartData = useMemo(() => {
        return transformDataForChart(filteredData);
    }, [filteredData]);
    const colors = useChartPalette({
        chartThemeName: theme,
        customPalette: customPalette || (lineColor ? [lineColor] : undefined),
        themePaletteName: "lineChartPalette",
        dataLength: 1,
    });
    const transformedKeys = useMemo(() => ({ [DATA_KEY]: DATA_KEY }), []);
    const chartConfig = useMemo(() => {
        return get2dChartConfig([DATA_KEY], colors, transformedKeys);
    }, [colors, transformedKeys]);
    return (_jsx(ChartContainer, { config: chartConfig, style: {
            width: size,
            height: size,
            aspectRatio: 1 / 1,
            minHeight: 100,
            minWidth: 100,
        }, rechartsProps: {
            aspect: 1 / 1,
        }, onClick: onLineClick, ref: containerRef, className: clsx("openui-charts-mini-line-chart-container", className), children: _jsxs(RechartsLineChart, { accessibilityLayer: true, data: chartData, margin: {
                top: 10,
            }, children: [_jsx(XAxis, { dataKey: "label", hide: true }), _jsx(Line, { dataKey: DATA_KEY, type: variant, stroke: `var(--color-${DATA_KEY})`, strokeWidth: strokeWidth, dot: false, isAnimationActive: isAnimationActive })] }) }));
};
