// @openuidev/react-ui/dist/components/RadioItem/index.cjs
Object.defineProperty(exports, Symbol.toStringTag, { value: "Module" });
const require_chunk = require("../../chunk-CKQMccvm.cjs");
let clsx = require("clsx");
clsx = require_chunk.__toESM(clsx);
let react = require("react");
let react_jsx_runtime = require("react/jsx-runtime");
let _radix_ui_react_radio_group = require("@radix-ui/react-radio-group");
_radix_ui_react_radio_group = require_chunk.__toESM(_radix_ui_react_radio_group);
//#region src/components/RadioItem/RadioItem.tsx
const RadioItem = (0, react.forwardRef)((props, ref) => {
	const { label, description, className, style, disabled, required, value } = props;
	const id = (0, react.useId)();
	return /* @__PURE__ */ (0, react_jsx_runtime.jsxs)("label", {
		htmlFor: id,
		className: "openui-radio-item-container",
		children: [/* @__PURE__ */ (0, react_jsx_runtime.jsx)(_radix_ui_react_radio_group.Item, {
			ref,
			id,
			className: (0, clsx.default)("openui-radio-item-root", className),
			value,
			disabled,
			required,
			style,
			children: /* @__PURE__ */ (0, react_jsx_runtime.jsxs)("svg", {
				width: 16,
				height: 16,
				fill: "none",
				viewBox: "0 0 16 16",
				className: "openui-radio-item-svg",
				children: [
					/* @__PURE__ */ (0, react_jsx_runtime.jsx)("path", {
						fill: "currentColor",
						d: "M1 8a7 7 0 1 1 14 0A7 7 0 0 1 1 8Z",
						className: "openui-radio-item-svg-path"
					}),
					/* @__PURE__ */ (0, react_jsx_runtime.jsx)("path", {
						stroke: "currentColor",
						d: "M1.5 8a6.5 6.5 0 1 1 13 0 6.5 6.5 0 0 1-13 0Z",
						className: "openui-radio-item-svg-border"
					}),
					/* @__PURE__ */ (0, react_jsx_runtime.jsx)("path", {
						fill: "currentColor",
						d: "M4 8a4 4 0 1 1 8 0 4 4 0 0 1-8 0Z",
						className: "openui-radio-item-svg-inner"
					})
				]
			})
		}), /* @__PURE__ */ (0, react_jsx_runtime.jsxs)("div", {
			className: "openui-radio-item-content",
			children: [label && /* @__PURE__ */ (0, react_jsx_runtime.jsx)("label", {
				htmlFor: id,
				className: "openui-radio-item-label",
				children: label
			}), description && /* @__PURE__ */ (0, react_jsx_runtime.jsx)("p", {
				className: "openui-radio-item-description",
				children: description
			})]
		})]
	});
});
RadioItem.displayName = "RadioItem";
//#endregion
exports.RadioItem = RadioItem;
