// @openuidev/react-ui/dist/genui-lib/FollowUpBlock/index.js
"use client";
import { jsx as _jsx } from "react/jsx-runtime";
import { defineComponent, useTriggerAction } from "@openuidev/react-lang";
import { z } from "zod/v4";
import { FollowUpBlock as OpenUIFollowUpBlock } from "../../components/FollowUpBlock";
import { FollowUpItem as OpenUIFollowUpItem } from "../../components/FollowUpItem";
import { FollowUpItem } from "../FollowUpItem";
export const FollowUpBlock = defineComponent({
    name: "FollowUpBlock",
    props: z.object({
        items: z.array(FollowUpItem.ref),
    }),
    description: "List of clickable follow-up suggestions placed at the end of a response",
    component: ({ props }) => {
        const triggerAction = useTriggerAction();
        const items = (props.items ?? []);
        return (_jsx(OpenUIFollowUpBlock, { children: items.map((item, i) => {
                const text = String(item?.props?.text ?? "");
                return _jsx(OpenUIFollowUpItem, { text: text, onClick: () => triggerAction(text) }, i);
            }) }));
    },
});
