// @openuidev/react-ui/dist/components/Tag/index.cjs
Object.defineProperty(exports, Symbol.toStringTag, { value: "Module" });
const require_chunk = require("../../chunk-CKQMccvm.cjs");
let clsx = require("clsx");
clsx = require_chunk.__toESM(clsx);
let react = require("react");
let react_jsx_runtime = require("react/jsx-runtime");
//#region src/components/Tag/Tag.tsx
const sizeMap = {
	sm: "openui-tag-sm",
	md: "openui-tag-md",
	lg: "openui-tag-lg"
};
const variantMap = {
	neutral: "openui-tag-neutral",
	info: "openui-tag-info",
	success: "openui-tag-success",
	warning: "openui-tag-warning",
	danger: "openui-tag-danger"
};
const Tag = (0, react.forwardRef)((props, ref) => {
	const { className, styles, icon, text, size = "md", variant = "neutral", ...rest } = props;
	return /* @__PURE__ */ (0, react_jsx_runtime.jsxs)("div", {
		ref,
		className: (0, clsx.default)("openui-tag", sizeMap[size], variantMap[variant], className),
		style: styles,
		...rest,
		children: [icon && /* @__PURE__ */ (0, react_jsx_runtime.jsx)("span", {
			className: "openui-tag-icon",
			children: icon
		}), /* @__PURE__ */ (0, react_jsx_runtime.jsx)("span", {
			className: "openui-tag-text",
			children: text
		})]
	});
});
Tag.displayName = "Tag";
//#endregion
exports.Tag = Tag;
