// @openuidev/react-ui/dist/components/Shell/stories/Shell.stories.js
import { jsx as _jsx, Fragment as _Fragment, jsxs as _jsxs } from "react/jsx-runtime";
import { ChatProvider } from "@openuidev/react-headless";
import { MessageSquare, Share, Sparkles, Zap } from "lucide-react";
import { Button } from "../../Button";
import { IconButton } from "../../IconButton";
import { Container } from "../Container";
import { ConversationStarter } from "../ConversationStarter";
import { MobileHeader } from "../MobileHeader";
import { NewChatButton } from "../NewChatButton";
import { SidebarContainer, SidebarContent, SidebarHeader, SidebarSeparator } from "../Sidebar";
import { Composer, MessageLoading, Messages, ScrollArea, ThreadContainer, ThreadHeader, } from "../Thread";
import { ThreadList } from "../ThreadList";
import { WelcomeScreen } from "../WelcomeScreen";
import logoUrl from "./thesysdev_logo.jpeg";
function mockSSEResponse(text, delayMs = 500) {
    return new Promise((resolve) => {
        setTimeout(() => {
            const events = `data: ${JSON.stringify({ type: "TEXT_MESSAGE_CONTENT", delta: text })}\n\ndata: [DONE]\n\n`;
            const stream = new ReadableStream({
                start(controller) {
                    controller.enqueue(new TextEncoder().encode(events));
                    controller.close();
                },
            });
            resolve(new Response(stream));
        }, delayMs);
    });
}
function getLastUserContent(messages) {
    const lastUser = [...messages].reverse().find((m) => m.role === "user");
    if (!lastUser)
        return "";
    return typeof lastUser.content === "string" ? lastUser.content : "";
}
export default {
    title: "Components/Shell",
    tags: ["dev"],
    argTypes: {
        variant: {
            control: "select",
            options: ["short", "long"],
            description: "Conversation starter variant",
        },
    },
};
const SAMPLE_STARTERS = [
    {
        displayText: "Help me get started",
        prompt: "Help me get started",
        icon: _jsx(Sparkles, { size: 16 }),
    },
    {
        displayText: "What can you do?",
        prompt: "What can you do?",
    },
    {
        displayText: "Tell me about your features",
        prompt: "Tell me about your features",
        icon: _jsx(MessageSquare, { size: 16 }),
    },
    {
        displayText: "Show me some examples (no icon)",
        prompt: "Show me some examples",
        icon: _jsx(_Fragment, {}),
    },
];
const LONG_STARTERS = [
    {
        displayText: "Help me get started with this application and guide me through the features",
        prompt: "Help me get started with this application",
        icon: _jsx(Sparkles, { size: 16 }),
    },
    {
        displayText: "What can you do? I'd like to know all your capabilities and how you can help me",
        prompt: "What can you do?",
        icon: _jsx(Zap, { size: 16 }),
    },
    {
        displayText: "Tell me about your advanced features and how I can use them effectively",
        prompt: "Tell me about your features",
    },
];
export const Default = {
    args: {
        variant: "short",
    },
    render: ({ variant }) => (_jsx(ChatProvider, { processMessage: async () => {
            await new Promise((r) => setTimeout(r, 100));
            return mockSSEResponse("This is a response from the AI assistant.", 1000);
        }, fetchThreadList: async () => {
            await new Promise((resolve) => setTimeout(resolve, 1000));
            return {
                threads: [
                    { id: "1", title: "test", createdAt: Date.now() },
                    { id: "2", title: "test 2", createdAt: Date.now() },
                    { id: "3", title: "test 3", createdAt: Date.now() },
                ],
            };
        }, createThread: async () => ({
            id: crypto.randomUUID(),
            title: "test",
            createdAt: Date.now(),
        }), deleteThread: async () => { }, updateThread: async (t) => t, loadThread: async (threadId) => {
            if (!threadId)
                return [];
            return [
                { id: crypto.randomUUID(), role: "user", content: "Hello" },
                {
                    id: crypto.randomUUID(),
                    role: "assistant",
                    content: "Hello! How can I help you today?",
                },
            ];
        }, children: _jsxs(Container, { logoUrl: logoUrl, agentName: "OpenUI", children: [_jsxs(SidebarContainer, { children: [_jsx(SidebarHeader, { children: _jsx(NewChatButton, {}) }), _jsxs(SidebarContent, { children: [_jsx(SidebarSeparator, {}), _jsx(ThreadList, {})] })] }), _jsxs(ThreadContainer, { children: [_jsx(MobileHeader, { rightChildren: _jsx(IconButton, { icon: _jsx(Share, { size: 16 }), "aria-label": "Share", size: "medium", variant: "secondary" }) }), _jsx(ThreadHeader, { children: _jsx(Button, { iconLeft: _jsx(Share, { size: 16 }), variant: "secondary", size: "small", children: "Share" }) }), _jsx(ScrollArea, { children: _jsx(Messages, { loader: _jsx(MessageLoading, {}) }) }), _jsx(ConversationStarter, { starters: SAMPLE_STARTERS, variant: variant }), _jsx(Composer, {})] })] }) })),
};
export const WithThreadHeader = {
    args: {
        variant: "short",
    },
    render: ({ variant }) => (_jsx(ChatProvider, { processMessage: async ({ messages }) => {
            const content = getLastUserContent(messages);
            return mockSSEResponse(`You asked: "${content}"`, 1000);
        }, fetchThreadList: async () => ({ threads: [] }), createThread: async () => ({
            id: crypto.randomUUID(),
            title: "New Chat",
            createdAt: Date.now(),
        }), deleteThread: async () => { }, updateThread: async (t) => t, loadThread: async () => [], children: _jsxs(Container, { logoUrl: logoUrl, agentName: "OpenUI", children: [_jsxs(SidebarContainer, { children: [_jsx(SidebarHeader, { children: _jsx(NewChatButton, {}) }), _jsxs(SidebarContent, { children: [_jsx(SidebarSeparator, {}), _jsx(ThreadList, {})] })] }), _jsxs(ThreadContainer, { children: [_jsx(MobileHeader, {}), _jsxs(ThreadHeader, { children: [_jsxs("select", { style: {
                                        padding: "4px 8px",
                                        borderRadius: "6px",
                                        border: "1px solid var(--openui-stroke-default, #ccc)",
                                        background: "var(--openui-bg-fill, #fff)",
                                        fontSize: "13px",
                                    }, children: [_jsx("option", { children: "GPT-4o" }), _jsx("option", { children: "Claude 3.5 Sonnet" }), _jsx("option", { children: "Gemini Pro" })] }), _jsx(Button, { iconLeft: _jsx(Share, { size: 16 }), variant: "secondary", size: "small", children: "Share" })] }), _jsx(ScrollArea, { children: _jsx(Messages, { loader: _jsx(MessageLoading, {}) }) }), _jsx(ConversationStarter, { starters: SAMPLE_STARTERS, variant: variant }), _jsx(Composer, {})] })] }) })),
};
export const WithConversationStarter = {
    args: {
        variant: "short",
    },
    render: ({ variant }) => (_jsx(ChatProvider, { processMessage: async ({ messages }) => {
            const content = getLastUserContent(messages);
            return mockSSEResponse(`You asked: "${content}"`, 1000);
        }, fetchThreadList: async () => ({ threads: [] }), createThread: async () => ({
            id: crypto.randomUUID(),
            title: "New Chat",
            createdAt: Date.now(),
        }), deleteThread: async () => { }, updateThread: async (t) => t, loadThread: async () => [], children: _jsxs(Container, { logoUrl: logoUrl, agentName: "OpenUI", children: [_jsxs(SidebarContainer, { children: [_jsx(SidebarHeader, { children: _jsx(NewChatButton, {}) }), _jsxs(SidebarContent, { children: [_jsx(SidebarSeparator, {}), _jsx(ThreadList, {})] })] }), _jsxs(ThreadContainer, { children: [_jsx(MobileHeader, { rightChildren: _jsx(IconButton, { icon: _jsx(Share, { size: 16 }), "aria-label": "Share", size: "medium", variant: "secondary" }) }), _jsx(ScrollArea, { children: _jsx(Messages, { loader: _jsx(MessageLoading, {}) }) }), _jsx(ConversationStarter, { starters: SAMPLE_STARTERS, variant: variant }), _jsx(Composer, {})] })] }) })),
};
export const LongVariant = {
    args: {
        variant: "long",
    },
    render: ({ variant }) => (_jsx(ChatProvider, { processMessage: async ({ messages }) => {
            const content = getLastUserContent(messages);
            return mockSSEResponse(`You asked: "${content}"`, 1000);
        }, fetchThreadList: async () => ({ threads: [] }), createThread: async () => ({
            id: crypto.randomUUID(),
            title: "New Chat",
            createdAt: Date.now(),
        }), deleteThread: async () => { }, updateThread: async (t) => t, loadThread: async () => [], children: _jsxs(Container, { logoUrl: logoUrl, agentName: "OpenUI", children: [_jsxs(SidebarContainer, { children: [_jsx(SidebarHeader, { children: _jsx(NewChatButton, {}) }), _jsxs(SidebarContent, { children: [_jsx(SidebarSeparator, {}), _jsx(ThreadList, {})] })] }), _jsxs(ThreadContainer, { children: [_jsx(MobileHeader, { rightChildren: _jsx(IconButton, { icon: _jsx(Share, { size: 16 }), "aria-label": "Share", size: "medium", variant: "secondary" }) }), _jsx(ScrollArea, { children: _jsx(Messages, { loader: _jsx(MessageLoading, {}) }) }), _jsx(ConversationStarter, { starters: LONG_STARTERS, variant: variant }), _jsx(Composer, {})] })] }) })),
};
export const WithWelcomeScreen = {
    args: {
        variant: "short",
    },
    render: ({ variant }) => (_jsx(ChatProvider, { processMessage: async ({ messages }) => {
            const content = getLastUserContent(messages);
            return mockSSEResponse(`You asked: "${content}"`, 1000);
        }, fetchThreadList: async () => ({ threads: [] }), createThread: async () => ({
            id: crypto.randomUUID(),
            title: "New Chat",
            createdAt: Date.now(),
        }), deleteThread: async () => { }, updateThread: async (t) => t, loadThread: async () => [], children: _jsxs(Container, { logoUrl: logoUrl, agentName: "OpenUI Assistant", children: [_jsxs(SidebarContainer, { children: [_jsx(SidebarHeader, { children: _jsx(NewChatButton, {}) }), _jsxs(SidebarContent, { children: [_jsx(SidebarSeparator, {}), _jsx(ThreadList, {})] })] }), _jsxs(ThreadContainer, { children: [_jsx(MobileHeader, { rightChildren: _jsx(IconButton, { icon: _jsx(Share, { size: 16 }), "aria-label": "Share", size: "medium", variant: "secondary" }) }), _jsx(WelcomeScreen, { title: "Hi, I'm OpenUI Assistant", description: "I can help you with questions about your account, products, and more.", image: { url: logoUrl }, starters: SAMPLE_STARTERS, starterVariant: variant }), _jsx(ScrollArea, { children: _jsx(Messages, { loader: _jsx(MessageLoading, {}) }) }), _jsx(ConversationStarter, { starters: SAMPLE_STARTERS, variant: variant }), _jsx(Composer, {})] })] }) })),
};
export const WithCustomWelcomeScreen = {
    args: {
        variant: "short",
    },
    render: ({ variant }) => (_jsx(ChatProvider, { processMessage: async ({ messages }) => {
            const content = getLastUserContent(messages);
            return mockSSEResponse(`You asked: "${content}"`, 1000);
        }, fetchThreadList: async () => ({ threads: [] }), createThread: async () => ({
            id: crypto.randomUUID(),
            title: "New Chat",
            createdAt: Date.now(),
        }), deleteThread: async () => { }, updateThread: async (t) => t, loadThread: async () => [], children: _jsxs(Container, { logoUrl: logoUrl, agentName: "OpenUI Assistant", children: [_jsxs(SidebarContainer, { children: [_jsx(SidebarHeader, { children: _jsx(NewChatButton, {}) }), _jsxs(SidebarContent, { children: [_jsx(SidebarSeparator, {}), _jsx(ThreadList, {})] })] }), _jsxs(ThreadContainer, { children: [_jsx(MobileHeader, { rightChildren: _jsx(IconButton, { icon: _jsx(Share, { size: 16 }), "aria-label": "Share", size: "medium", variant: "secondary" }) }), _jsx(WelcomeScreen, { children: _jsxs("div", { style: { textAlign: "center" }, children: [_jsx("div", { style: {
                                            width: 80,
                                            height: 80,
                                            borderRadius: 16,
                                            background: "linear-gradient(135deg, #667eea 0%, #764ba2 100%)",
                                            display: "flex",
                                            alignItems: "center",
                                            justifyContent: "center",
                                            margin: "0 auto 16px",
                                        }, children: _jsx(Sparkles, { size: 40, color: "white" }) }), _jsx("h2", { style: { margin: "0 0 8px", fontSize: 24, fontWeight: 600 }, children: "Welcome to AI Assistant" }), _jsx("p", { style: { margin: 0, color: "rgba(0,0,0,0.5)", fontSize: 16 }, children: "Your personal AI helper for all your questions" })] }) }), _jsx(ScrollArea, { children: _jsx(Messages, { loader: _jsx(MessageLoading, {}) }) }), _jsx(ConversationStarter, { starters: SAMPLE_STARTERS, variant: variant }), _jsx(Composer, {})] })] }) })),
};
