// @openuidev/react-ui/dist/components/Charts/shared/XAxisTick/XAxisTick.js
import { jsx as _jsx } from "react/jsx-runtime";
import clsx from "clsx";
import React, { useLayoutEffect, useRef, useState } from "react";
import { LabelTooltip } from "../LabelTooltip/LabelTooltip";
const XAxisTick = React.forwardRef((props, ref) => {
    const { x, y, payload, className, variant = "multiLine", widthOfGroup = 70, labelHeight = 20, onMouseEnter, onMouseLeave, } = props;
    const rawValue = payload?.value;
    const value = String(rawValue || "");
    const foreignObjectRef = useRef(null);
    const spanRef = useRef(null);
    const [isTruncated, setIsTruncated] = useState(false);
    // Check if text is truncated
    useLayoutEffect(() => {
        if (spanRef.current) {
            const element = spanRef.current;
            const isCurrentlyTruncated = variant === "multiLine"
                ? element.scrollHeight > element.clientHeight
                : element.scrollWidth > element.clientWidth;
            setIsTruncated(isCurrentlyTruncated);
        }
    }, [value, variant, widthOfGroup]);
    if (x === undefined || y === undefined) {
        return null;
    }
    // The x position from Recharts is the center of the group
    // To center the foreignObject, we need to offset by half of widthOfGroup
    // +2 is for the multiLine variant to center the text
    // +5 is for the singleLine variant to center the text
    // offset for padding - 4 and -10 respectively
    const calX = variant === "multiLine" ? x - widthOfGroup / 2 + 2 : x - widthOfGroup / 2 + 5;
    const calWidth = variant === "multiLine" ? widthOfGroup - 4 : widthOfGroup - 10;
    const spanClassName = variant === "multiLine"
        ? "openui-chart-x-axis-tick-multi-line"
        : "openui-chart-x-axis-tick-single-line";
    return (_jsx("g", { ref: ref, transform: `translate(${calX},${y})`, children: _jsx("foreignObject", { ref: foreignObjectRef, transform: "translate(0, 0)", width: calWidth, height: labelHeight, className: "openui-chart-x-axis-tick-foreign", children: _jsx("div", { style: {
                    width: "100%",
                    height: "100%",
                    boxSizing: "border-box",
                }, onMouseEnter: () => onMouseEnter?.(props), onMouseLeave: () => onMouseLeave?.(props), children: _jsx(LabelTooltip, { content: value, side: "top", disabled: !isTruncated, children: _jsx("span", { ref: spanRef, style: {
                            textAlign: "center",
                            wordBreak: "break-word",
                        }, className: clsx(spanClassName, className), children: value }) }) }) }) }));
});
XAxisTick.displayName = "XAxisTick";
export { XAxisTick };
