// @openuidev/react-lang/dist/index.mjs
import { ACTION_STEPS, ACTION_STEPS as ACTION_STEPS$1, BuiltinActionType, BuiltinActionType as BuiltinActionType$1, ToolNotFoundError, ToolNotFoundError as ToolNotFoundError$1, builtInValidators, createLibrary as createLibrary$1, createParser, createQueryManager, createStore, createStreamingParser, createStreamingParser as createStreamingParser$1, defineComponent as defineComponent$1, enrichErrors, evaluate, evaluateElementProps, extractToolResult, extractToolResult as extractToolResult$1, generatePrompt, isReactiveAssign, markReactive, mergeStatements, parseRules, parseStructuredRules, resolveStateField, tagSchemaId, validate, validate as validate$1 } from "@openuidev/lang-core";
import { Component, Fragment, createContext, useCallback, useContext, useEffect, useInsertionEffect, useMemo, useRef, useState, useSyncExternalStore } from "react";
import { jsx, jsxs } from "react/jsx-runtime";
//#region src/library.ts
function defineComponent(config) {
	return defineComponent$1(config);
}
function createLibrary(input) {
	return createLibrary$1(input);
}
//#endregion
//#region src/context.ts
const OpenUIContext = createContext(null);
/**
* Access the full OpenUI context. Throws if used outside a <Renderer />.
*/
function useOpenUI() {
	const ctx = useContext(OpenUIContext);
	if (!ctx) throw new Error("useOpenUI must be used within a <Renderer /> component.");
	return ctx;
}
/**
* Get the renderNode function for rendering nested component values.
*/
function useRenderNode() {
	return useOpenUI().renderNode;
}
/**
* Get the triggerAction function for firing structured action events.
*
* @example
* ```tsx
* const triggerAction = useTriggerAction();
* <button onClick={() => triggerAction("Submit", "myForm")}>
* ```
*/
function useTriggerAction() {
	return useOpenUI().triggerAction;
}
/**
* Whether the LLM is currently streaming content.
*/
function useIsStreaming() {
	return useOpenUI().isStreaming;
}
/**
* Get a form field value from the form state context.
*
* @example
* ```tsx
* const getFieldValue = useGetFieldValue();
* const name = getFieldValue("contactForm", "name");
* ```
*/
function useGetFieldValue() {
	return useOpenUI().getFieldValue;
}
/**
* Get the setFieldValue function for updating form field values.
*
* @example
* ```tsx
* const setFieldValue = useSetFieldValue();
* <input onChange={(e) => setFieldValue("contactForm", "Input", "name", e.target.value, false)} />
* ```
*/
function useSetFieldValue() {
	return useOpenUI().setFieldValue;
}
const FormNameContext = createContext(void 0);
/**
* Get the current form name (set by the nearest parent Form component).
* Returns undefined if not inside a Form.
*/
function useFormName() {
	return useContext(FormNameContext);
}
/**
* Persists a component's default/initial value into form state once streaming
* finishes — but only if the user hasn't already set a value.
*
* Call this inside any form component that has a `defaultValue` or
* `defaultChecked` prop. It is a no-op during streaming so that LLM
* prop changes don't fight with partial state.
*
* @param shouldTriggerSaveCallback — defaults to `false` (only local state, no message persistence)
*/
function useSetDefaultValue({ formName, componentType, name, existingValue, defaultValue, shouldTriggerSaveCallback = false }) {
	const setFieldValue = useSetFieldValue();
	const isStreaming = useIsStreaming();
	useEffect(() => {
		if (!isStreaming && existingValue === void 0 && defaultValue !== void 0) setFieldValue(formName, componentType, name, defaultValue, shouldTriggerSaveCallback);
	}, [
		defaultValue,
		existingValue,
		formName,
		componentType,
		name,
		isStreaming,
		setFieldValue,
		shouldTriggerSaveCallback
	]);
}
//#endregion
//#region src/hooks/useOpenUIState.ts
/** Unwrap { value, componentType } wrapper from form field entries. Returns raw value. */
function unwrapFieldValue(v) {
	if (v && typeof v === "object" && !Array.isArray(v) && "value" in v) return v.value;
	return v;
}
/**
* Core state hook — extracts all form state, action handling, parser
* management, and context assembly out of the Renderer component.
*
* Store holds everything: $bindings as top-level keys, form fields nested
* under formName as plain values.
*/
function useOpenUIState({ response, library, isStreaming, onAction, onStateUpdate, initialState, toolProvider, onError }, renderDeep) {
	const sp = useMemo(() => createStreamingParser$1(library.toJSONSchema(), library.root), [library]);
	const parseExceptionRef = useRef(null);
	const result = useMemo(() => {
		parseExceptionRef.current = null;
		if (!response) return null;
		try {
			return sp.set(response);
		} catch (e) {
			parseExceptionRef.current = {
				source: "parser",
				code: "parse-exception",
				message: `Parser crashed: ${e instanceof Error ? e.message : String(e)}`,
				hint: "The response may contain syntax the parser cannot handle"
			};
			return null;
		}
	}, [sp, response]);
	const store = useMemo(() => createStore(), []);
	const queryManager = useMemo(() => createQueryManager(toolProvider ?? null), [toolProvider]);
	useEffect(() => {
		queryManager.activate();
		return () => queryManager.dispose();
	}, [queryManager]);
	const storeInitKeyRef = useRef(Symbol());
	useEffect(() => {
		if (!result?.stateDeclarations && !initialState) return;
		const key = `${JSON.stringify(result?.stateDeclarations)}::${JSON.stringify(initialState)}`;
		if (storeInitKeyRef.current === key) return;
		storeInitKeyRef.current = key;
		const bindingDefaults = {};
		if (initialState) for (const [key, value] of Object.entries(initialState)) if (key.startsWith("$")) bindingDefaults[key] = value;
		else store.set(key, value);
		store.initialize(result?.stateDeclarations ?? {}, bindingDefaults);
	}, [
		result?.stateDeclarations,
		store,
		initialState
	]);
	const storeSnapshot = useSyncExternalStore(store.subscribe, store.getSnapshot, store.getSnapshot);
	const querySnapshot = useSyncExternalStore(queryManager.subscribe, queryManager.getSnapshot, queryManager.getSnapshot);
	const evaluationContext = useMemo(() => ({
		getState: (name) => unwrapFieldValue(store.get(name)),
		resolveRef: (name) => {
			const mutResult = queryManager.getMutationResult(name);
			if (mutResult) return mutResult;
			return queryManager.getResult(name);
		}
	}), [store, queryManager]);
	useEffect(() => {
		if (isStreaming) return;
		const evaluatedNodes = (result?.queryStatements ?? []).map((qn) => {
			const relevantDeps = {};
			if (qn.deps) for (const ref of qn.deps) relevantDeps[ref] = storeSnapshot[ref];
			return {
				statementId: qn.statementId,
				toolName: qn.toolAST ? evaluate(qn.toolAST, evaluationContext) : "",
				args: qn.argsAST ? evaluate(qn.argsAST, evaluationContext) : null,
				defaults: qn.defaultsAST ? evaluate(qn.defaultsAST, evaluationContext) : null,
				refreshInterval: qn.refreshAST ? evaluate(qn.refreshAST, evaluationContext) : void 0,
				deps: Object.keys(relevantDeps).length > 0 ? relevantDeps : void 0,
				complete: qn.complete
			};
		});
		queryManager.evaluateQueries(evaluatedNodes);
	}, [
		isStreaming,
		result?.queryStatements,
		evaluationContext,
		queryManager,
		storeSnapshot
	]);
	useEffect(() => {
		if (isStreaming) return;
		const nodes = (result?.mutationStatements ?? []).map((mn) => ({
			statementId: mn.statementId,
			toolName: mn.toolAST ? evaluate(mn.toolAST, evaluationContext) : ""
		}));
		queryManager.registerMutations(nodes);
	}, [
		isStreaming,
		result?.mutationStatements,
		evaluationContext,
		queryManager
	]);
	const propsRef = useRef({
		onAction,
		onStateUpdate,
		onError
	});
	propsRef.current = {
		onAction,
		onStateUpdate,
		onError
	};
	const resultRef = useRef(result);
	resultRef.current = result;
	const lastInitSnapshotRef = useRef(null);
	useEffect(() => {
		lastInitSnapshotRef.current = store.getSnapshot();
		return store.subscribe(() => {
			const currentSnapshot = store.getSnapshot();
			if (currentSnapshot === lastInitSnapshotRef.current) return;
			lastInitSnapshotRef.current = null;
			propsRef.current.onStateUpdate?.(currentSnapshot);
		});
	}, [store]);
	const getFieldValue = useCallback((formName, name) => {
		if (!formName) return unwrapFieldValue(store.get(name));
		const formData = store.get(formName);
		if (!formData || typeof formData !== "object" || Array.isArray(formData)) return void 0;
		return unwrapFieldValue(formData[name]);
	}, [store]);
	const setFieldValue = useCallback((formName, componentType, name, value, shouldTriggerSaveCallback = true) => {
		const wrapped = {
			value,
			componentType
		};
		if (!formName) store.set(name, wrapped);
		else {
			const raw = store.get(formName);
			const formData = raw && typeof raw === "object" && !Array.isArray(raw) ? raw : {};
			store.set(formName, {
				...formData,
				[name]: wrapped
			});
		}
		if (shouldTriggerSaveCallback) propsRef.current.onStateUpdate?.(store.getSnapshot());
	}, [store]);
	const getFormPayload = useCallback((formName) => {
		if (formName) {
			const raw = store.get(formName);
			if (raw && typeof raw === "object" && !Array.isArray(raw)) return { [formName]: raw };
		}
		return store.getSnapshot();
	}, [store]);
	const triggerAction = useCallback(async (userMessage, formName, action) => {
		const formPayload = getFormPayload(formName);
		const { onAction: handler } = propsRef.current;
		if (action && !("steps" in action)) {
			const actionType = action.type || BuiltinActionType$1.ContinueConversation;
			const params = { ...action.params || {} };
			if (action.url) params.url = action.url;
			if (action.context) params.context = action.context;
			handler?.({
				type: actionType,
				params,
				humanFriendlyMessage: userMessage,
				formState: formPayload,
				formName
			});
			return;
		}
		const actionPlan = action;
		if (actionPlan?.steps) {
			for (const step of actionPlan.steps) switch (step.type) {
				case ACTION_STEPS$1.Run:
					if (step.refType === "mutation") {
						const mn = resultRef.current?.mutationStatements?.find((m) => m.statementId === step.statementId);
						const evaluatedArgs = mn?.argsAST ? evaluate(mn.argsAST, evaluationContext) : {};
						if (!await queryManager.fireMutation(step.statementId, evaluatedArgs)) return;
					} else queryManager.invalidate([step.statementId]);
					break;
				case ACTION_STEPS$1.ToAssistant:
					handler?.({
						type: BuiltinActionType$1.ContinueConversation,
						params: step.context ? { context: step.context } : {},
						humanFriendlyMessage: step.message,
						formState: formPayload,
						formName
					});
					break;
				case ACTION_STEPS$1.OpenUrl:
					handler?.({
						type: BuiltinActionType$1.OpenUrl,
						params: { url: step.url },
						humanFriendlyMessage: "",
						formState: formPayload,
						formName
					});
					break;
				case ACTION_STEPS$1.Set: {
					if (!step.valueAST) {
						console.warn(`[openui] Set action for ${step.target} has no valueAST — skipping`);
						break;
					}
					const value = evaluate(step.valueAST, evaluationContext);
					store.set(step.target, value);
					break;
				}
				case ACTION_STEPS$1.Reset: {
					const decls = resultRef.current?.stateDeclarations ?? {};
					for (const target of step.targets) store.set(target, decls[target] ?? null);
					break;
				}
			}
			return;
		}
		handler?.({
			type: BuiltinActionType$1.ContinueConversation,
			params: {},
			humanFriendlyMessage: userMessage,
			formState: formPayload,
			formName
		});
	}, [
		queryManager,
		evaluationContext,
		getFormPayload,
		store
	]);
	const renderErrorsRef = useRef([]);
	const isStreamingRef = useRef(isStreaming);
	isStreamingRef.current = isStreaming;
	const reportError = useCallback((error) => {
		if (isStreamingRef.current) return;
		renderErrorsRef.current.push(error);
	}, []);
	const contextValue = useMemo(() => ({
		library,
		renderNode: renderDeep,
		triggerAction,
		isStreaming,
		getFieldValue,
		setFieldValue,
		store,
		evaluationContext,
		reportError
	}), [
		library,
		renderDeep,
		isStreaming,
		triggerAction,
		getFieldValue,
		setFieldValue,
		store,
		evaluationContext,
		reportError
	]);
	const runtimeErrorsRef = useRef([]);
	const evaluatedResult = useMemo(() => {
		if (!result?.root) return result;
		const errors = [];
		const evalCtx = {
			ctx: evaluationContext,
			library,
			store,
			errors
		};
		try {
			const evaluatedRoot = evaluateElementProps(result.root, evalCtx);
			runtimeErrorsRef.current = errors;
			return {
				...result,
				root: evaluatedRoot
			};
		} catch (e) {
			const msg = e instanceof Error ? e.message : String(e);
			errors.push({
				source: "runtime",
				code: "runtime-error",
				message: `Prop evaluation failed: ${msg}`
			});
			runtimeErrorsRef.current = errors;
			return result;
		}
	}, [
		result,
		evaluationContext,
		library,
		store,
		storeSnapshot,
		querySnapshot
	]);
	const isQueryLoading = querySnapshot.__openui_loading.length > 0;
	const lastErrorKeyRef = useRef("");
	useEffect(() => {
		if (isStreaming) {
			if (lastErrorKeyRef.current !== "") {
				lastErrorKeyRef.current = "";
				propsRef.current.onError?.([]);
			}
			return;
		}
		const errors = [];
		if (parseExceptionRef.current) errors.push(parseExceptionRef.current);
		if (response && !result?.root && !parseExceptionRef.current) errors.push({
			source: "parser",
			code: "parse-failed",
			message: result ? "Code parsed but produced no renderable root component" : "Response could not be parsed as valid openui-lang",
			hint: `The entire response must be valid openui-lang code starting with root = ${library.root ?? "Root"}(...)`
		});
		if (result?.meta?.errors?.length) errors.push(...enrichErrors(result.meta.errors, library.toJSONSchema(), Object.keys(library.components)));
		errors.push(...runtimeErrorsRef.current);
		errors.push(...renderErrorsRef.current);
		renderErrorsRef.current = [];
		errors.push(...querySnapshot.__openui_errors ?? []);
		const key = JSON.stringify(errors);
		if (key === lastErrorKeyRef.current) return;
		lastErrorKeyRef.current = key;
		if (propsRef.current.onError) propsRef.current.onError(errors);
		else if (errors.length > 0) for (const e of errors) console.warn(`[openui] ${e.source}/${e.code}: ${e.message}`);
	}, [
		isStreaming,
		response,
		result,
		evaluatedResult,
		querySnapshot,
		library
	]);
	return {
		result: evaluatedResult,
		parseResult: result,
		contextValue,
		isQueryLoading
	};
}
//#endregion
//#region src/Renderer.tsx
/**
* Error boundary that intentionally shows the last successfully rendered
* children when a render error occurs. This "show last good state" behavior
* prevents the UI from going blank during streaming or transient evaluation
* errors, and auto-recovers when new valid children arrive.
*/
var ElementErrorBoundary = class extends Component {
	lastValidChildren = null;
	constructor(props) {
		super(props);
		this.state = { hasError: false };
	}
	static getDerivedStateFromError() {
		return { hasError: true };
	}
	componentDidMount() {
		if (!this.state.hasError) this.lastValidChildren = this.props.children;
	}
	componentDidUpdate(prevProps) {
		if (!this.state.hasError) this.lastValidChildren = this.props.children;
		if (this.state.hasError && prevProps.children !== this.props.children) this.setState({ hasError: false });
	}
	componentDidCatch(error) {
		const name = this.props.componentName ?? "Unknown";
		this.props.onError?.({
			source: "runtime",
			code: "render-error",
			component: name,
			message: `Component ${name} render failed: ${error.message}`
		});
	}
	render() {
		if (this.state.hasError) return this.lastValidChildren;
		return this.props.children;
	}
};
/**
* Recursively renders a parsed value (element, array, primitive)
* into React nodes.
*/
function renderDeep(value) {
	if (value == null) return null;
	if (typeof value === "string") return value;
	if (typeof value === "number") return String(value);
	if (typeof value === "boolean") return String(value);
	if (Array.isArray(value)) return value.map((v, i) => /* @__PURE__ */ jsx(Fragment, { children: renderDeep(v) }, i));
	if (typeof value === "object" && value !== null) {
		const obj = value;
		if (obj.type === "element") return /* @__PURE__ */ jsx(RenderNode, { node: obj });
	}
	return null;
}
/**
* Renders a single ElementNode.
*/
function RenderNode({ node }) {
	const { library, reportError } = useOpenUI();
	const Comp = library.components[node.typeName]?.component;
	if (!Comp) return null;
	return /* @__PURE__ */ jsx(ElementErrorBoundary, {
		componentName: node.typeName,
		onError: reportError,
		children: /* @__PURE__ */ jsx(RenderNodeInner, {
			el: node,
			Comp
		})
	});
}
/**
* Renders a resolved element using its renderer.
* Props are already evaluated by evaluate-tree — no AST awareness needed.
*/
function RenderNodeInner({ el, Comp }) {
	const renderNode = useRenderNode();
	return /* @__PURE__ */ jsx(Comp, {
		props: el.props,
		renderNode
	});
}
let loadingStyleInjected = false;
function ensureLoadingStyle() {
	if (loadingStyleInjected || typeof document === "undefined") return;
	loadingStyleInjected = true;
	const style = document.createElement("style");
	style.textContent = `@keyframes openui-spin { 0% { transform: rotate(0deg); } 100% { transform: rotate(360deg); } }`;
	document.head.appendChild(style);
}
const DefaultQueryLoader = () => /* @__PURE__ */ jsx("div", { style: {
	position: "absolute",
	top: 8,
	right: 8,
	width: 16,
	height: 16,
	border: "2px solid #e5e7eb",
	borderTopColor: "#3b82f6",
	borderRadius: "50%",
	animation: "openui-spin 0.6s linear infinite",
	zIndex: 10
} });
function Renderer({ response, library, isStreaming = false, onAction, onStateUpdate, initialState, onParseResult, toolProvider, queryLoader, onError }) {
	useInsertionEffect(() => {
		ensureLoadingStyle();
	}, []);
	const onParseResultRef = useRef(onParseResult);
	onParseResultRef.current = onParseResult;
	const toolProviderInputRef = useRef(toolProvider);
	toolProviderInputRef.current = toolProvider;
	const stableToolProvider = useRef({ async callTool(toolName, args) {
		const current = toolProviderInputRef.current ?? null;
		if (current == null) throw new Error("[openui] toolProvider is null");
		if (typeof current.callTool === "function") return extractToolResult$1(await current.callTool({
			name: toolName,
			arguments: args
		}));
		const map = current;
		const fn = map[toolName];
		if (!fn) throw new ToolNotFoundError$1(toolName, Object.keys(map));
		return fn(args);
	} });
	const { result, parseResult, contextValue, isQueryLoading } = useOpenUIState({
		response,
		library,
		isStreaming,
		onAction,
		onStateUpdate,
		initialState,
		toolProvider: toolProvider != null ? stableToolProvider.current : null,
		onError
	}, renderDeep);
	useEffect(() => {
		onParseResultRef.current?.(parseResult);
	}, [parseResult]);
	if (!result?.root) return null;
	return /* @__PURE__ */ jsx(OpenUIContext.Provider, {
		value: contextValue,
		children: /* @__PURE__ */ jsxs("div", {
			style: { position: "relative" },
			children: [isQueryLoading && (queryLoader ?? /* @__PURE__ */ jsx(DefaultQueryLoader, {})), /* @__PURE__ */ jsx("div", {
				style: {
					opacity: isQueryLoading ? .7 : 1,
					transition: "opacity 0.2s ease"
				},
				children: /* @__PURE__ */ jsx(RenderNode, { node: result.root })
			})]
		})
	});
}
//#endregion
//#region src/runtime/reactive.ts
/**
* Mark a schema prop as reactive so runtime evaluation can preserve $bindings.
*
* The widened return type carries the eventual value shape into helpers like
* `useStateField()`. The actual bound value is still resolved at render time.
*/
function reactive(schema) {
	markReactive(schema);
	return schema;
}
//#endregion
//#region src/hooks/useStateField.ts
function useStateField(name, value) {
	const ctx = useOpenUI();
	const formName = useFormName();
	return resolveStateField(name, value, ctx.store ?? null, ctx.evaluationContext ?? null, (fieldName) => ctx.getFieldValue(formName, fieldName), (fieldName, nextValue) => ctx.setFieldValue(formName, void 0, fieldName, nextValue));
}
//#endregion
//#region src/hooks/useFormValidation.ts
const FormValidationContext = createContext(null);
function useFormValidation() {
	return useContext(FormValidationContext);
}
function useCreateFormValidation() {
	const [errors, setErrors] = useState({});
	const errorsRef = useRef(errors);
	errorsRef.current = errors;
	const fieldsRef = useRef({});
	const getFieldError = useCallback((name) => errorsRef.current[name], []);
	const validateField = useCallback((name, value, rules) => {
		const error = validate$1(value, rules);
		setErrors((prev) => {
			if (prev[name] === error) return prev;
			return {
				...prev,
				[name]: error
			};
		});
		return !error;
	}, []);
	const registerField = useCallback((name, rules, getValue) => {
		fieldsRef.current[name] = {
			rules,
			getValue
		};
	}, []);
	const unregisterField = useCallback((name) => {
		delete fieldsRef.current[name];
	}, []);
	const validateForm = useCallback(() => {
		let allValid = true;
		const newErrors = {};
		for (const [name, reg] of Object.entries(fieldsRef.current)) {
			let value = reg.getValue();
			if (value != null && typeof value === "object" && "value" in value && "componentType" in value) value = value.value;
			const error = validate$1(value, reg.rules);
			newErrors[name] = error;
			if (error) allValid = false;
		}
		setErrors(newErrors);
		return allValid;
	}, []);
	const clearFieldError = useCallback((name) => {
		setErrors((prev) => {
			if (prev[name] === void 0) return prev;
			return {
				...prev,
				[name]: void 0
			};
		});
	}, []);
	return useMemo(() => ({
		errors,
		getFieldError,
		validateField,
		registerField,
		unregisterField,
		validateForm,
		clearFieldError
	}), [
		errors,
		getFieldError,
		validateField,
		registerField,
		unregisterField,
		validateForm,
		clearFieldError
	]);
}
//#endregion
export { ACTION_STEPS, BuiltinActionType, FormNameContext, FormValidationContext, Renderer, ToolNotFoundError, builtInValidators, createLibrary, createParser, createStreamingParser, defineComponent, extractToolResult, generatePrompt, isReactiveAssign, mergeStatements, parseRules, parseStructuredRules, reactive, tagSchemaId, useCreateFormValidation, useFormName, useFormValidation, useGetFieldValue, useIsStreaming, useRenderNode, useSetDefaultValue, useSetFieldValue, useStateField, useTriggerAction, validate };
