// @openuidev/react-ui/dist/components/Carousel/Carousel.js
import { jsx as _jsx } from "react/jsx-runtime";
import clsx from "clsx";
import { createContext, forwardRef, useCallback, useContext, useEffect, useImperativeHandle, useMemo, useRef, useState, } from "react";
import { IconButton } from "../IconButton";
const CarouselContext = createContext(null);
const useCarousel = () => {
    const context = useContext(CarouselContext);
    if (!context)
        throw new Error("useCarousel must be used within a Carousel");
    return context;
};
export const Carousel = forwardRef(({ itemsToScroll = 1, noSnap, showButtons = true, variant = "card", className, children, onScrollLeftEnabled, onScrollRightEnabled, ...props }, ref) => {
    const scrollDivRef = useRef(null);
    const [isPrevVisible, setIsPrevVisible] = useState(false);
    const [isNextVisible, setIsNextVisible] = useState(false);
    const scroll = useCallback((direction) => {
        if (!scrollDivRef.current) {
            return;
        }
        const container = scrollDivRef.current;
        const items = noSnap
            ? Array.from(container.children[0]?.children ?? [])
            : Array.from(container.children);
        if (items.length === 0) {
            return;
        }
        const containerRect = container.getBoundingClientRect();
        const visibleIndex = items.findIndex((child) => {
            const rect = child.getBoundingClientRect();
            return rect.left >= containerRect.left;
        });
        let currentIndex = visibleIndex;
        if (visibleIndex === -1) {
            // Scrolled to the far right, so the last item is the current one
            currentIndex = items.length - 1;
        }
        const targetIndex = direction === "left"
            ? Math.max(0, currentIndex - itemsToScroll)
            : Math.min(items.length - 1, currentIndex + itemsToScroll);
        const targetElement = items[targetIndex];
        if (targetElement) {
            container.scrollTo({
                left: targetElement.offsetLeft,
                behavior: "smooth",
            });
        }
    }, [noSnap, itemsToScroll]);
    useEffect(() => {
        if (!scrollDivRef.current)
            return;
        const container = scrollDivRef.current;
        const handleScroll = () => {
            const canScrollLeft = container.scrollLeft > 0;
            const canScrollRight = Math.ceil(container.scrollLeft) + container.offsetWidth < container.scrollWidth;
            setIsPrevVisible(canScrollLeft);
            setIsNextVisible(canScrollRight);
        };
        handleScroll();
        const resizeObserver = new ResizeObserver(handleScroll);
        resizeObserver.observe(container);
        const mutationObserver = new MutationObserver(handleScroll);
        mutationObserver.observe(container, { childList: true, subtree: true });
        container.addEventListener("scroll", handleScroll);
        return () => {
            container.removeEventListener("scroll", handleScroll);
            resizeObserver.disconnect();
            mutationObserver.disconnect();
        };
    }, []);
    useEffect(() => {
        onScrollLeftEnabled?.(isPrevVisible);
    }, [isPrevVisible, onScrollLeftEnabled]);
    useEffect(() => {
        onScrollRightEnabled?.(isNextVisible);
    }, [isNextVisible, onScrollRightEnabled]);
    useImperativeHandle(ref, () => {
        return {
            scroll,
            scrollDivRef,
        };
    }, [scroll]);
    const contextValue = useMemo(() => ({
        scrollDivRef,
        scroll,
        itemsToScroll,
        noSnap,
        showButtons,
        variant,
        isPrevVisible,
        isNextVisible,
    }), [
        scrollDivRef,
        scroll,
        itemsToScroll,
        noSnap,
        showButtons,
        variant,
        isPrevVisible,
        isNextVisible,
    ]);
    return (_jsx(CarouselContext.Provider, { value: contextValue, children: _jsx("div", { className: clsx("openui-carousel", `openui-carousel--${variant}`, className), ...props, children: children }) }));
});
export const CarouselContent = forwardRef(({ className, children, ...props }, _ref) => {
    const { scrollDivRef, noSnap, isPrevVisible, isNextVisible } = useCarousel();
    const content = noSnap ? (_jsx("div", { className: "openui-carousel-content-wrapper", children: children })) : (children);
    return (_jsx("div", { ref: scrollDivRef, className: clsx("openui-carousel-content", {
            "openui-carousel-content--mask-left": isPrevVisible,
            "openui-carousel-content--mask-right": isNextVisible,
        }, className), ...props, children: content }));
});
export const CarouselItem = forwardRef(({ className, children, ...props }, ref) => (_jsx("div", { ref: ref, className: clsx("openui-carousel-item", className), ...props, children: children })));
export const CarouselPrevious = forwardRef(({ className, style, ...props }, ref) => {
    const { scroll, showButtons, isPrevVisible } = useCarousel();
    if (!isPrevVisible || !showButtons)
        return null;
    return (_jsx("div", { className: clsx("openui-carousel-button openui-carousel-button-left", className), children: _jsx(IconButton, { ref: ref, shape: "square", variant: "secondary", size: "small", onClick: () => scroll("left"), style: style, ...props }) }));
});
export const CarouselNext = forwardRef(({ className, style, ...props }, ref) => {
    const { scroll, showButtons, isNextVisible } = useCarousel();
    if (!isNextVisible || !showButtons)
        return null;
    return (_jsx("div", { className: clsx("openui-carousel-button openui-carousel-button-right", className), children: _jsx(IconButton, { ref: ref, shape: "square", variant: "secondary", size: "small", onClick: () => scroll("right"), style: style, ...props }) }));
});
