// @openuidev/react-ui/dist/components/Carousel/index.cjs
Object.defineProperty(exports, Symbol.toStringTag, { value: "Module" });
const require_chunk = require("../../chunk-CKQMccvm.cjs");
const require_components_IconButton_index = require("../IconButton/index.cjs");
let clsx = require("clsx");
clsx = require_chunk.__toESM(clsx);
let react = require("react");
let react_jsx_runtime = require("react/jsx-runtime");
//#region src/components/Carousel/Carousel.tsx
const CarouselContext = (0, react.createContext)(null);
const useCarousel = () => {
	const context = (0, react.useContext)(CarouselContext);
	if (!context) throw new Error("useCarousel must be used within a Carousel");
	return context;
};
const Carousel = (0, react.forwardRef)(({ itemsToScroll = 1, noSnap, showButtons = true, variant = "card", className, children, onScrollLeftEnabled, onScrollRightEnabled, ...props }, ref) => {
	const scrollDivRef = (0, react.useRef)(null);
	const [isPrevVisible, setIsPrevVisible] = (0, react.useState)(false);
	const [isNextVisible, setIsNextVisible] = (0, react.useState)(false);
	const scroll = (0, react.useCallback)((direction) => {
		if (!scrollDivRef.current) return;
		const container = scrollDivRef.current;
		const items = noSnap ? Array.from(container.children[0]?.children ?? []) : Array.from(container.children);
		if (items.length === 0) return;
		const containerRect = container.getBoundingClientRect();
		const visibleIndex = items.findIndex((child) => {
			return child.getBoundingClientRect().left >= containerRect.left;
		});
		let currentIndex = visibleIndex;
		if (visibleIndex === -1) currentIndex = items.length - 1;
		const targetElement = items[direction === "left" ? Math.max(0, currentIndex - itemsToScroll) : Math.min(items.length - 1, currentIndex + itemsToScroll)];
		if (targetElement) container.scrollTo({
			left: targetElement.offsetLeft,
			behavior: "smooth"
		});
	}, [noSnap, itemsToScroll]);
	(0, react.useEffect)(() => {
		if (!scrollDivRef.current) return;
		const container = scrollDivRef.current;
		const handleScroll = () => {
			const canScrollLeft = container.scrollLeft > 0;
			const canScrollRight = Math.ceil(container.scrollLeft) + container.offsetWidth < container.scrollWidth;
			setIsPrevVisible(canScrollLeft);
			setIsNextVisible(canScrollRight);
		};
		handleScroll();
		const resizeObserver = new ResizeObserver(handleScroll);
		resizeObserver.observe(container);
		const mutationObserver = new MutationObserver(handleScroll);
		mutationObserver.observe(container, {
			childList: true,
			subtree: true
		});
		container.addEventListener("scroll", handleScroll);
		return () => {
			container.removeEventListener("scroll", handleScroll);
			resizeObserver.disconnect();
			mutationObserver.disconnect();
		};
	}, []);
	(0, react.useEffect)(() => {
		onScrollLeftEnabled?.(isPrevVisible);
	}, [isPrevVisible, onScrollLeftEnabled]);
	(0, react.useEffect)(() => {
		onScrollRightEnabled?.(isNextVisible);
	}, [isNextVisible, onScrollRightEnabled]);
	(0, react.useImperativeHandle)(ref, () => {
		return {
			scroll,
			scrollDivRef
		};
	}, [scroll]);
	const contextValue = (0, react.useMemo)(() => ({
		scrollDivRef,
		scroll,
		itemsToScroll,
		noSnap,
		showButtons,
		variant,
		isPrevVisible,
		isNextVisible
	}), [
		scrollDivRef,
		scroll,
		itemsToScroll,
		noSnap,
		showButtons,
		variant,
		isPrevVisible,
		isNextVisible
	]);
	return /* @__PURE__ */ (0, react_jsx_runtime.jsx)(CarouselContext.Provider, {
		value: contextValue,
		children: /* @__PURE__ */ (0, react_jsx_runtime.jsx)("div", {
			className: (0, clsx.default)("openui-carousel", `openui-carousel--${variant}`, className),
			...props,
			children
		})
	});
});
const CarouselContent = (0, react.forwardRef)(({ className, children, ...props }, _ref) => {
	const { scrollDivRef, noSnap, isPrevVisible, isNextVisible } = useCarousel();
	const content = noSnap ? /* @__PURE__ */ (0, react_jsx_runtime.jsx)("div", {
		className: "openui-carousel-content-wrapper",
		children
	}) : children;
	return /* @__PURE__ */ (0, react_jsx_runtime.jsx)("div", {
		ref: scrollDivRef,
		className: (0, clsx.default)("openui-carousel-content", {
			"openui-carousel-content--mask-left": isPrevVisible,
			"openui-carousel-content--mask-right": isNextVisible
		}, className),
		...props,
		children: content
	});
});
const CarouselItem = (0, react.forwardRef)(({ className, children, ...props }, ref) => /* @__PURE__ */ (0, react_jsx_runtime.jsx)("div", {
	ref,
	className: (0, clsx.default)("openui-carousel-item", className),
	...props,
	children
}));
const CarouselPrevious = (0, react.forwardRef)(({ className, style, ...props }, ref) => {
	const { scroll, showButtons, isPrevVisible } = useCarousel();
	if (!isPrevVisible || !showButtons) return null;
	return /* @__PURE__ */ (0, react_jsx_runtime.jsx)("div", {
		className: (0, clsx.default)("openui-carousel-button openui-carousel-button-left", className),
		children: /* @__PURE__ */ (0, react_jsx_runtime.jsx)(require_components_IconButton_index.IconButton, {
			ref,
			shape: "square",
			variant: "secondary",
			size: "small",
			onClick: () => scroll("left"),
			style,
			...props
		})
	});
});
const CarouselNext = (0, react.forwardRef)(({ className, style, ...props }, ref) => {
	const { scroll, showButtons, isNextVisible } = useCarousel();
	if (!isNextVisible || !showButtons) return null;
	return /* @__PURE__ */ (0, react_jsx_runtime.jsx)("div", {
		className: (0, clsx.default)("openui-carousel-button openui-carousel-button-right", className),
		children: /* @__PURE__ */ (0, react_jsx_runtime.jsx)(require_components_IconButton_index.IconButton, {
			ref,
			shape: "square",
			variant: "secondary",
			size: "small",
			onClick: () => scroll("right"),
			style,
			...props
		})
	});
});
//#endregion
exports.Carousel = Carousel;
exports.CarouselContent = CarouselContent;
exports.CarouselItem = CarouselItem;
exports.CarouselNext = CarouselNext;
exports.CarouselPrevious = CarouselPrevious;
