// @openuidev/react-ui/dist/genui-lib/ListBlock/index.js
"use client";
import { jsx as _jsx } from "react/jsx-runtime";
import { defineComponent, useTriggerAction } from "@openuidev/react-lang";
import { ChevronRight } from "lucide-react";
import { z } from "zod/v4";
import { ListBlock as OpenUIListBlock } from "../../components/ListBlock";
import { ListItem as OpenUIListItem } from "../../components/ListItem";
import { ListItem } from "../ListItem";
export const ListBlock = defineComponent({
    name: "ListBlock",
    props: z.object({
        items: z.array(ListItem.ref),
        variant: z.enum(["number", "image"]).optional(),
    }),
    description: "A list of items with number or image indicators. Each item can optionally have an action.",
    component: ({ props }) => {
        const triggerAction = useTriggerAction();
        const items = (props.items ?? []);
        const variant = props.variant ?? "number";
        const listHasSubtitle = items.some((item) => !!item?.props?.subtitle);
        return (_jsx(OpenUIListBlock, { variant: variant, children: items.map((item, index) => {
                const title = String(item?.props?.title ?? "");
                const subtitle = item?.props?.subtitle ? String(item.props.subtitle) : undefined;
                const image = item?.props?.image;
                const actionLabel = item?.props?.actionLabel ? String(item.props.actionLabel) : undefined;
                const action = item?.props?.action;
                const hasAction = !!action;
                const handleClick = hasAction
                    ? () => triggerAction(title, undefined, action)
                    : undefined;
                return (_jsx(OpenUIListItem, { title: title, subtitle: subtitle, listHasSubtitle: listHasSubtitle, image: variant === "image" ? image : undefined, actionLabel: hasAction ? actionLabel : undefined, actionIcon: hasAction ? _jsx(ChevronRight, { size: 16 }) : undefined, onClick: handleClick }, index));
            }) }));
    },
});
