// @ag-ui/core/dist/index.mjs
import { z } from "zod";

//#region src/types.ts
const FunctionCallSchema = z.object({
	name: z.string(),
	arguments: z.string()
});
const ToolCallSchema = z.object({
	id: z.string(),
	type: z.literal("function"),
	function: FunctionCallSchema,
	encryptedValue: z.string().optional()
});
const BaseMessageSchema = z.object({
	id: z.string(),
	role: z.string(),
	content: z.string().optional(),
	name: z.string().optional(),
	encryptedValue: z.string().optional()
});
const TextInputContentSchema = z.object({
	type: z.literal("text"),
	text: z.string()
});
const BinaryInputContentObjectSchema = z.object({
	type: z.literal("binary"),
	mimeType: z.string(),
	id: z.string().optional(),
	url: z.string().optional(),
	data: z.string().optional(),
	filename: z.string().optional()
});
const ensureBinaryPayload = (value, ctx) => {
	if (!value.id && !value.url && !value.data) ctx.addIssue({
		code: z.ZodIssueCode.custom,
		message: "BinaryInputContent requires at least one of id, url, or data.",
		path: ["id"]
	});
};
const BinaryInputContentSchema = BinaryInputContentObjectSchema.superRefine((value, ctx) => {
	ensureBinaryPayload(value, ctx);
});
const InputContentBaseSchema = z.discriminatedUnion("type", [TextInputContentSchema, BinaryInputContentObjectSchema]);
const InputContentSchema = InputContentBaseSchema.superRefine((value, ctx) => {
	if (value.type === "binary") ensureBinaryPayload(value, ctx);
});
const DeveloperMessageSchema = BaseMessageSchema.extend({
	role: z.literal("developer"),
	content: z.string()
});
const SystemMessageSchema = BaseMessageSchema.extend({
	role: z.literal("system"),
	content: z.string()
});
const AssistantMessageSchema = BaseMessageSchema.extend({
	role: z.literal("assistant"),
	content: z.string().optional(),
	toolCalls: z.array(ToolCallSchema).optional()
});
const UserMessageSchema = BaseMessageSchema.extend({
	role: z.literal("user"),
	content: z.union([z.string(), z.array(InputContentSchema)])
});
const ToolMessageSchema = z.object({
	id: z.string(),
	content: z.string(),
	role: z.literal("tool"),
	toolCallId: z.string(),
	error: z.string().optional(),
	encryptedValue: z.string().optional()
});
const ActivityMessageSchema = z.object({
	id: z.string(),
	role: z.literal("activity"),
	activityType: z.string(),
	content: z.record(z.any())
});
const ReasoningMessageSchema = z.object({
	id: z.string(),
	role: z.literal("reasoning"),
	content: z.string(),
	encryptedValue: z.string().optional()
});
const MessageSchema = z.discriminatedUnion("role", [
	DeveloperMessageSchema,
	SystemMessageSchema,
	AssistantMessageSchema,
	UserMessageSchema,
	ToolMessageSchema,
	ActivityMessageSchema,
	ReasoningMessageSchema
]);
const RoleSchema = z.union([
	z.literal("developer"),
	z.literal("system"),
	z.literal("assistant"),
	z.literal("user"),
	z.literal("tool"),
	z.literal("activity"),
	z.literal("reasoning")
]);
const ContextSchema = z.object({
	description: z.string(),
	value: z.string()
});
const ToolSchema = z.object({
	name: z.string(),
	description: z.string(),
	parameters: z.any()
});
const RunAgentInputSchema = z.object({
	threadId: z.string(),
	runId: z.string(),
	parentRunId: z.string().optional(),
	state: z.any(),
	messages: z.array(MessageSchema),
	tools: z.array(ToolSchema),
	context: z.array(ContextSchema),
	forwardedProps: z.any()
});
const StateSchema = z.any();
var AGUIError = class extends Error {
	constructor(message) {
		super(message);
	}
};
var AGUIConnectNotImplementedError = class extends AGUIError {
	constructor() {
		super("Connect not implemented. This method is not supported by the current agent.");
	}
};

//#endregion
//#region src/events.ts
const TextMessageRoleSchema = z.union([
	z.literal("developer"),
	z.literal("system"),
	z.literal("assistant"),
	z.literal("user")
]);
let EventType = /* @__PURE__ */ function(EventType) {
	EventType["TEXT_MESSAGE_START"] = "TEXT_MESSAGE_START";
	EventType["TEXT_MESSAGE_CONTENT"] = "TEXT_MESSAGE_CONTENT";
	EventType["TEXT_MESSAGE_END"] = "TEXT_MESSAGE_END";
	EventType["TEXT_MESSAGE_CHUNK"] = "TEXT_MESSAGE_CHUNK";
	EventType["TOOL_CALL_START"] = "TOOL_CALL_START";
	EventType["TOOL_CALL_ARGS"] = "TOOL_CALL_ARGS";
	EventType["TOOL_CALL_END"] = "TOOL_CALL_END";
	EventType["TOOL_CALL_CHUNK"] = "TOOL_CALL_CHUNK";
	EventType["TOOL_CALL_RESULT"] = "TOOL_CALL_RESULT";
	/**
	* @deprecated Use REASONING_START instead. Will be removed in 1.0.0.
	*/
	EventType["THINKING_START"] = "THINKING_START";
	/**
	* @deprecated Use REASONING_END instead. Will be removed in 1.0.0.
	*/
	EventType["THINKING_END"] = "THINKING_END";
	/**
	* @deprecated Use REASONING_MESSAGE_START instead. Will be removed in 1.0.0.
	*/
	EventType["THINKING_TEXT_MESSAGE_START"] = "THINKING_TEXT_MESSAGE_START";
	/**
	* @deprecated Use REASONING_MESSAGE_CONTENT instead. Will be removed in 1.0.0.
	*/
	EventType["THINKING_TEXT_MESSAGE_CONTENT"] = "THINKING_TEXT_MESSAGE_CONTENT";
	/**
	* @deprecated Use REASONING_MESSAGE_END instead. Will be removed in 1.0.0.
	*/
	EventType["THINKING_TEXT_MESSAGE_END"] = "THINKING_TEXT_MESSAGE_END";
	EventType["STATE_SNAPSHOT"] = "STATE_SNAPSHOT";
	EventType["STATE_DELTA"] = "STATE_DELTA";
	EventType["MESSAGES_SNAPSHOT"] = "MESSAGES_SNAPSHOT";
	EventType["ACTIVITY_SNAPSHOT"] = "ACTIVITY_SNAPSHOT";
	EventType["ACTIVITY_DELTA"] = "ACTIVITY_DELTA";
	EventType["RAW"] = "RAW";
	EventType["CUSTOM"] = "CUSTOM";
	EventType["RUN_STARTED"] = "RUN_STARTED";
	EventType["RUN_FINISHED"] = "RUN_FINISHED";
	EventType["RUN_ERROR"] = "RUN_ERROR";
	EventType["STEP_STARTED"] = "STEP_STARTED";
	EventType["STEP_FINISHED"] = "STEP_FINISHED";
	EventType["REASONING_START"] = "REASONING_START";
	EventType["REASONING_MESSAGE_START"] = "REASONING_MESSAGE_START";
	EventType["REASONING_MESSAGE_CONTENT"] = "REASONING_MESSAGE_CONTENT";
	EventType["REASONING_MESSAGE_END"] = "REASONING_MESSAGE_END";
	EventType["REASONING_MESSAGE_CHUNK"] = "REASONING_MESSAGE_CHUNK";
	EventType["REASONING_END"] = "REASONING_END";
	EventType["REASONING_ENCRYPTED_VALUE"] = "REASONING_ENCRYPTED_VALUE";
	return EventType;
}({});
const BaseEventSchema = z.object({
	type: z.nativeEnum(EventType),
	timestamp: z.number().optional(),
	rawEvent: z.any().optional()
}).passthrough();
const TextMessageStartEventSchema = BaseEventSchema.extend({
	type: z.literal(EventType.TEXT_MESSAGE_START),
	messageId: z.string(),
	role: TextMessageRoleSchema.default("assistant")
});
const TextMessageContentEventSchema = BaseEventSchema.extend({
	type: z.literal(EventType.TEXT_MESSAGE_CONTENT),
	messageId: z.string(),
	delta: z.string().refine((s) => s.length > 0, "Delta must not be an empty string")
});
const TextMessageEndEventSchema = BaseEventSchema.extend({
	type: z.literal(EventType.TEXT_MESSAGE_END),
	messageId: z.string()
});
const TextMessageChunkEventSchema = BaseEventSchema.extend({
	type: z.literal(EventType.TEXT_MESSAGE_CHUNK),
	messageId: z.string().optional(),
	role: TextMessageRoleSchema.optional(),
	delta: z.string().optional()
});
/**
* @deprecated Use ReasoningMessageStartEventSchema instead. Will be removed in 1.0.0.
*/
const ThinkingTextMessageStartEventSchema = BaseEventSchema.extend({ type: z.literal(EventType.THINKING_TEXT_MESSAGE_START) });
/**
* @deprecated Use ReasoningMessageContentEventSchema instead. Will be removed in 1.0.0.
*/
const ThinkingTextMessageContentEventSchema = TextMessageContentEventSchema.omit({
	messageId: true,
	type: true
}).extend({ type: z.literal(EventType.THINKING_TEXT_MESSAGE_CONTENT) });
/**
* @deprecated Use ReasoningMessageEndEventSchema instead. Will be removed in 1.0.0.
*/
const ThinkingTextMessageEndEventSchema = BaseEventSchema.extend({ type: z.literal(EventType.THINKING_TEXT_MESSAGE_END) });
const ToolCallStartEventSchema = BaseEventSchema.extend({
	type: z.literal(EventType.TOOL_CALL_START),
	toolCallId: z.string(),
	toolCallName: z.string(),
	parentMessageId: z.string().optional()
});
const ToolCallArgsEventSchema = BaseEventSchema.extend({
	type: z.literal(EventType.TOOL_CALL_ARGS),
	toolCallId: z.string(),
	delta: z.string()
});
const ToolCallEndEventSchema = BaseEventSchema.extend({
	type: z.literal(EventType.TOOL_CALL_END),
	toolCallId: z.string()
});
const ToolCallResultEventSchema = BaseEventSchema.extend({
	messageId: z.string(),
	type: z.literal(EventType.TOOL_CALL_RESULT),
	toolCallId: z.string(),
	content: z.string(),
	role: z.literal("tool").optional()
});
const ToolCallChunkEventSchema = BaseEventSchema.extend({
	type: z.literal(EventType.TOOL_CALL_CHUNK),
	toolCallId: z.string().optional(),
	toolCallName: z.string().optional(),
	parentMessageId: z.string().optional(),
	delta: z.string().optional()
});
/**
* @deprecated Use ReasoningStartEventSchema instead. Will be removed in 1.0.0.
*/
const ThinkingStartEventSchema = BaseEventSchema.extend({
	type: z.literal(EventType.THINKING_START),
	title: z.string().optional()
});
/**
* @deprecated Use ReasoningEndEventSchema instead. Will be removed in 1.0.0.
*/
const ThinkingEndEventSchema = BaseEventSchema.extend({ type: z.literal(EventType.THINKING_END) });
const StateSnapshotEventSchema = BaseEventSchema.extend({
	type: z.literal(EventType.STATE_SNAPSHOT),
	snapshot: StateSchema
});
const StateDeltaEventSchema = BaseEventSchema.extend({
	type: z.literal(EventType.STATE_DELTA),
	delta: z.array(z.any())
});
const MessagesSnapshotEventSchema = BaseEventSchema.extend({
	type: z.literal(EventType.MESSAGES_SNAPSHOT),
	messages: z.array(MessageSchema)
});
const ActivitySnapshotEventSchema = BaseEventSchema.extend({
	type: z.literal(EventType.ACTIVITY_SNAPSHOT),
	messageId: z.string(),
	activityType: z.string(),
	content: z.record(z.any()),
	replace: z.boolean().optional().default(true)
});
const ActivityDeltaEventSchema = BaseEventSchema.extend({
	type: z.literal(EventType.ACTIVITY_DELTA),
	messageId: z.string(),
	activityType: z.string(),
	patch: z.array(z.any())
});
const RawEventSchema = BaseEventSchema.extend({
	type: z.literal(EventType.RAW),
	event: z.any(),
	source: z.string().optional()
});
const CustomEventSchema = BaseEventSchema.extend({
	type: z.literal(EventType.CUSTOM),
	name: z.string(),
	value: z.any()
});
const RunStartedEventSchema = BaseEventSchema.extend({
	type: z.literal(EventType.RUN_STARTED),
	threadId: z.string(),
	runId: z.string(),
	parentRunId: z.string().optional(),
	input: RunAgentInputSchema.optional()
});
const RunFinishedEventSchema = BaseEventSchema.extend({
	type: z.literal(EventType.RUN_FINISHED),
	threadId: z.string(),
	runId: z.string(),
	result: z.any().optional()
});
const RunErrorEventSchema = BaseEventSchema.extend({
	type: z.literal(EventType.RUN_ERROR),
	message: z.string(),
	code: z.string().optional()
});
const StepStartedEventSchema = BaseEventSchema.extend({
	type: z.literal(EventType.STEP_STARTED),
	stepName: z.string()
});
const StepFinishedEventSchema = BaseEventSchema.extend({
	type: z.literal(EventType.STEP_FINISHED),
	stepName: z.string()
});
const ReasoningEncryptedValueSubtypeSchema = z.union([z.literal("tool-call"), z.literal("message")]);
const ReasoningStartEventSchema = BaseEventSchema.extend({
	type: z.literal(EventType.REASONING_START),
	messageId: z.string()
});
const ReasoningMessageStartEventSchema = BaseEventSchema.extend({
	type: z.literal(EventType.REASONING_MESSAGE_START),
	messageId: z.string(),
	role: z.literal("reasoning")
});
const ReasoningMessageContentEventSchema = BaseEventSchema.extend({
	type: z.literal(EventType.REASONING_MESSAGE_CONTENT),
	messageId: z.string(),
	delta: z.string().refine((s) => s.length > 0, "Delta must not be an empty string")
});
const ReasoningMessageEndEventSchema = BaseEventSchema.extend({
	type: z.literal(EventType.REASONING_MESSAGE_END),
	messageId: z.string()
});
const ReasoningMessageChunkEventSchema = BaseEventSchema.extend({
	type: z.literal(EventType.REASONING_MESSAGE_CHUNK),
	messageId: z.string().optional(),
	delta: z.string().optional()
});
const ReasoningEndEventSchema = BaseEventSchema.extend({
	type: z.literal(EventType.REASONING_END),
	messageId: z.string()
});
const ReasoningEncryptedValueEventSchema = BaseEventSchema.extend({
	type: z.literal(EventType.REASONING_ENCRYPTED_VALUE),
	subtype: ReasoningEncryptedValueSubtypeSchema,
	entityId: z.string(),
	encryptedValue: z.string()
});
const EventSchemas = z.discriminatedUnion("type", [
	TextMessageStartEventSchema,
	TextMessageContentEventSchema,
	TextMessageEndEventSchema,
	TextMessageChunkEventSchema,
	ThinkingStartEventSchema,
	ThinkingEndEventSchema,
	ThinkingTextMessageStartEventSchema,
	ThinkingTextMessageContentEventSchema,
	ThinkingTextMessageEndEventSchema,
	ToolCallStartEventSchema,
	ToolCallArgsEventSchema,
	ToolCallEndEventSchema,
	ToolCallChunkEventSchema,
	ToolCallResultEventSchema,
	StateSnapshotEventSchema,
	StateDeltaEventSchema,
	MessagesSnapshotEventSchema,
	ActivitySnapshotEventSchema,
	ActivityDeltaEventSchema,
	RawEventSchema,
	CustomEventSchema,
	RunStartedEventSchema,
	RunFinishedEventSchema,
	RunErrorEventSchema,
	StepStartedEventSchema,
	StepFinishedEventSchema,
	ReasoningStartEventSchema,
	ReasoningMessageStartEventSchema,
	ReasoningMessageContentEventSchema,
	ReasoningMessageEndEventSchema,
	ReasoningMessageChunkEventSchema,
	ReasoningEndEventSchema,
	ReasoningEncryptedValueEventSchema
]);

//#endregion
export { AGUIConnectNotImplementedError, AGUIError, ActivityDeltaEventSchema, ActivityMessageSchema, ActivitySnapshotEventSchema, AssistantMessageSchema, BaseEventSchema, BaseMessageSchema, BinaryInputContentSchema, ContextSchema, CustomEventSchema, DeveloperMessageSchema, EventSchemas, EventType, FunctionCallSchema, InputContentSchema, MessageSchema, MessagesSnapshotEventSchema, RawEventSchema, ReasoningEncryptedValueEventSchema, ReasoningEncryptedValueSubtypeSchema, ReasoningEndEventSchema, ReasoningMessageChunkEventSchema, ReasoningMessageContentEventSchema, ReasoningMessageEndEventSchema, ReasoningMessageSchema, ReasoningMessageStartEventSchema, ReasoningStartEventSchema, RoleSchema, RunAgentInputSchema, RunErrorEventSchema, RunFinishedEventSchema, RunStartedEventSchema, StateDeltaEventSchema, StateSchema, StateSnapshotEventSchema, StepFinishedEventSchema, StepStartedEventSchema, SystemMessageSchema, TextInputContentSchema, TextMessageChunkEventSchema, TextMessageContentEventSchema, TextMessageEndEventSchema, TextMessageStartEventSchema, ThinkingEndEventSchema, ThinkingStartEventSchema, ThinkingTextMessageContentEventSchema, ThinkingTextMessageEndEventSchema, ThinkingTextMessageStartEventSchema, ToolCallArgsEventSchema, ToolCallChunkEventSchema, ToolCallEndEventSchema, ToolCallResultEventSchema, ToolCallSchema, ToolCallStartEventSchema, ToolMessageSchema, ToolSchema, UserMessageSchema };
