// @openuidev/react-ui/dist/components/BottomTray/stories/BottomTray.stories.js
import { jsx as _jsx, jsxs as _jsxs, Fragment as _Fragment } from "react/jsx-runtime";
import { ChatProvider } from "@openuidev/react-headless";
import { MessageSquare, Sparkles, Zap } from "lucide-react";
import { useState } from "react";
import { Composer, Container, ConversationStarter, Header, MessageLoading, Messages, ScrollArea, ThreadContainer, Trigger, WelcomeScreen, } from "../../BottomTray";
// @ts-ignore
import styles from "./style.module.css";
import logoUrl from "./thesysdev_logo.jpeg";
function mockSSEResponse(text, delayMs = 500) {
    return new Promise((resolve) => {
        setTimeout(() => {
            const events = `data: ${JSON.stringify({ type: "TEXT_MESSAGE_CONTENT", delta: text })}\n\ndata: [DONE]\n\n`;
            const stream = new ReadableStream({
                start(controller) {
                    controller.enqueue(new TextEncoder().encode(events));
                    controller.close();
                },
            });
            resolve(new Response(stream));
        }, delayMs);
    });
}
export default {
    title: "Components/BottomTray",
    tags: ["dev"],
    argTypes: {
        defaultOpen: {
            control: "boolean",
            description: "Whether the tray starts open",
        },
        variant: {
            control: "select",
            options: ["short", "long"],
            description: "Conversation starter variant",
        },
    },
};
const BottomTrayStory = ({ defaultOpen = false, variant = "short", }) => {
    const [isOpen, setIsOpen] = useState(defaultOpen);
    return (_jsxs("div", { className: styles.container, children: [_jsxs("div", { className: styles.content, children: [_jsx("h1", { children: "Bottom Tray Example (Composable)" }), _jsxs("p", { children: ["The chat interface uses a ", _jsx("strong", { children: "composition pattern" }), ". The trigger button is separate from the container, giving you full control over placement and styling."] }), _jsxs("p", { children: [_jsx("strong", { children: "Try it:" }), " Click the pill button (bottom-right) or the custom button below to toggle the tray."] }), _jsxs("button", { onClick: () => setIsOpen(!isOpen), className: styles.toggleButton, children: [isOpen ? "Close" : "Open", " Tray"] })] }), _jsxs(ChatProvider, { processMessage: async ({ threadId, messages, abortController }) => {
                    await new Promise((resolve) => setTimeout(resolve, 1000));
                    return mockSSEResponse("This is a response from the bottom tray assistant!", 0);
                }, fetchThreadList: async () => {
                    await new Promise((resolve) => setTimeout(resolve, 1000));
                    return {
                        threads: [
                            { id: "1", title: "test", createdAt: Date.now() },
                            { id: "2", title: "test 2", createdAt: Date.now() },
                            { id: "3", title: "test 3", createdAt: Date.now() },
                        ],
                    };
                }, createThread: async () => ({
                    id: crypto.randomUUID(),
                    title: "test",
                    createdAt: Date.now(),
                }), deleteThread: async () => { }, updateThread: async (t) => t, loadThread: async (threadId) => {
                    if (!threadId)
                        return [];
                    return [
                        {
                            id: crypto.randomUUID(),
                            role: "user",
                            content: "Hello",
                        },
                        {
                            id: crypto.randomUUID(),
                            role: "assistant",
                            content: "Hello! How can I help you today?",
                        },
                    ];
                }, children: [_jsx(Trigger, { onClick: () => setIsOpen(!isOpen), isOpen: isOpen }), _jsx(Container, { logoUrl: logoUrl, agentName: "OpenUI Assistant", isOpen: isOpen, children: _jsxs(ThreadContainer, { children: [_jsx(Header, { onMinimize: () => setIsOpen(false) }), _jsx(ScrollArea, { children: _jsx(Messages, { loader: _jsx(MessageLoading, {}) }) }), _jsx(ConversationStarter, { variant: variant, starters: [
                                        {
                                            displayText: "Tell me about my portfolio",
                                            prompt: "Tell me about the latest stock market trends and how they affect my portfolio",
                                            icon: _jsx(Sparkles, { size: 16 }),
                                        },
                                        {
                                            displayText: "Who is the president of Venezuela and where is he currently located (icon was not passed)",
                                            prompt: "Who is the president of Venezuela and where is he currently located?",
                                            // icon undefined = shows default lightbulb
                                        },
                                        {
                                            displayText: "Tell me about major stock (no icon with empty fragment)",
                                            prompt: "Tell me about major stock",
                                            icon: _jsx(_Fragment, {}), // Empty fragment = no icon
                                        },
                                    ] }), _jsx(Composer, {})] }) })] })] }));
};
export const Default = {
    args: {
        defaultOpen: false,
        variant: "short",
    },
    render: (args) => _jsx(BottomTrayStory, { ...args }),
};
export const OpenByDefault = {
    args: {
        defaultOpen: true,
        variant: "short",
    },
    render: (args) => _jsx(BottomTrayStory, { ...args }),
};
export const LongVariant = {
    args: {
        defaultOpen: true,
        variant: "long",
    },
    render: (args) => _jsx(BottomTrayStory, { ...args }),
};
// Example with custom trigger
const CustomTriggerStory = ({ defaultOpen = false, variant = "short", }) => {
    const [isOpen, setIsOpen] = useState(defaultOpen);
    return (_jsxs("div", { className: styles.container, children: [_jsxs("div", { className: styles.content, children: [_jsx("h1", { children: "Custom Trigger Example" }), _jsx("p", { children: "Use a fully custom trigger with your own styling and content." })] }), _jsxs(ChatProvider, { processMessage: async () => {
                    await new Promise((resolve) => setTimeout(resolve, 1000));
                    return mockSSEResponse("This is a response from the assistant!", 0);
                }, fetchThreadList: async () => {
                    await new Promise((resolve) => setTimeout(resolve, 1000));
                    return {
                        threads: [
                            { id: "1", title: "test", createdAt: Date.now() },
                            { id: "2", title: "test 2", createdAt: Date.now() },
                        ],
                    };
                }, createThread: async () => ({
                    id: crypto.randomUUID(),
                    title: "test",
                    createdAt: Date.now(),
                }), deleteThread: async () => { }, updateThread: async (t) => t, loadThread: async (threadId) => {
                    if (!threadId)
                        return [];
                    return [
                        { id: crypto.randomUUID(), role: "user", content: "Hello" },
                        {
                            id: crypto.randomUUID(),
                            role: "assistant",
                            content: "Hello! How can I help you today?",
                        },
                    ];
                }, children: [_jsx(Trigger, { onClick: () => setIsOpen(!isOpen), isOpen: isOpen, className: styles.customTrigger, children: "\uD83D\uDCAC Need Help?" }), _jsx(Container, { logoUrl: logoUrl, agentName: "Support", isOpen: isOpen, children: _jsxs(ThreadContainer, { children: [_jsx(Header, { onMinimize: () => setIsOpen(false) }), _jsx(ScrollArea, { children: _jsx(Messages, { loader: _jsx(MessageLoading, {}) }) }), _jsx(ConversationStarter, { variant: variant, starters: [
                                        {
                                            displayText: "Help me understand what features are available and how to get started with this application",
                                            prompt: "Help me understand what features are available and how to get started with this application",
                                            icon: _jsx(Zap, { size: 16 }),
                                        },
                                        {
                                            displayText: "What can you help me with today? I need assistance with multiple tasks",
                                            prompt: "What can you help me with today? I need assistance with multiple tasks",
                                            icon: _jsx(MessageSquare, { size: 16 }),
                                        },
                                        {
                                            displayText: "No icon example - this is a shorter prompt",
                                            prompt: "No icon example - this is a shorter prompt",
                                            icon: _jsx(_Fragment, {}), // Empty fragment = no icon
                                        },
                                    ] }), _jsx(Composer, {})] }) })] })] }));
};
export const CustomTrigger = {
    args: {
        defaultOpen: false,
        variant: "short",
    },
    render: (args) => _jsx(CustomTriggerStory, { ...args }),
};
export const CustomTriggerLongVariant = {
    args: {
        defaultOpen: true,
        variant: "long",
    },
    render: (args) => _jsx(CustomTriggerStory, { ...args }),
};
// Example with WelcomeScreen
const WelcomeScreenStory = ({ defaultOpen = true, variant = "short", }) => {
    const [isOpen, setIsOpen] = useState(defaultOpen);
    return (_jsxs("div", { className: styles.container, children: [_jsxs("div", { className: styles.content, children: [_jsx("h1", { children: "Welcome Screen Example" }), _jsx("p", { children: "This example shows the WelcomeScreen component with title, description, and logo." })] }), _jsxs(ChatProvider, { processMessage: async ({ messages }) => {
                    await new Promise((resolve) => setTimeout(resolve, 1000));
                    const lastUser = messages.filter((m) => m.role === "user").pop();
                    const content = typeof lastUser?.content === "string" ? lastUser.content : "";
                    return mockSSEResponse(`You asked: "${content}"`, 0);
                }, fetchThreadList: async () => ({ threads: [] }), createThread: async () => ({
                    id: crypto.randomUUID(),
                    title: "New Chat",
                    createdAt: Date.now(),
                }), deleteThread: async () => { }, updateThread: async (t) => t, loadThread: async () => [], children: [_jsx(Trigger, { onClick: () => setIsOpen(!isOpen), isOpen: isOpen }), _jsx(Container, { logoUrl: logoUrl, agentName: "OpenUI Assistant", isOpen: isOpen, children: _jsxs(ThreadContainer, { children: [_jsx(Header, { onMinimize: () => setIsOpen(false) }), _jsx(WelcomeScreen, { title: "Hi, I'm OpenUI Assistant", description: "I can help you with questions about your account, products, and more.", image: { url: logoUrl } }), _jsx(ScrollArea, { children: _jsx(Messages, { loader: _jsx(MessageLoading, {}) }) }), _jsx(ConversationStarter, { variant: variant, starters: [
                                        {
                                            displayText: "Help me get started",
                                            prompt: "Help me get started",
                                            icon: _jsx(Sparkles, { size: 16 }),
                                        },
                                        {
                                            displayText: "What can you do?",
                                            prompt: "What can you do?",
                                        },
                                    ] }), _jsx(Composer, {})] }) })] })] }));
};
export const WithWelcomeScreen = {
    args: {
        defaultOpen: true,
        variant: "short",
    },
    render: (args) => _jsx(WelcomeScreenStory, { ...args }),
};
export const WithWelcomeScreenLongVariant = {
    args: {
        defaultOpen: true,
        variant: "long",
    },
    render: (args) => _jsx(WelcomeScreenStory, { ...args }),
};
// Example with custom children in WelcomeScreen
const CustomWelcomeScreenStory = ({ defaultOpen = true, variant = "short", }) => {
    const [isOpen, setIsOpen] = useState(defaultOpen);
    return (_jsxs("div", { className: styles.container, children: [_jsxs("div", { className: styles.content, children: [_jsx("h1", { children: "Custom Welcome Screen Example" }), _jsx("p", { children: "This example shows WelcomeScreen with custom children instead of props." })] }), _jsxs(ChatProvider, { processMessage: async ({ messages }) => {
                    await new Promise((resolve) => setTimeout(resolve, 1000));
                    const lastUser = messages.filter((m) => m.role === "user").pop();
                    const content = typeof lastUser?.content === "string" ? lastUser.content : "";
                    return mockSSEResponse(`You asked: "${content}"`, 0);
                }, fetchThreadList: async () => ({ threads: [] }), createThread: async () => ({
                    id: crypto.randomUUID(),
                    title: "New Chat",
                    createdAt: Date.now(),
                }), deleteThread: async () => { }, updateThread: async (t) => t, loadThread: async () => [], children: [_jsx(Trigger, { onClick: () => setIsOpen(!isOpen), isOpen: isOpen }), _jsx(Container, { logoUrl: logoUrl, agentName: "OpenUI Assistant", isOpen: isOpen, children: _jsxs(ThreadContainer, { children: [_jsx(Header, { onMinimize: () => setIsOpen(false) }), _jsx(WelcomeScreen, { children: _jsxs("div", { style: { textAlign: "center" }, children: [_jsx("div", { style: {
                                                    width: 80,
                                                    height: 80,
                                                    borderRadius: 16,
                                                    background: "linear-gradient(135deg, #667eea 0%, #764ba2 100%)",
                                                    display: "flex",
                                                    alignItems: "center",
                                                    justifyContent: "center",
                                                    margin: "0 auto 16px",
                                                }, children: _jsx(Sparkles, { size: 40, color: "white" }) }), _jsx("h2", { style: { margin: "0 0 8px", fontSize: 20, fontWeight: 600 }, children: "Welcome to AI Assistant" }), _jsx("p", { style: { margin: 0, color: "rgba(0,0,0,0.5)", fontSize: 14 }, children: "Your personal AI helper for all your questions" })] }) }), _jsx(ScrollArea, { children: _jsx(Messages, { loader: _jsx(MessageLoading, {}) }) }), _jsx(ConversationStarter, { variant: variant, starters: [
                                        {
                                            displayText: "Help me get started",
                                            prompt: "Help me get started",
                                            icon: _jsx(Sparkles, { size: 16 }),
                                        },
                                        {
                                            displayText: "What can you do?",
                                            prompt: "What can you do?",
                                        },
                                    ] }), _jsx(Composer, {})] }) })] })] }));
};
export const WithCustomWelcomeScreen = {
    args: {
        defaultOpen: true,
        variant: "short",
    },
    render: (args) => _jsx(CustomWelcomeScreenStory, { ...args }),
};
