// @openuidev/react-ui/dist/components/SectionBlock/index.cjs
Object.defineProperty(exports, Symbol.toStringTag, { value: "Module" });
const require_chunk = require("../../chunk-CKQMccvm.cjs");
const require_components_Separator_index = require("../Separator/index.cjs");
const require_components_IconButton_index = require("../IconButton/index.cjs");
let _radix_ui_react_accordion = require("@radix-ui/react-accordion");
_radix_ui_react_accordion = require_chunk.__toESM(_radix_ui_react_accordion);
let clsx = require("clsx");
clsx = require_chunk.__toESM(clsx);
let lucide_react = require("lucide-react");
let react = require("react");
react = require_chunk.__toESM(react);
let react_jsx_runtime = require("react/jsx-runtime");
//#region src/components/SectionBlock/FoldableSection.tsx
const FoldableSectionRoot = (0, react.forwardRef)(({ className, style, ...props }, ref) => /* @__PURE__ */ (0, react_jsx_runtime.jsx)(_radix_ui_react_accordion.Root, {
	ref,
	className: (0, clsx.default)("openui-foldable-section-root", className),
	style,
	...props
}));
FoldableSectionRoot.displayName = "FoldableSectionRoot";
const FoldableSectionItem = (0, react.forwardRef)(({ className, style, value, ...props }, ref) => /* @__PURE__ */ (0, react_jsx_runtime.jsx)(_radix_ui_react_accordion.Item, {
	ref,
	className: (0, clsx.default)("openui-foldable-section-item", className),
	style,
	value,
	...props
}));
FoldableSectionItem.displayName = "FoldableSectionItem";
const FoldableSectionTrigger = (0, react.forwardRef)(({ className, style, text, ...props }, ref) => /* @__PURE__ */ (0, react_jsx_runtime.jsx)(_radix_ui_react_accordion.Header, {
	className: "openui-foldable-section-header",
	children: /* @__PURE__ */ (0, react_jsx_runtime.jsx)(_radix_ui_react_accordion.Trigger, {
		ref,
		className: (0, clsx.default)("openui-foldable-section-trigger", className),
		style,
		...props,
		children: /* @__PURE__ */ (0, react_jsx_runtime.jsxs)("div", {
			className: "openui-foldable-section-trigger-content-wrapper",
			children: [/* @__PURE__ */ (0, react_jsx_runtime.jsx)(require_components_Separator_index.Separator, { className: "openui-foldable-section-trigger-content-separator" }), /* @__PURE__ */ (0, react_jsx_runtime.jsxs)("div", {
				className: "openui-foldable-section-trigger-content-icon-button-wrapper",
				children: [/* @__PURE__ */ (0, react_jsx_runtime.jsx)(require_components_IconButton_index.IconButton, {
					icon: /* @__PURE__ */ (0, react_jsx_runtime.jsx)(lucide_react.ChevronRight, { className: "openui-foldable-section-trigger-content-icon-button-icon" }),
					size: "3-extra-small",
					variant: "secondary",
					className: "openui-foldable-section-trigger-content-icon-button"
				}), /* @__PURE__ */ (0, react_jsx_runtime.jsx)("div", {
					className: "openui-foldable-section-trigger-content-text",
					children: text
				})]
			})]
		})
	})
}));
FoldableSectionTrigger.displayName = "FoldableSectionTrigger";
const FoldableSectionContent = (0, react.forwardRef)(({ className, style, children, ...props }, ref) => /* @__PURE__ */ (0, react_jsx_runtime.jsx)(_radix_ui_react_accordion.Content, {
	ref,
	className: (0, clsx.default)("openui-foldable-section-content", className),
	style,
	...props,
	children
}));
FoldableSectionContent.displayName = "FoldableSectionContent";
//#endregion
//#region src/components/SectionBlock/SectionV2.tsx
const SectionV2 = ({ trigger, children }) => {
	return /* @__PURE__ */ (0, react_jsx_runtime.jsx)("div", {
		className: "openui-section-v2",
		children: /* @__PURE__ */ (0, react_jsx_runtime.jsxs)("div", {
			className: "openui-section-v2-wrapper",
			children: [
				/* @__PURE__ */ (0, react_jsx_runtime.jsx)(require_components_Separator_index.Separator, { orientation: "horizontal" }),
				/* @__PURE__ */ (0, react_jsx_runtime.jsx)("div", {
					className: "openui-section-v2-header",
					children: /* @__PURE__ */ (0, react_jsx_runtime.jsx)("div", {
						className: "openui-section-v2-header-trigger",
						children: trigger
					})
				}),
				/* @__PURE__ */ (0, react_jsx_runtime.jsx)("div", {
					className: "openui-section-v2-content",
					children
				})
			]
		})
	});
};
//#endregion
exports.FoldableSectionContent = FoldableSectionContent;
exports.FoldableSectionItem = FoldableSectionItem;
exports.FoldableSectionRoot = FoldableSectionRoot;
exports.FoldableSectionTrigger = FoldableSectionTrigger;
exports.SectionV2 = SectionV2;
